---
title: "Impact of DNA source on genetic variant detection from human whole-genome sequencing data"
subtitle: "Microbiome taxonomic and functional analysis"
output:
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
---

```{R, setup_r, include=FALSE}
library(reticulate)
library(kableExtra)
library(knitr)
library(exactRankTests)
library(nlme)
library(dplyr)
library(ggplot2)
library(compositions)
library(vegan)

opts_knit$set(root.dir = '/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/')
#reticulate::py_config()
```

```{python, setup_python, results='hide'}
import os
import pandas as pd
import csv
import numpy as np
import math
from matplotlib.lines import Line2D
import matplotlib as mpl
from matplotlib.patches import Patch
import matplotlib.pyplot as plt
import random
from scipy.spatial import distance
from scipy import stats
from sklearn import manifold
from sklearn.decomposition import PCA
import statsmodels.stats.multitest as sm
```

# Summary

This file includes the results of a microbiome analysis performed on samples taken from four individuals that were originally used to determine the [**"Impact of DNA source on genetic variant detection from human whole-genome sequencing data"**](https://jmg.bmj.com/content/56/12/809).

This included blood, saliva and buccal samples taken from four individuals (blood samples were taken at a different time than saliva and buccal samples). Additionally, a methylation-based enrichment for eukaryotic DNA was performed on the saliva and buccal samples.

# Basic processing steps

## Obtaining reads

Fastq.gz files were downloaded from the ENA database, project accession number [PRJNA523344](https://www.ebi.ac.uk/ena/browser/view/PRJNA523344)

## Kneaddata

[Kneaddata](https://huttenhower.sph.harvard.edu/kneaddata/) was used for quality control and removal of human sequences. This included:<br/>
- **Trimmomatic 0.39**: "SLIDINGWINDOW:4:20 MINLEN:50‚Äù<br/>
- **Bowtie2** with the GRCh38_PhiX database (to remove human and PhiX reads): "--fast --dovetail"<br/>

## Reads kept after Kneaddata {.tabset}

```{python, basic_1, results='hide'}
#3 
#set up colors function (to get up to 120 colors, but with up to 40 unique colors)
def get_cols(num):
    colormap_20, colormap_40b, colormap_40c = mpl.cm.get_cmap('tab20', 256), mpl.cm.get_cmap('tab20b', 256), mpl.cm.get_cmap('tab20c', 256)
    norm, norm2 = mpl.colors.Normalize(vmin=0, vmax=19), mpl.colors.Normalize(vmin=20, vmax=39)
    m1, m2, m3 = mpl.cm.ScalarMappable(norm=norm, cmap=colormap_20), mpl.cm.ScalarMappable(norm=norm, cmap=colormap_40b), mpl.cm.ScalarMappable(norm=norm2, cmap=colormap_40c)
    colors_20 = [m1.to_rgba(a) for a in range(20)]
    colors_40 = [m2.to_rgba(a) for a in range(20)]+[m3.to_rgba(a) for a in range(20,40)]
    if num < 21: return colors_20
    elif num < 41: return colors_40
    else: return colors_40+colors_40+colors_40
#and colors and shapes for different participants and body sites
colors_dict, shapes_dict = {'Blood':'#900C3F', 'Saliva':'#016F85', 'Buccal':'#ff8300', 'Saliva_euk':'#02aed1', 'Buccal_euk':'#FFC300'}, {'Huref':'o', 'PGPC-0002':'^', 'PGPC-0005':'*', 'PGPC-0006':'s', 'PGPC-0050':'p'}
```

```{python, basic_2, results='hide'}
#4
#get numbers of reads for different steps
reads = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/read_counts.txt', sep='\t', index_col=3, header=0)
participant_dict, site_dict, full_name_dict = {}, {}, {}
samples = list(reads.index.values)
for s in samples:
    participant_dict[s] = reads.loc[s, 'Participant']
    site_dict[s] = reads.loc[s, 'Body site']
    full_name_dict[s] = reads.loc[s, 'Participant']+' '+reads.loc[s, 'Body site']
total_reads = pd.DataFrame(reads.loc[:, 'cat_reads'])
sample_names = [participant_dict[name]+' '+site_dict[name] for name in samples]
colors = [colors_dict[s] for s in list(reads.loc[:, 'Body site'].values)]
shapes = [shapes_dict[s] for s in list(reads.loc[:, 'Participant'].values)]
```

### Percentage reads remaining
```{python, basic_3, cache=TRUE, results='hide',fig.keep='all'}
#5
plt.figure(figsize=(10, 5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
plt.sca(ax1)
plt.bar(list(reads.index.values), reads.loc[:, 'Percentage'].values, color=colors, edgecolor='k')
plt.xticks(list(reads.index.values), sample_names, rotation=90)
plt.ylabel('Reads kept (%)')
plt.xlim([-0.5,20.5])

plt.sca(ax2)
plt.bar(list(reads.index.values), reads.loc[:, 'Percentage'].values, color=colors, edgecolor='k')
plt.semilogy()
plt.xticks(list(reads.index.values), sample_names, rotation=90)
plt.ylabel('Log reads kept (%)')
handles = [Patch(facecolor=colors_dict[color], edgecolor='k', label=color) for color in colors_dict]
ax2.legend(handles=handles, bbox_to_anchor=(1.4,1.05))
plt.xlim([-0.5,20.5])
plt.tight_layout()
plt.show()
```

### Number of reads remaining
```{python, basic_4, cache=TRUE, results='hide',fig.keep='all'}
#6
plt.figure(figsize=(10, 5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
plt.sca(ax1)
plt.bar(list(reads.index.values), reads.loc[:, 'cat_reads'].values, color=colors, edgecolor='k')
plt.xticks(list(reads.index.values), sample_names, rotation=90)
plt.xlim([-0.5,20.5])
plt.ylabel('Reads remaining')

plt.sca(ax2)
plt.bar(list(reads.index.values), reads.loc[:, 'cat_reads'].values, color=colors, edgecolor='k')
plt.semilogy()
plt.xticks(list(reads.index.values), sample_names, rotation=90)
plt.ylabel('Log reads remaining')
handles = [Patch(facecolor=colors_dict[color], edgecolor='k', label=color) for color in colors_dict]
ax2.legend(handles=handles, bbox_to_anchor=(1.4,1.05))
plt.xlim([-0.5,20.5])
plt.tight_layout()
plt.show()
```

### Table of reads remaining
```{python, basic_5, cache=TRUE}
#7
reads_remain = reads.loc[:, ['Percentage', 'cat_reads']].rename(index=full_name_dict, columns = {'cat_reads':'Number'})
```

```{r, basic_6, cache=TRUE}
#8
py$reads_remain %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "350px", height = "400px")
```

# Taxonomic profiling

The taxonomy has been profiled using:<br/>
1. [HUMAnN](https://huttenhower.sph.harvard.edu/humann)<br/>
    - HUMAnN2 and MetaPhlAn2<br/>
    - HUMAnN3 and MetaPhlAn3<br/>
2. [Kraken2](https://ccb.jhu.edu/software/kraken2/) with [Bracken](https://ccb.jhu.edu/software/bracken/)<br/>
    - GTDB (no confidence parameter set) - using the database constructed using [Struo](https://github.com/leylabmpi/Struo), release 89<br/>
    - GTDB (confidence = 0.1)<br/>
    - Minikraken v1 (no human genome, no confidence parameter set)<br/>
    - Minikraken v1 (no human genome, confidence = 0.1)<br/>
    - Minikraken v2 (with human genome, no confidence parameter set)<br/>
    - Minikraken v2 (with human genome, confidence = 0.1)<br/>
    - RefSeq Complete v93 (no confidence parameter set)<br/>
    - RefSeq Complete v93 (confidence = 0.1)<br/>

## MetaPhlAn2 taxonomy nMDS plots {.tabset}
```{python, tax_humann_1, cache=TRUE}
#9
#get the taxonomy file and sort it to strain and genus level
taxa = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_90/metaphlan_merged.tsv', sep='\t', header=0, index_col=0)
tax_names = list(taxa.index.values)
keeping = []
for a in range(len(tax_names)):
    if 't__' in tax_names[a]:
        keeping.append(True)
    elif 'unclassified' in tax_names[a]:
        keeping.append(True)
    else:
        keeping.append(False)
strain = taxa.loc[keeping, :]
strain_names = list(strain.index.values)
strain_dict = {}
for i in range(len(strain_names)):
    strain_dict[strain_names[i]] = strain_names[i].split('|s__')[0].split('|g__')[1]
genus = strain.rename(index=strain_dict)
genus = genus.groupby(by=genus.index, axis=0).sum()
```

```{python, tax_humann_2}
#10
#define the function that calculates the nmds plots
def transform_for_NMDS(df, dist_met='braycurtis'):
    X = df.iloc[0:].values
    y = df.iloc[:,0].values
    seed = np.random.RandomState(seed=3)
    X_true = X
    similarities = distance.cdist(X_true, X_true, dist_met)
    mds = manifold.MDS(n_components=2, max_iter=3000, eps=1e-9, random_state=seed,
                   dissimilarity="precomputed", n_jobs=1)
    #print(similarities)
    pos = mds.fit(similarities).embedding_
    nmds = manifold.MDS(n_components=2, metric=False, max_iter=3000, eps=1e-12,
                        dissimilarity="precomputed", random_state=seed, n_jobs=1,
                        n_init=1)
    npos = nmds.fit_transform(similarities, init=pos)
    # Rescale the data
    pos *= np.sqrt((X_true ** 2).sum()) / np.sqrt((pos ** 2).sum())
    npos *= np.sqrt((X_true ** 2).sum()) / np.sqrt((npos ** 2).sum())
    # Rotate the data
    clf = PCA()
    X_true = clf.fit_transform(X_true)
    pos = clf.fit_transform(pos)
    npos = clf.fit_transform(npos)
    return pos, npos, nmds.stress_
```
```{python, tax_humann_2b, cache=TRUE}
strain_t = strain.transpose()
genus_t = genus.transpose()

handles1 = [Patch(facecolor=colors_dict[color], edgecolor='k', label=color) for color in colors_dict]
handles2 = [Line2D([0], [0], marker=shapes_dict[shape], color='w', label=shape, markerfacecolor='k', markersize=15) for shape in shapes_dict]
```

### Bray-Curtis distance
```{python, tax_humann_3, cache=TRUE, results='hide',fig.keep='all'}
#11
strain_pos, strain_npos, strain_stress = transform_for_NMDS(strain_t, 'braycurtis')
genus_pos, genus_npos, genus_stress = transform_for_NMDS(genus_t, 'braycurtis')

plt.figure(figsize=(10,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
for a in range(len(strain_npos)):
    ax1.scatter(strain_npos[a,0], strain_npos[a,1], marker=shapes[a], color=colors[a], s=100)
    ax2.scatter(genus_npos[a,0], genus_npos[a,1], marker=shapes[a],  color=colors[a], s=100)
ax1.set_xlabel('nMDS1')
ax2.set_xlabel('nMDS1')
ax1.set_ylabel('nMDS2')
ax1.set_title('Strain')
ax2.set_title('Genus')

ax2.legend(handles=handles1+handles2, bbox_to_anchor=(1,1))
plt.tight_layout()
plt.show()
```

### Euclidean distance
```{python, tax_humann_4, cache=TRUE, results='hide',fig.keep='all'}
#12
strain_pos, strain_npos, strain_stress = transform_for_NMDS(strain_t, 'euclidean')
genus_pos, genus_npos, genus_stress = transform_for_NMDS(genus_t, 'euclidean')

plt.figure(figsize=(10,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
for a in range(len(strain_npos)):
    ax1.scatter(strain_npos[a,0], strain_npos[a,1], marker=shapes[a], color=colors[a], s=100)
    ax2.scatter(genus_npos[a,0], genus_npos[a,1], marker=shapes[a],  color=colors[a], s=100)
ax1.set_xlabel('nMDS1')
ax2.set_xlabel('nMDS1')
ax1.set_ylabel('nMDS2')
ax1.set_title('Strain')
ax2.set_title('Genus')

ax2.legend(handles=handles1+handles2, bbox_to_anchor=(1.4,1))
plt.tight_layout()
plt.show()
```

### Jaccard distance
```{python, tax_humann_5, cache=TRUE, results='hide',fig.keep='all'}
#13
strain_pos, strain_npos, strain_stress = transform_for_NMDS(strain_t, 'jaccard')
genus_pos, genus_npos, genus_stress = transform_for_NMDS(genus_t, 'jaccard')

plt.figure(figsize=(10,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
for a in range(len(strain_npos)):
    ax1.scatter(strain_npos[a,0], strain_npos[a,1], marker=shapes[a], color=colors[a], s=100)
    ax2.scatter(genus_npos[a,0], genus_npos[a,1], marker=shapes[a],  color=colors[a], s=100)
ax1.set_xlabel('nMDS1')
ax2.set_xlabel('nMDS1')
ax1.set_ylabel('nMDS2')
ax1.set_title('Strain')
ax2.set_title('Genus')

ax2.legend(handles=handles1+handles2, bbox_to_anchor=(1.4,1))
plt.tight_layout()
plt.show()
```

## MetaPhlAn2 taxonomy relative abundance {.tabset}

### Kingdom

Here the relative abundance of taxa calulated by MetaPhlAn2 are plotted at the Kingdom level for each sample. 

```{python, tax_humann_6, cache=TRUE, results='hide',fig.keep='all'}
#14
plt.figure(figsize=(7,5))
ax1 = plt.subplot(111)
plt.bar(list(taxa.columns.values), taxa.loc['k__Viruses', :].values, color='#C70039', edgecolor='k')
plt.bar(list(taxa.columns.values), taxa.loc['k__Bacteria', :].values, bottom=taxa.loc['k__Viruses', :].values, color='#026B81', edgecolor='k')
#plt.xticks(list(taxa.columns.values), sample_names, rotation=90)
empty = []
for x in range(0,21):
    empty.append('')
    ax1.text(x, -2, sample_names[x], color=colors[x], rotation=90, va='top', ha='center')
plt.xticks(range(0, 21), empty)
plt.xlim([-0.5,20.5])

handles = [Patch(facecolor='#C70039', edgecolor='k', label='Viruses'), Patch(facecolor='#026B81', edgecolor='k', label='Bacteria')]
plt.legend(handles=handles, bbox_to_anchor=(1,1.05))
plt.tight_layout()
plt.show()
```

### Genus

Here the relative abundance of taxa calulated by MetaPhlAn2 are plotted at the Genus level for each sample. Genera with below 1% maximum relative abundance have been removed.
```{python, tax_humann_7, results='hide',fig.keep='all', cache=TRUE}
#15
genus = genus[genus.max(axis=1) > 1]
genera = list(genus.index.values)
plt.figure(figsize=(10,5))
ax1 = plt.subplot(111)
gen_colors = get_cols(len(genus.index.values))
handles = []
for g in range(len(genera)):
    this_gen = genus.loc[genera[g], :].values
    if g == 0:
        ax1.bar(list(genus.columns.values), this_gen, color=gen_colors[g], edgecolor='k')
        total = this_gen
    else:
        ax1.bar(list(genus.columns.values), this_gen, bottom=total, color=gen_colors[g], edgecolor='k')
        total = this_gen+total
    handles.append(Patch(facecolor=gen_colors[g], edgecolor='k', label=genera[g]))
empty = []
for x in range(0,21):
    empty.append('')
    ax1.text(x, -2, sample_names[x], color=colors[x], rotation=90, va='top', ha='center')
plt.xticks(range(0, 21), empty)
plt.xlim([-0.5,20.5])
plt.legend(handles=handles, bbox_to_anchor=(1,1.05), ncol=2)
plt.tight_layout()
plt.show()
```

## MetaPhlAn3 taxonomy nMDS plots {.tabset}

```{python, tax_humann3_1, cache=TRUE}
#get the taxonomy files and sort them to strain and genus level
files = sorted(os.listdir('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/humann3_final_out/bugs_lists'))
all_genus, samples, all_strain = [], [], []
for f in files:
  #if f != 'SRR8595490_metaphlan_bugs_list.tsv': continue
  samples.append(f.split('_')[0])
  fn = '/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/humann3_final_out/bugs_lists/'+f
  taxa = pd.read_csv(fn, header=0, index_col=0, sep='\t')
  taxa.drop(['additional_species', 'NCBI_tax_id'], axis=1, inplace=True)
  taxa.rename(columns={'relative_abundance':f.split('_')[0]}, inplace=True)
  if taxa.shape[0] == 1:
    all_genus.append(taxa)
    all_strain.append(taxa)
    continue
  tax_names = list(taxa.index.values)
  keeping = []
  unclassified, ind = [], []
  for a in range(len(tax_names)):
    if 's__' in tax_names[a]:
        keeping.append(True)
    elif 'UNKNOWN' in tax_names[a]:
        keeping.append(True)
    elif 'unclassified' in tax_names[a].split('__')[-1]:
        keeping.append(True)
        unclassified.append(tax_names[a])
        ind.append(a)
    else:
        keeping.append(False)
  for a in range(len(unclassified)):
    for b in range(len(unclassified)):
      if unclassified[a] in unclassified[b]:
        keeping[ind[a]] = False
  strain = taxa.loc[keeping, :]
  all_strain.append(strain)
  strain_names = list(strain.index.values)
  strain_dict = {}
  for i in range(len(strain_names)):
    if 's__' in strain_names[i]:
      strain_dict[strain_names[i]] = strain_names[i].split('|s__')[0].split('|g__')[1].replace('_', ' ')
    else:
      strain_dict[strain_names[i]] = strain_names[i].split('__')[1].replace('_', ' ')
  genus = strain.rename(index=strain_dict)
  genus = genus.groupby(by=genus.index, axis=0).sum()
  all_genus.append(genus)
all_genus = pd.concat(all_genus).fillna(value=0)
all_genus = all_genus.groupby(by=all_genus.index, axis=0).sum()
all_strain = pd.concat(all_strain).fillna(value=0)
all_strain = all_strain.groupby(by=all_strain.index, axis=0).sum()
```

```{python, tax_humann3_2, cache=TRUE}
strain_t = all_strain.transpose()
genus_t = all_genus.transpose()

handles1 = [Patch(facecolor=colors_dict[color], edgecolor='k', label=color) for color in colors_dict]
handles2 = [Line2D([0], [0], marker=shapes_dict[shape], color='w', label=shape, markerfacecolor='k', markersize=15) for shape in shapes_dict]
```
<br/>

### Bray-Curtis distance

```{python, tax_humann3_3, cache=TRUE, results='hide',fig.keep='all'}
strain_pos, strain_npos, strain_stress = transform_for_NMDS(strain_t, 'braycurtis')
genus_pos, genus_npos, genus_stress = transform_for_NMDS(genus_t, 'braycurtis')

plt.figure(figsize=(10,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
for a in range(len(strain_npos)):
    ax1.scatter(strain_npos[a,0], strain_npos[a,1], marker=shapes[a], color=colors[a], s=100)
    ax2.scatter(genus_npos[a,0], genus_npos[a,1], marker=shapes[a],  color=colors[a], s=100)
ax1.set_xlabel('nMDS1')
ax2.set_xlabel('nMDS1')
ax1.set_ylabel('nMDS2')
ax1.set_title('Strain')
ax2.set_title('Genus')

ax2.legend(handles=handles1+handles2, bbox_to_anchor=(1,1))
plt.tight_layout()
plt.show()
```

<br/>

### Euclidean distance

```{python, tax_humann3_4, cache=TRUE, results='hide',fig.keep='all'}
strain_pos, strain_npos, strain_stress = transform_for_NMDS(strain_t, 'euclidean')
genus_pos, genus_npos, genus_stress = transform_for_NMDS(genus_t, 'euclidean')

plt.figure(figsize=(10,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
for a in range(len(strain_npos)):
    ax1.scatter(strain_npos[a,0], strain_npos[a,1], marker=shapes[a], color=colors[a], s=100)
    ax2.scatter(genus_npos[a,0], genus_npos[a,1], marker=shapes[a],  color=colors[a], s=100)
ax1.set_xlabel('nMDS1')
ax2.set_xlabel('nMDS1')
ax1.set_ylabel('nMDS2')
ax1.set_title('Strain')
ax2.set_title('Genus')

ax2.legend(handles=handles1+handles2, bbox_to_anchor=(1.4,1))
plt.tight_layout()
plt.show()
```

<br/>

### Jaccard distance

```{python, tax_humann3_5, cache=TRUE, results='hide',fig.keep='all'}
strain_pos, strain_npos, strain_stress = transform_for_NMDS(strain_t, 'jaccard')
genus_pos, genus_npos, genus_stress = transform_for_NMDS(genus_t, 'jaccard')

plt.figure(figsize=(10,5))
ax1, ax2 = plt.subplot(121), plt.subplot(122)
for a in range(len(strain_npos)):
    ax1.scatter(strain_npos[a,0], strain_npos[a,1], marker=shapes[a], color=colors[a], s=100)
    ax2.scatter(genus_npos[a,0], genus_npos[a,1], marker=shapes[a],  color=colors[a], s=100)
ax1.set_xlabel('nMDS1')
ax2.set_xlabel('nMDS1')
ax1.set_ylabel('nMDS2')
ax1.set_title('Strain')
ax2.set_title('Genus')

ax2.legend(handles=handles1+handles2, bbox_to_anchor=(1.4,1))
plt.tight_layout()
plt.show()
```

<br/>

## MetaPhlAn3 taxonomy relative abundance {.tabset}

<br/>

### Kingdom

I haven't included this plot for MetaPhlAn3 as all taxa (that are classified) are classified as bacteria. There are now no viruses in the output, but those samples that were all viral reads in MetaPhlAn2 are now all unclassified/'unknown'.

<br/>

### Genus

Here the relative abundance of taxa calulated by MetaPhlAn3 are plotted at the Genus level for each sample. Genera with below 1% maximum relative abundance have been removed.

```{python, tax_humann3_7, results='hide',fig.keep='all', cache=TRUE}
genus = all_genus[all_genus.max(axis=1) > 1]
genera = list(genus.index.values)
plt.figure(figsize=(10,5))
ax1 = plt.subplot(111)
gen_colors = get_cols(len(genus.index.values))
handles = []
for g in range(len(genera)):
    this_gen = genus.loc[genera[g], :].values
    if g == 0:
        ax1.bar(list(genus.columns.values), this_gen, color=gen_colors[g], edgecolor='k')
        total = this_gen
    else:
        ax1.bar(list(genus.columns.values), this_gen, bottom=total, color=gen_colors[g], edgecolor='k')
        total = this_gen+total
    if 'XIII' in genera[g]:
      genera[g] = genera[g].replace('XIII', 'XIII\n')
    handles.append(Patch(facecolor=gen_colors[g], edgecolor='k', label=genera[g]))
empty = []
for x in range(0,21):
    empty.append('')
    ax1.text(x, -2, sample_names[x], color=colors[x], rotation=90, va='top', ha='center')
plt.xticks(range(0, 21), empty)
plt.xlim([-0.5,20.5])
plt.legend(handles=handles, bbox_to_anchor=(1,1.05), ncol=2)
plt.tight_layout()
plt.show()
```

<br/>

## Kraken2 reads classified {.tabset}

```{python, tax_kraken_1, results='hide'}
#16
#get all samples into dataframes based on the database that they use
folders = sorted(os.listdir('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2'))
del folders[0]
kraken_columns = {0:'Percent fragments clade', 1:'Number fragments clade', 2:'Number fragments taxon', 3:'Level', 4:'NCBI ID', 5:'Taxon name'}
kraken_all_db, bracken_all_db, all_domains = [], [], {}
for fol in folders:
    if fol == 'db_genera' or fol == 'db_genera_in_common':
      continue
    if not os.path.isdir('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/'+fol):
        continue
    bracken, kraken, bracken_kreport = [], [], []
    bracken_pd, kraken_pd = [], []
    for fi in sorted(os.listdir('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/'+fol)):
        if fi[-7:] == 'bracken':
            bracken.append(fi)
        elif fi[-7:] == 'kreport' and 'bracken' not in fi:
            kraken.append(fi)
        elif fi[-7:] == 'kreport':
            bracken_kreport.append(fi)
    for bk in bracken_kreport:
        with open('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/'+fol+'/'+bk, 'rU') as f:
            bk = []
            domains = {}
            this_domain, domain_name = [], ''
            for row in csv.reader(f, delimiter='\t'):
                bk.append(row)
                row[5] = row[5].lstrip()
                if row[3] == 'D':
                    if domain_name != '':
                        domains[domain_name] = this_domain
                    this_domain, domain_name = [], row[5]
                else:
                    if row[3] != 'R' and row[3] != 'U' and 'D' not in row[3]:
                        this_domain.append(row[5])
            domains[domain_name] = this_domain
            for domain in domains:
                if domain in all_domains:
                    all_domains[domain] = list(set(all_domains[domain]+domains[domain]))
                else:
                    all_domains[domain] = list(set(domains[domain]))
    for b in bracken:
        if len(b) > 22:
            continue
        sample = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/'+fol+'/'+b, sep='\t', header=0, index_col=0)
        b = b.replace('_150', '')
        sample.drop(['taxonomy_id', 'taxonomy_lvl', 'kraken_assigned_reads', 'added_reads', 'fraction_total_reads'], axis=1, inplace=True)
        sample.rename(columns={'new_est_reads':b[:-8]}, inplace=True)
        bracken_pd.append(sample)
    for k in kraken:
        sample = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/'+fol+'/'+k, sep='\t', header=None, index_col=3)
        sample = sample.loc[['U', 'D'], :]
        sample = sample.rename(columns=kraken_columns).drop(['Number fragments taxon', 'NCBI ID'], axis=1).rename(columns={'Percent fragments clade':k[:-8]+'_percent', 'Number fragments clade':k[:-8]+'_reads'}).set_index('Taxon name')
        taxa = list(sample.index.values)
        taxa_dict = {}
        for t in taxa:
            taxa_dict[t] = t.replace(' ', '')
        sample = sample.rename(index=taxa_dict)
        kraken_pd.append(sample)
    bracken = pd.concat(bracken_pd, join='outer')
    kraken = pd.concat(kraken_pd, join='outer')
    kraken = kraken.rename(index={'d__Bacteria':'Bacteria', 'd__Archaea':'Archaea'})
    kraken = kraken.groupby(by=kraken.index, axis=0).sum()
    bracken = bracken.groupby(by=bracken.index, axis=0).sum().fillna(value=0)
    kraken_all_db.append(kraken), bracken_all_db.append(bracken)
```
```{python, tax_kraken_2, results='hide', cache=TRUE}
#17
x1 = [x for x in range(21)]
x2 = [x+0.3 for x in range(21)]
tax_plotting = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses', 'unclassified']
color_plotting = ['#EDBB99', '#5499C7', '#7DCEA0', '#F7DC6F', '#CCD1D1']
tax_paper = ['Bacteria', 'Eukaryota', 'Other', 'Unclassified']
color_paper = ['#5499C7', '#7DCEA0', '#CD6155', '#CCD1D1']
from_paper = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/from_paper.csv', header=0, index_col=0)

def get_summary_reads(kraken_db):
    fig = plt.figure(figsize=(15,15))
    ax1, ax2, ax3, ax4, ax5 = plt.subplot(321), plt.subplot(322), plt.subplot(323), plt.subplot(324), plt.subplot(325)
    ax1.set_title('Minikraken V1\n(without human genome)'), ax2.set_title('Minikraken V2\n(with human genome)'), ax3.set_title('GTDB'), ax4.set_title('RefSeq Complete V93')
    ax5.set_title('From paper')
    ax_plot = [ax3, ax3, ax1, ax1, ax2, ax2, ax4, ax4]
    x_plot = [x1, x2, x1, x2, x1, x2, x1, x2]
    axs = [ax1, ax2, ax3, ax4, ax5]
    for s in range(len(samples)):
        if s == 0:
            continue
        bottom = 0
        for t in range(len(tax_paper)):
            ax5.bar(x1[s], from_paper.loc[tax_paper[t], samples[s]], bottom=bottom, color=color_paper[t], edgecolor='k', width=0.6)
            bottom += from_paper.loc[tax_paper[t], samples[s]]
    for db in range(len(kraken_db)):
        ax_using = ax_plot[db]
        x = x_plot[db]
        db = kraken_db[db]
        handles = []
        for tax in range(len(tax_plotting)):
            handles.append(Patch(facecolor=color_plotting[tax], edgecolor='k', label=tax_plotting[tax]))
            tax = tax_plotting[tax]
            if tax not in list(db.index.values):
                db.loc[tax] = [0 for i in range(db.shape[1])]
        handles.append(Patch(facecolor=color_paper[2], edgecolor='k', label='Other'))
        db = db.fillna(value=0)
        for s in range(len(samples)):
            bottom = 0
            for t in range(len(tax_plotting)):
                prop = db.loc[tax_plotting[t], samples[s]+'_reads']
                cat = total_reads.loc[samples[s], 'cat_reads']
                prop = (prop/cat)*100
                ax_using.bar(x[s], prop, bottom=bottom, color=color_plotting[t], edgecolor='k', width=0.3)
                bottom += prop
    ax2.legend(handles=handles, bbox_to_anchor=(1,1.05))
    for ax in axs:
        plt.sca(ax)
        plt.xticks(x1, ['' for x in x1])
        plt.ylim([0, 100])
        plt.xlim([-0.5, 20.5])
    for x in x1:
        ax5.text(x, -2, sample_names[x], color=colors[x], rotation=90, va='top', ha='center')
        ax4.text(x, -2, sample_names[x], color=colors[x], rotation=90, va='top', ha='center')
    ax1.set_ylabel('Classified (%)'), ax3.set_ylabel('Classified (%)'), ax5.set_ylabel('Classified(%)')
    #plt.tight_layout()
    return

def get_summary_bacteria(kraken_db):
    tax_plotting = ['Bacteria']
    alpha = ['#5499C7', '#F1C40F', '#5499C7', '#F1C40F', '#5499C7', '#F1C40F', '#5499C7', '#F1C40F']
    
    fig = plt.figure(figsize=(10,8))
    ax1, ax2, ax3, ax4 = plt.subplot(221), plt.subplot(222), plt.subplot(223), plt.subplot(224)
    ax1.set_title('Minikraken V1\n(without human genome)'), ax2.set_title('Minikraken V2\n(with human genome)'), ax3.set_title('GTDB'), ax4.set_title('RefSeq Complete V93')
    ax_plot = [ax3, ax3, ax1, ax1, ax2, ax2, ax4, ax4]
    x_plot = [x1, x2, x1, x2, x1, x2, x1, x2]
    axs = [ax1, ax2, ax3, ax4]
    
    for db in range(len(kraken_db)):
        ax_using = ax_plot[db]
        x = x_plot[db]
        alp = alpha[db]
        db = kraken_db[db]
        for tax in range(len(tax_plotting)):
            tax = tax_plotting[tax]
            if tax not in list(db.index.values):
                db.loc[tax] = [0 for i in range(db.shape[1])]
        db = db.fillna(value=0)
        for s in range(len(samples)):
            bottom = 0
            for t in range(len(tax_plotting)):
                prop = db.loc[tax_plotting[t], samples[s]+'_reads']
                cat = total_reads.loc[samples[s], 'cat_reads']
                ax_using.bar(x[s], prop, bottom=bottom, color=alp, edgecolor='k', width=0.3)
                bottom += prop
    handles = []
    handles.append(Patch(facecolor=alpha[0], edgecolor='k', label='No confidence value'))
    handles.append(Patch(facecolor=alpha[1], edgecolor='k', label='Confidence=0.1'))
    ax2.legend(handles=handles, bbox_to_anchor=(1.6,1.03))
    for ax in axs:
        plt.sca(ax)
        plt.xticks(x1, ['' for x in x1])
        plt.semilogy()
        plt.xlim([-0.5, 20.5])
        #plt.ylim([0, 100])
        plt.xlim([-0.5, 20.5])
    for x in x1:
        pl = ((1/21)*(x+1))-0.02
        ax3.text(pl, -0.03, sample_names[x], color=colors[x], rotation=90, va='top', ha='center', transform=ax3.transAxes)
        ax4.text(pl, -0.03, sample_names[x], color=colors[x], rotation=90, va='top', ha='center', transform=ax4.transAxes)
    ax1.set_ylabel('Number of reads'), ax3.set_ylabel('Number of reads')
    #plt.tight_layout()
    return
```

### Percent classified

A summary of the percentage of reads classified as different domains with different databases.
Note that the 'From paper' plot uses the classifications given in the original paper, where 10,000 unmapped reads were classified using BLAST searches of the NCBI database.

```{python, tax_kraken_3, cache=TRUE, results='hide',fig.keep='all'}
#18
get_summary_reads(kraken_all_db)
plt.show()
```

### Summary of number of bacteria

Summary of the number of reads that are classified as bacteria by each database.

```{python, tax_kraken_4, cache=TRUE, results='hide',fig.keep='all'}
#19
get_summary_bacteria(kraken_all_db)
plt.tight_layout()
plt.show()
```

## Kraken2 taxonomy nMDS plots {.tabset}

These first plots are all separately with the confidence parameter set. See the last tab for those without the confidence parameter set. 

```{python, tax_kraken_5, results='hide', cache=TRUE}
#20
db_names = ['gtdb', 'gtdb_conf', 'minikraken', 'minikraken_conf', 'minikraken_human', 'minikraken_human_conf', 'refseq', 'refseq_conf']
bacteria = all_domains['Bacteria']+all_domains['d__Bacteria']
genera, gen_names, genera_1, gen_names_1, strain, gen_sums = [], [], [], [], [], []

for db in range(len(bracken_all_db)):
    d = int(db)
    db = bracken_all_db[db]
    species = list(db.index.values)
    keeping = []
    species_dict = {}
    for sp in species:
        if sp in bacteria:
            keeping.append(True)
            new_sp = sp.split('__')
            if len(new_sp) > 1:
                new_sp = new_sp[1]
            else:
                new_sp = new_sp[0]
            species_dict[sp] = new_sp.split(' ')[0].replace("'", '')
        else:
            keeping.append(False)
    in_len = db.shape[0]
    db = db.loc[keeping, :]
    strain.append(db)
    db = db.rename(index=species_dict)
    db = db.groupby(by=db.index, axis=0).sum()
    sums = db.sum(axis=0)
    gen_sums.append(sums)
    db = db.divide(sums, axis=1).multiply(100)
    genera.append(db)
    db.to_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera/'+db_names[d]+'.csv')
    gen_names = gen_names+list(db.index.values)
    db = db[db.max(axis=1) > 1]
    genera_1.append(db)
    gen_names_1 = gen_names_1+list(db.index.values)
gen_names = list(set(gen_names))
gen_names_1 = list(set(gen_names_1))
```

```{python, tax_kraken_6, results='hide', cache=TRUE}
#21
def plot_four_nmds(dbs, metric, name):
    fig = plt.figure(figsize=(15,10))
    #fig.suptitle(name+metric+'\n\n\n')
    ax1, ax2, ax3, ax4 = plt.subplot(221), plt.subplot(222), plt.subplot(223), plt.subplot(224)
    axs = [ax3, ax1, ax2, ax4]
    ax1.set_title('Minikraken V1\n(without human genome)'), ax2.set_title('Minikraken V2\n(with human genome)'), ax3.set_title('GTDB'), ax4.set_title('RefSeq Complete V93')
    
    for db in range(len(dbs)):
        n = db
        db = dbs[db].transpose()
        pos, npos, stress = transform_for_NMDS(db, metric)
        for a in range(len(npos)):
            axs[n].scatter(npos[a,0], npos[a,1], marker=shapes[a], color=colors[a], s=100, edgecolor='k')
        axs[n].set_xlabel('nMDS1')
        axs[n].set_ylabel('nMDS2')
    handles1 = [Patch(facecolor=colors_dict[color], edgecolor='k', label=color) for color in colors_dict]
    handles2 = [Line2D([0], [0], marker=shapes_dict[shape], color='w', label=shape, markerfacecolor='k', markersize=15) for shape in shapes_dict]

    ax2.legend(handles=handles1+handles2, bbox_to_anchor=(1,1))
    plt.tight_layout()
    return
```
### Bray-Curtis distance strain level
```{python, tax_kraken_7, results='hide',fig.keep='all', cache=TRUE}
#22
plot_four_nmds([strain[1], strain[3], strain[5], strain[7]], 'braycurtis', 'NMDS confidence=0.1 strain ')
plt.show()
```

### Bray-Curtis distance genus level
```{python, tax_kraken_8, results='hide',fig.keep='all', cache=TRUE}
#23
plot_four_nmds([genera[1], genera[3], genera[5], genera[7]], 'braycurtis', 'NMDS confidence=0.1 genera ')
plt.show()
```

### Euclidean distance strain level
```{python, tax_kraken_9, results='hide',fig.keep='all', cache=TRUE}
#24
plot_four_nmds([strain[1], strain[3], strain[5], strain[7]], 'euclidean', 'NMDS confidence=0.1 strain ')
plt.show()
```

### Euclidean distance genus level
```{python, tax_kraken_10, results='hide',fig.keep='all', cache=TRUE}
#25
plot_four_nmds([genera[1], genera[3], genera[5], genera[7]], 'euclidean', 'NMDS confidence=0.1 genera ')
plt.show()
```

### Jaccard distance strain level
```{python, tax_kraken_11, results='hide',fig.keep='all', cache=TRUE}
#26
plot_four_nmds([strain[1], strain[3], strain[5], strain[7]], 'jaccard', 'NMDS confidence=0.1 strain ')
plt.show()
```

### Jaccard distance genus level
```{python, tax_kraken_12, results='hide',fig.keep='all', cache=TRUE}
#27
plot_four_nmds([genera[1], genera[3], genera[5], genera[7]], 'jaccard', 'NMDS confidence=0.1 genera ')
plt.show()
```

### All plots with no confidence parameter set

**Bray-curtis distance at strain level**
```{python, tax_kraken_13, results='hide',fig.keep='all', cache=TRUE}
#28
plot_four_nmds([strain[0], strain[2], strain[4], strain[6]], 'braycurtis', 'NMDS no confidence strain ')
plt.show()
```
**Bray-curtis distance at genus level**
```{python, tax_kraken_14, results='hide',fig.keep='all', cache=TRUE}
#29
plot_four_nmds([genera[0], genera[2], genera[4], genera[6]], 'braycurtis', 'NMDS no confidence genera ')
plt.show()
```
**Euclidean distance at strain level**
```{python, tax_kraken_15, results='hide',fig.keep='all', cache=TRUE}
#30
plot_four_nmds([strain[0], strain[2], strain[4], strain[6]], 'euclidean', 'NMDS no confidence strain ')
plt.show()
```
**Euclidean distance at genus level**
```{python, tax_kraken_16, results='hide',fig.keep='all', cache=TRUE}
#31
plot_four_nmds([genera[0], genera[2], genera[4], genera[6]], 'euclidean', 'NMDS no confidence genera ')
plt.show()
```
**Jaccard distance at strain level**
```{python, tax_kraken_17, results='hide',fig.keep='all', cache=TRUE}
#32
plot_four_nmds([strain[0], strain[2], strain[4], strain[6]], 'jaccard', 'NMDS no confidence strain ')
plt.show()
```
**Jaccard distance at genus level**
```{python, tax_kraken_18, results='hide',fig.keep='all', cache=TRUE}
#33
plot_four_nmds([genera[0], genera[2], genera[4], genera[6]], 'jaccard', 'NMDS no confidence genera ')
plt.show()
```

## Kraken2 taxonomy relative abundance {.tabset}

These plots are now only calculated for the classifications that used confidence = 0.1. Genera with below 1% maximum relative abundance are removed and the numbers in brackets are the number of reads that were classified as bacteria. 

```{python, tax_kraken_19, results='hide', cache=TRUE}
#34
db_names = ['gtdb', 'gtdb_conf', 'minikraken', 'minikraken_conf', 'minikraken_human', 'minikraken_human_conf', 'refseq', 'refseq_conf']
#bacteria = all_domains['Bacteria']+all_domains['d__Bacteria']
#genera, gen_names, genera_1, gen_names_1, strain, gen_sums = [], [], [], [], [], []
gen_names_1 = sorted(gen_names_1)
def plot_genera(db, sums, tax_cols, gen_names_1, dname):
    plt.figure(figsize=(10,5))
    ax1 = plt.subplot(111)
    bottom = [0 for x in range(len(db.columns))]
    handles = []
    for g in range(len(gen_names_1)):
        if gen_names_1[g] in db.index.values:
            this_row = db.loc[gen_names_1[g], :].values
            ax1.bar(x1, this_row, bottom=bottom, color=tax_cols[g], edgecolor='k')
            handles.append(Patch(facecolor=tax_cols[g], edgecolor='k', label=gen_names_1[g]))
            for b in range(len(bottom)):
                bottom[b] += this_row[b]
    ax1.legend(handles=handles, bbox_to_anchor=(1, 1.03), ncol=3)
    plt.xticks(x1, ['' for x in x1])
    plt.ylabel('Relative abundance(%)')
    plt.xlim([-0.5, 20.5])
    plt.ylim([0, 100])
    for x in x1:
        n = str(int(sums[samples[x]]))
        ax1.text(x, -2, sample_names[x]+' ('+n+')', color=colors[x], rotation=90, va='top', ha='center')
    plt.tight_layout()
    return

gen_plot = [genera_1[1], genera_1[3], genera_1[5], genera_1[7]]
db_name = ['GTDB', 'Minikraken V1', 'Minikraken V2', 'RefSeq Complete V93']
all_sums = [gen_sums[1], gen_sums[3], gen_sums[5], gen_sums[7]]
tax_cols = get_cols(len(gen_names_1))
```

### Minikraken v1
```{python, tax_kraken_20, results='hide',fig.keep='all', cache=TRUE}
#35
plot_genera(gen_plot[1], all_sums[1], tax_cols, gen_names_1, db_name[1])
plt.show()
```

### Minikraken v2
```{python, tax_kraken_21, results='hide',fig.keep='all', cache=TRUE}
#36
plot_genera(gen_plot[2], all_sums[2], tax_cols, gen_names_1, db_name[2])
plt.show()
```

### GTDB
```{python, tax_kraken_22, results='hide',fig.keep='all', cache=TRUE}
#37
plot_genera(gen_plot[0], all_sums[0], tax_cols, gen_names_1, db_name[0])
plt.show()
```

### RefSeq Complete v93
```{python, tax_kraken_23, results='hide', fig.keep='all', cache=TRUE}
#38
plot_genera(gen_plot[3], all_sums[3], tax_cols, gen_names_1, db_name[3])
plt.show()
```

## Kraken2 similarities and differences between body sites {.tabset}

Here I have carried out ANCOM2 tests for differential abundance of genera between body sites. All genera are then plotted on a heatmap with mean values for each body site and stars to denote significant differences as determined by ANCOM.

While many of these results vary between databases, it is worth noting that some differences are consistent, e.g.:<br/>
1. *Prevotella* are always more abundant in saliva samples<br/>
2. *Streptococcus* are always more abundant in buccal samples<br/>
3. *Klebsiella* (where present) are always more abundant in blood samples<br/>

```{python, tax_kraken_24, results='hide', fig.keep='all', cache=TRUE}
#39
def get_differences(new_genus, db_name):
    new_genus = new_genus.drop(['SRR8595488'], axis=1)
    new_genus = new_genus[new_genus.max(axis=1) > 0.1]
    samples = list(new_genus.columns)
    
    list_comparisons = [['Blood', 'Saliva'], ['Blood', 'Buccal'], ['Saliva', 'Buccal'], ['Saliva', 'Saliva_euk'], ['Buccal', 'Buccal_euk'], ['Blood', 'Saliva_euk'], ['Blood', 'Buccal_euk'], ['Saliva_euk', 'Buccal_euk']]
    comparisons, metadata, comp_len = [], [], []
    for a in range(len(list_comparisons)):
        keeping = []
        this_md = []
        for b in range(len(samples)):
            if site_dict[samples[b]] in list_comparisons[a]:
                keeping.append(True)
                this_md.append([samples[b], site_dict[samples[b]]])
            else:
                keeping.append(False)
        this_comp = new_genus.loc[:, keeping]
        this_comp = this_comp[this_comp.max(axis=1) > 0.1]
        comparisons.append(this_comp)
        this_md = pd.DataFrame(this_md, columns=['Samples', 'Groups'])
        #this_md.set_index('Samples', inplace=True)
        metadata.append(this_md)
        comp_len.append(this_comp.shape[0])
    new_genus = new_genus.rename(columns=site_dict, inplace=False)
    return comparisons, metadata, new_genus, comp_len
db_name = ['GTDB', 'Minikraken V1', 'Minikraken V2', 'RefSeq Complete V93']
comparison_names = [r'Blood vs saliva', r'Blood vs buccal', r'Saliva vs buccal', r'Saliva vs saliva_euk', r'Buccal vs buccal_euk', r'Blood vs saliva_euk', r'Blood vs buccal_euk', r'Saliva_euk vs buccal_euk']
```

```{r, tax_kraken_25, results='hide'}
source("/Users/robynwright/Documents/OneDrive/Github/R-notebooks/ancom_v2.1.R")

get_ancom <- function(ft, md) {
all_ancom = list()
for (a in 1:8){
  feature_table = ft[a]
  meta_data = md[a]
  process = feature_table_pre_process(feature_table, meta_data, 'Samples', 'Groups', lib_cut=10, neg_lb=TRUE)
  ancom_out = ANCOM(process$feature_table, process$meta_data, process$struc_zero, main_var='Groups')
  all_ancom[[a]] <- ancom_out$out
}
return(all_ancom)
}

```

```{python, tax_kraken_26, results='hide'}
def sort_ancom_results(r_ancom):
  ancom_lists_09, ancom_lists_08, ancom_lists_07, ancom_lists_06 = [], [], [], []
  for a in range(len(r_ancom)):
    this_sig_09, this_sig_08, this_sig_07, this_sig_06 = [], [], [], []
    r_ancom[a].set_index('taxa_id', inplace=True)
    all_sp = list(r_ancom[a].index.values)
    for b in range(len(all_sp)):
      if r_ancom[a].loc[all_sp[b], 'detected_0.9'] == True:
        this_sig_09.append(all_sp[b])
      if r_ancom[a].loc[all_sp[b], 'detected_0.8'] == True:
        this_sig_08.append(all_sp[b])
      if r_ancom[a].loc[all_sp[b], 'detected_0.7'] == True:
        this_sig_07.append(all_sp[b])
      if r_ancom[a].loc[all_sp[b], 'detected_0.6'] == True:
        this_sig_06.append(all_sp[b])
    ancom_lists_09.append(this_sig_09), ancom_lists_08.append(this_sig_08), ancom_lists_07.append(this_sig_07), ancom_lists_06.append(this_sig_06)
  return [ancom_lists_09, ancom_lists_08, ancom_lists_07, ancom_lists_06]

```

```{python, tax_kraken_27, results='hide'}
def plot_heatmap(new_genus, ANCOM):
    print(ANCOM)
    names = ['Blood', 'Saliva', 'Buccal', 'Saliva_euk', 'Buccal_euk']
    other_names = ['Abundance', r'Blood $vs$ saliva', r'Blood $vs$ buccal', r'Saliva $vs$ buccal', r'Saliva $vs$ saliva_euk', r'Buccal $vs$ buccal_euk', r'Blood $vs$ saliva_euk', r'Blood $vs$ buccal_euk', r'Saliva_euk $vs$ buccal_euk']
    colormap, norm = mpl.cm.get_cmap('plasma', 256), mpl.colors.Normalize(vmin=0, vmax=1)
    m = mpl.cm.ScalarMappable(norm=norm, cmap=colormap)
    if list(new_genus.index.values)[0][0] == 'A':
        new_genus = new_genus.iloc[::-1]
    figure = plt.figure(figsize=(5,new_genus.shape[0]*0.2))
    ax1 = plt.subplot(111)
    genus_names = list(new_genus.index.values)
    y = []
    for g in range(len(genus_names)):
        this_row = new_genus.loc[genus_names[g], :]
        values = [(np.mean(this_row[name].values)) for name in names]
        bottom, top, x = [g for a in range(5)], [1 for a in range(5)], [a for a in range(5)]
        y.append(g+0.5)
        ma = max(values)
        values = [v/ma for v in values]
        values = [m.to_rgba(v) for v in values]
        ax1.bar(x, top, bottom=bottom, color=values, edgecolor='k', width=1)
        x.append(5)
        ax1.scatter(x[-1], bottom[-1]+0.5, color='k', s=ma*5)
        sig, x_plt = [], x[-1]
        for a in range(len(ANCOM)):
            x_plt += 0.75
            if genus_names[g] in ANCOM[a]: ax1.scatter(x_plt, bottom[-1]+0.5, marker='*', color='k')
            x.append(x_plt)
    plt.ylim([0, len(genus_names)]), plt.xlim([-0.5, x[-1]+0.5])
    plt.yticks(y, genus_names)
    plt.xticks(x, names+other_names, rotation=90)
    ax1.xaxis.set_ticks_position('top')
    plt.tight_layout()
    plt.show()
    return
```

### Minikraken v1 ANCOM

These differences are determined by ANCOM2 and the most conservative of these is the ANCOM 0.9 set, which is then used for plotting heatmaps.

```{python, tax_kraken_28, results='hide', cache=TRUE}
this_genera_db = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera/minikraken_conf.csv', header=0, index_col=0)
comparisons, md, new_genus, comp_len = get_differences(this_genera_db, db_name[1])
c1, c2, c3, c4, c5, c6, c7, c8 = comparisons[0], comparisons[1], comparisons[2], comparisons[3], comparisons[4], comparisons[5], comparisons[6], comparisons[7]
md1, md2, md3, md4, md5, md6, md7, md8 = md[0], md[1], md[2], md[3], md[4], md[5], md[6], md[7]
```

```{r, tax_kraken_29, results='hide', cache=TRUE}
ft = list(py$c1, py$c2, py$c3, py$c4, py$c5, py$c6, py$c7, py$c8)
md = list(py$md1, py$md2, py$md3, py$md4, py$md5, py$md6, py$md7, py$md8)
all_ancom = get_ancom(ft, md)
```

```{python, tax_kraken_30, results='hide', cache=TRUE}
ancom = sort_ancom_results(r.all_ancom)
ancom_df = []
for c in range(len(comparison_names)):
    this_row = [comparison_names[c], comp_len[c], len(ancom[3][c]), len(ancom[2][c]), len(ancom[1][c]), len(ancom[0][c])]
    ancom_df.append(this_row)
ancom_df = pd.DataFrame(ancom_df, columns=['Comparison', 'Shared genera', 'ANCOM 0.6', 'ANCOM 0.7', 'ANCOM 0.8', 'ANCOM 0.9'])
```

```{R, tax_kraken_31, cache=TRUE}
py$ancom_df %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```
### Minikraken v1 heatmap

Relative abundance is scaled within each genus, with yellow being highest in abundance and purple lowest. Blobs are scaled to maximum relative abundance. Stars denote genera that are significantly differentially abundant between body sites (ANCOM2 0.9).

```{python, tax_kraken_32, results='hide', cache=TRUE}
plot_heatmap(new_genus, ancom[0])
```

### Minikraken v2 ANCOM

These differences are determined by ANCOM2 and the most conservative of these is the ANCOM 0.9 set, which is then used for plotting heatmaps.

```{python, tax_kraken_33, results='hide', cache=TRUE}
this_genera_db = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera/minikraken_human_conf.csv', header=0, index_col=0)
comparisons, md, new_genus, comp_len = get_differences(this_genera_db, db_name[1])
c1, c2, c3, c4, c5, c6, c7, c8 = comparisons[0], comparisons[1], comparisons[2], comparisons[3], comparisons[4], comparisons[5], comparisons[6], comparisons[7]
md1, md2, md3, md4, md5, md6, md7, md8 = md[0], md[1], md[2], md[3], md[4], md[5], md[6], md[7]
```

```{r, tax_kraken_34, results='hide', cache=TRUE}
ft = list(py$c1, py$c2, py$c3, py$c4, py$c5, py$c6, py$c7, py$c8)
md = list(py$md1, py$md2, py$md3, py$md4, py$md5, py$md6, py$md7, py$md8)
all_ancom = get_ancom(ft, md)
```

```{python, tax_kraken_35, results='hide', cache=TRUE}
ancom = sort_ancom_results(r.all_ancom)
ancom_df = []
for c in range(len(comparison_names)):
    this_row = [comparison_names[c], comp_len[c], len(ancom[3][c]), len(ancom[2][c]), len(ancom[1][c]), len(ancom[0][c])]
    ancom_df.append(this_row)
ancom_df = pd.DataFrame(ancom_df, columns=['Comparison', 'Shared genera', 'ANCOM 0.6', 'ANCOM 0.7', 'ANCOM 0.8', 'ANCOM 0.9'])
```

```{R, tax_kraken_36, cache=TRUE}
py$ancom_df %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```
### Minikraken v2 heatmap

Relative abundance is scaled within each genus, with yellow being highest in abundance and purple lowest. Blobs are scaled to maximum relative abundance. Stars denote genera that are significantly differentially abundant between body sites (ANCOM2 0.9).

```{python, tax_kraken_37, results='hide', cache=TRUE}
plot_heatmap(new_genus, ancom[0])
```

### GTDB ANCOM

These differences are determined by ANCOM2 and the most conservative of these is the ANCOM 0.9 set, which is then used for plotting heatmaps.

```{python, tax_kraken_38, results='hide', cache=TRUE}
this_genera_db = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera/gtdb_conf.csv', header=0, index_col=0)
comparisons, md, new_genus, comp_len = get_differences(this_genera_db, db_name[1])
c1, c2, c3, c4, c5, c6, c7, c8 = comparisons[0], comparisons[1], comparisons[2], comparisons[3], comparisons[4], comparisons[5], comparisons[6], comparisons[7]
md1, md2, md3, md4, md5, md6, md7, md8 = md[0], md[1], md[2], md[3], md[4], md[5], md[6], md[7]
```

```{r, tax_kraken_39, results='hide', cache=TRUE}
ft = list(py$c1, py$c2, py$c3, py$c4, py$c5, py$c6, py$c7, py$c8)
md = list(py$md1, py$md2, py$md3, py$md4, py$md5, py$md6, py$md7, py$md8)
all_ancom = get_ancom(ft, md)
```

```{python, tax_kraken_40, results='hide', cache=TRUE}
ancom = sort_ancom_results(r.all_ancom)
ancom_df = []
for c in range(len(comparison_names)):
    this_row = [comparison_names[c], comp_len[c], len(ancom[3][c]), len(ancom[2][c]), len(ancom[1][c]), len(ancom[0][c])]
    ancom_df.append(this_row)
ancom_df = pd.DataFrame(ancom_df, columns=['Comparison', 'Shared genera', 'ANCOM 0.6', 'ANCOM 0.7', 'ANCOM 0.8', 'ANCOM 0.9'])
```

```{R, tax_kraken_41, cache=TRUE}
py$ancom_df %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```
### GTDB heatmap

Relative abundance is scaled within each genus, with yellow being highest in abundance and purple lowest. Blobs are scaled to maximum relative abundance. Stars denote genera that are significantly differentially abundant between body sites (ANCOM2 0.9).

```{python, tax_kraken_42, results='hide', cache=TRUE}
plot_heatmap(new_genus, ancom[0])
```

### RefSeq Complete v93 ANCOM

These differences are determined by ANCOM2 and the most conservative of these is the ANCOM 0.9 set, which is then used for plotting heatmaps.

```{python, tax_kraken_43, results='hide', cache=TRUE}
this_genera_db = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera/refseq_conf.csv', header=0, index_col=0)
comparisons, md, new_genus, comp_len = get_differences(this_genera_db, db_name[1])
c1, c2, c3, c4, c5, c6, c7, c8 = comparisons[0], comparisons[1], comparisons[2], comparisons[3], comparisons[4], comparisons[5], comparisons[6], comparisons[7]
md1, md2, md3, md4, md5, md6, md7, md8 = md[0], md[1], md[2], md[3], md[4], md[5], md[6], md[7]
```

```{r, tax_kraken_44, results='hide', cache=TRUE}
ft = list(py$c1, py$c2, py$c3, py$c4, py$c5, py$c6, py$c7, py$c8)
md = list(py$md1, py$md2, py$md3, py$md4, py$md5, py$md6, py$md7, py$md8)
all_ancom = get_ancom(ft, md)
```

```{python, tax_kraken_45, results='hide', cache=TRUE}
ancom = sort_ancom_results(r.all_ancom)
ancom_df = []
for c in range(len(comparison_names)):
    this_row = [comparison_names[c], comp_len[c], len(ancom[3][c]), len(ancom[2][c]), len(ancom[1][c]), len(ancom[0][c])]
    ancom_df.append(this_row)
ancom_df = pd.DataFrame(ancom_df, columns=['Comparison', 'Shared genera', 'ANCOM 0.6', 'ANCOM 0.7', 'ANCOM 0.8', 'ANCOM 0.9'])
```

```{R, tax_kraken_46, cache=TRUE}
py$ancom_df %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```
### RefSeq Complete v93 heatmap

Relative abundance is scaled within each genus, with yellow being highest in abundance and purple lowest. Blobs are scaled to maximum relative abundance. Stars denote genera that are significantly differentially abundant between body sites (ANCOM2 0.9).

```{python, tax_kraken_47, results='hide', cache=TRUE}
plot_heatmap(new_genus, ancom[0])
```

## Kraken2 intersection between databases {.tabset}

Now we are looking at the taxa (grouped to genus level, to hopefully allow for differences between databases) that are in common between the different databases that have been used with Kraken2. We are also using only the versions that have confidence=0.1.

### Genera common to all databases (reads before and after)

The plots below here show the number of reads in each sample before and after removal of the genera that aren't present in all databases. 

```{python, tax_kraken_int_1, results='hide'}
#print(bracken_all_db)
def intersection(lst1, lst2): 
    lst3 = [value for value in lst1 if value in lst2] 
    return lst3

conf_bracken_genus = []
genus_in_each = []
for a in range(len(bracken_all_db)):
  if a not in [0, 2, 4, 6]:
    continue
  this_db = pd.DataFrame(bracken_all_db[a])
  taxa = list(this_db.index.values)
  tax_dict = {}
  for b in range(len(taxa)):
    orig_tax = taxa[b]
    if 's__' in taxa[b]:
      taxa[b] = taxa[b].replace('s__', '')
    taxa[b] = taxa[b].split(' ')[0]
    taxa[b] = taxa[b].replace("'", "")
    tax_dict[orig_tax] = taxa[b]
  this_db.rename(index=tax_dict, inplace=True)
  genus = this_db.groupby(by=this_db.index, axis=0).sum()
  conf_bracken_genus.append(genus)
  genus_in_each.append(list(genus.index.values))

[print(len(gen)) for gen in genus_in_each]
overall_genus = intersection(genus_in_each[0], genus_in_each[1])
overall_genus = intersection(overall_genus, genus_in_each[2])
overall_genus = intersection(overall_genus, genus_in_each[3])
print(len(overall_genus))

fig = plt.figure(figsize=(10,6))
ax = [plt.subplot(223), plt.subplot(221), plt.subplot(222), plt.subplot(224)]
x1 = [x for x in range(21)]
x2 = [x+0.4 for x in range(21)]
x3 = [x+0.2 for x in range(21)]

conf_bracken_overall = []
names = ['gtdb', 'minikrakenv1', 'minikrakenv2', 'refseq']
labels = ['GTDB', 'Minikraken v1 (without human)', 'Minikraken v2 (with human)', 'RefSeq Complete v93']
sum_reduced = []
for db in range(len(conf_bracken_genus)):
  new_db = pd.DataFrame(conf_bracken_genus[db].loc[overall_genus, :])
  conf_bracken_overall.append(new_db)
  new_db.to_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera_in_common/'+names[db]+'_common.csv')
  conf_bracken_genus[db].to_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera_in_common/'+names[db]+'.csv')
  sums = new_db.sum(axis=0)
  sum_reduced.append(sums)
  ax[db].bar(x2, sums, color=colors, edgecolor='k', width=0.4, alpha=0.5)
  sums = conf_bracken_genus[db].sum(axis=0)
  ax[db].bar(x1, sums, color=colors, edgecolor='k', width=0.4)
  ax[db].semilogy()
  ax[db].set_title(labels[db])
  plt.sca(ax[db])
  if db == 1 or db == 2:
    plt.xticks([])
  else:
    plt.xticks(x3, sample_names, rotation=90)
ax[0].set_ylabel('Number of reads'), ax[1].set_ylabel('Number of reads')
handles = [Patch(facecolor='k', edgecolor='k', label='All genera'), Patch(facecolor='k', edgecolor='k', alpha=0.5, label='Genera present in all')]
ax[2].legend(handles=handles, loc='upper left', bbox_to_anchor=(1,1))
plt.tight_layout()
plt.show()
```

<br/><br/>

### Genera common to all databases (reads remaining)

```{python, tax_kraken_int_2, results='hide', cache=TRUE}
fig = plt.figure(figsize=(8,4))
ax = plt.subplot(111)
x1 = [x*5 for x in range(21)]
x2 = [x+1 for x in x1]
x3 = [x+1 for x in x2]
x4 = [x+1 for x in x3]
xplt = [x+0.5 for x in x2]

x = [x3, x1, x2, x4]
al = [0.6, 1, 0.8, 0.4]
for db in range(len(sum_reduced)):
  ax.bar(x[db], sum_reduced[db], color=colors, width=1, edgecolor='k', alpha=al[db])
handles = [Patch(facecolor='k', edgecolor='k', label='Minikraken v1'), Patch(facecolor='k', edgecolor='k', alpha=0.8, label='Minikraken v2'), Patch(facecolor='k', edgecolor='k', alpha=0.6, label='GTDB'), Patch(facecolor='k', edgecolor='k', alpha=0.4, label='RefSeq Complete v93')]
ax.legend(handles=handles, loc='upper left', bbox_to_anchor=(1,1))
plt.xticks(xplt, sample_names, rotation=90)
plt.semilogy()
plt.ylabel('Number of reads')
plt.tight_layout()
plt.show()
```

<br/>

### Genera stacked bars (absolute abundance)

```{python, tax_kraken_int_3, results='hide', cache=TRUE}
for db in range(len(conf_bracken_overall)):
  rename_samples = {}
  for sample in list(conf_bracken_overall[db].columns):
    rename_samples[sample] = sample+'_'+names[db]
  this_db = pd.DataFrame(conf_bracken_overall[db].rename(columns=rename_samples))
  if db == 0:
    overall_genus = this_db
  else:
    overall_genus = pd.concat([overall_genus, this_db], axis=1)
overall_genus = pd.DataFrame(overall_genus)
#overall_genus = pd.DataFrame(overall_genus.divide(overall_genus.sum(axis=0)).multiply(100))
overall_genus = overall_genus[overall_genus.max(axis=1) > 25000]
sums = overall_genus.sum(axis=0)
gen_colors = get_cols(overall_genus.shape[0])

fig = plt.figure(figsize=(16,7))
ax1 = plt.subplot(234)
ax = [ax1, plt.subplot(231, sharey=ax1), plt.subplot(232, sharey=ax1), plt.subplot(235, sharey=ax1)]
x1 = [x for x in range(21)]
a = 0
col_samples = list(overall_genus.columns)
#line 20
new_db = []
for name in names:
  keeping = []
  for s in list(overall_genus.columns):
    if name in s:
      keeping.append(True)
    else:
      keeping.append(False)
  other_new_db = pd.DataFrame(overall_genus.loc[:, keeping])
  new_db.append(other_new_db)
for db in range(len(new_db)):
  this_genera = list(new_db[db].index.values)
  handles = []
  for g in range(len(this_genera)):
    if g == 0:
      bottom = [0 for c in x1]
    these_values = new_db[db].loc[this_genera[g], :].values
    ax[db].bar(x1, these_values, bottom=bottom, color=gen_colors[g], edgecolor='k', width=1)
    handles.append(Patch(facecolor=gen_colors[g], edgecolor='k', label=this_genera[g]))
    bottom = [bottom[x]+these_values[x] for x in range(len(bottom))]
  plt.sca(ax[db])
  if db == 1 or db == 2:
    plt.xticks([])
  else:
    plt.xticks(x1, sample_names, rotation=90)
  plt.semilogy()
  plt.xlim([x1[0]-0.5, x1[-1]+0.5])
  plt.title(labels[db])
ax[0].set_ylabel('Absolute abundance (reads)'), ax[1].set_ylabel('Absolute abundance (reads)'), 
ax[2].legend(handles=handles, loc='upper left', bbox_to_anchor=(1,1), ncol=3)
#plt.tight_layout()
plt.show()
```
<br/>

### Genera stacked bars (relative abundance)

```{python, tax_kraken_int_4, results='hide', cache=TRUE}
for db in range(len(conf_bracken_overall)):
  rename_samples = {}
  for sample in list(conf_bracken_overall[db].columns):
    rename_samples[sample] = sample+'_'+names[db]
  this_db = pd.DataFrame(conf_bracken_overall[db].rename(columns=rename_samples))
  if db == 0:
    overall_genus = this_db
  else:
    overall_genus = pd.concat([overall_genus, this_db], axis=1)
overall_genus = pd.DataFrame(overall_genus)
overall_genus = pd.DataFrame(overall_genus.divide(overall_genus.sum(axis=0)).multiply(100))
overall_genus = overall_genus[overall_genus.max(axis=1) > 1]
sums = overall_genus.sum(axis=0)
gen_colors = get_cols(overall_genus.shape[0])

fig = plt.figure(figsize=(16,7))
ax1 = plt.subplot(234)
ax = [ax1, plt.subplot(231, sharey=ax1), plt.subplot(232, sharey=ax1), plt.subplot(235, sharey=ax1)]
x1 = [x for x in range(21)]
a = 0
col_samples = list(overall_genus.columns)
#line 20
new_db = []
for name in names:
  keeping = []
  for s in list(overall_genus.columns):
    if name in s:
      keeping.append(True)
    else:
      keeping.append(False)
  other_new_db = pd.DataFrame(overall_genus.loc[:, keeping])
  new_db.append(other_new_db)
for db in range(len(new_db)):
  this_genera = list(new_db[db].index.values)
  handles = []
  for g in range(len(this_genera)):
    if g == 0:
      bottom = [0 for c in x1]
    these_values = new_db[db].loc[this_genera[g], :].values
    ax[db].bar(x1, these_values, bottom=bottom, color=gen_colors[g], edgecolor='k', width=1)
    handles.append(Patch(facecolor=gen_colors[g], edgecolor='k', label=this_genera[g]))
    bottom = [bottom[x]+these_values[x] for x in range(len(bottom))]
  plt.sca(ax[db])
  if db == 1 or db == 2:
    plt.xticks([])
  else:
    plt.xticks(x1, sample_names, rotation=90)
  plt.xlim([x1[0]-0.5, x1[-1]+0.5]), plt.ylim([0, 100])
  plt.title(labels[db])
ax[0].set_ylabel('Relative abundance (%)'), ax[1].set_ylabel('Relative abundance (%)')
ax[2].legend(handles=handles, loc='upper left', bbox_to_anchor=(1,1), ncol=2)
plt.show()
```

<br/>

### Presence/absence of genera

Here, each sample is plotted using each of the four different databases, from top to bottom: GTDB, Minikraken v1, Minikraken v2 and RefSeq Complete v93. Each genera is plotted as a column, with black indicating that it is present and white indicating that it is absent. 

Note that for some sample sets, e.g. saliva, there is far more consensus within samples than in other sample sets, e.g. blood. 

```{python, tax_kraken_int_5, results='hide', cache=TRUE}
count = 0
plt.figure(figsize=(6,6))
x, y, b0, b1, b2, b3 = [], [], [], [], [], []
for s in full_name_dict:
  print(s)
  ax = plt.subplot2grid((21, 10), (count,2), colspan=8)
  count += 1
  nums, pres_colors = [], []
  for db in conf_bracken_overall:
    nums.append(list(db.loc[:, s].values))
    pres_colors.append([])
  for a in range(len(nums[0])):
    if nums[0][a] > 0: pres_colors[0].append('k')
    else: pres_colors[0].append('w')
    if nums[1][a] > 0: pres_colors[1].append('k')
    else: pres_colors[1].append('w')
    if nums[2][a] > 0: pres_colors[2].append('k')
    else: pres_colors[2].append('w')
    if nums[3][a] > 0: pres_colors[3].append('k')
    else: pres_colors[3].append('w')
    if s == 'SRR8595488':
      x.append(a), y.append(1), b0.append(0), b1.append(1), b2.append(2), b3.append(3)
  ax.bar(x, y, bottom=b3, color=pres_colors[0], width=1)
  ax.bar(x, y, bottom=b2, color=pres_colors[1], width=1)
  ax.bar(x, y, bottom=b1, color=pres_colors[2], width=1)
  ax.bar(x, y, bottom=b0, color=pres_colors[3], width=1)
  plt.sca(ax)
  plt.xticks([]), plt.yticks([2], [full_name_dict[s]])
  plt.xlim([x[0]-0.5, x[-1]+0.5]), plt.ylim([0, 4])
plt.subplots_adjust(hspace=0.05)
plt.show()
```

<br/>

### Presence/absence of genera (in all 4 databases)

Now I am combining the data from the 4 databases, and only plotting a genus in a sample if it is present in all 4 databases. Again, black indicates presence while white indicates absence.

```{python, tax_kraken_int_6, results='hide', cache=TRUE}
count = 0
plt.figure(figsize=(6,6))
x, y = [], []
for s in full_name_dict:
  ax = plt.subplot2grid((21, 10), (count,2), colspan=8)
  count += 1
  nums, pres_colors = [], []
  for db in conf_bracken_overall:
    nums.append(list(db.loc[:, s].values))
  for a in range(len(nums[0])):
    if nums[0][a] > 0 and nums[1][a] > 0 and nums[2][a] > 0 and nums[3][a] > 0: pres_colors.append('k')
    else: pres_colors.append('w')
    if s == 'SRR8595488':
      x.append(a), y.append(1)
  ax.bar(x, y, color=pres_colors, width=1)
  plt.sca(ax)
  plt.xticks([]), plt.yticks([0.5], [full_name_dict[s]])
  plt.xlim([x[0]-0.5, x[-1]+0.5]), plt.ylim([0, 1])
plt.subplots_adjust(hspace=0.05)
plt.show()
```

<br/>

### Presence/absence of genera (in 3 of 4 databases)

Now I am combining the data from the 4 databases, and only plotting a genus in a sample if it is present in 3 of 4 databases. Again, black indicates presence while white indicates absence.

```{python, tax_kraken_int_7, results='hide', cache=TRUE}
count = 0
plt.figure(figsize=(6,6))
x, y = [], []
for s in full_name_dict:
  ax = plt.subplot2grid((21, 10), (count,2), colspan=8)
  count += 1
  nums, pres_colors = [], []
  for db in conf_bracken_overall:
    nums.append(list(db.loc[:, s].values))
  for a in range(len(nums[0])):
    this_num = [nums[0][a], nums[1][a], nums[2][a], nums[3][a]]
    if sum(this_num) >= 3: pres_colors.append('k')
    else: pres_colors.append('w')
    if s == 'SRR8595488':
      x.append(a), y.append(1)
  ax.bar(x, y, color=pres_colors, width=1)
  plt.sca(ax)
  plt.xticks([]), plt.yticks([0.5], [full_name_dict[s]])
  plt.xlim([x[0]-0.5, x[-1]+0.5]), plt.ylim([0, 1])
plt.subplots_adjust(hspace=0.05)
plt.show()
```

<br/>

### Presence/absence of genera (in 2 of 4 databases)

Now I am combining the data from the 4 databases, and only plotting a genus in a sample if it is present in 2 of 4 databases. Again, black indicates presence while white indicates absence.

```{python, tax_kraken_int_8, results='hide', cache=TRUE}
count = 0
plt.figure(figsize=(6,6))
x, y = [], []
for s in full_name_dict:
  ax = plt.subplot2grid((21, 10), (count,2), colspan=8)
  count += 1
  nums, pres_colors = [], []
  for db in conf_bracken_overall:
    nums.append(list(db.loc[:, s].values))
  for a in range(len(nums[0])):
    this_num = [nums[0][a], nums[1][a], nums[2][a], nums[3][a]]
    if sum(this_num) >= 2: pres_colors.append('k')
    else: pres_colors.append('w')
    if s == 'SRR8595488':
      x.append(a), y.append(1)
  ax.bar(x, y, color=pres_colors, width=1)
  plt.sca(ax)
  plt.xticks([]), plt.yticks([0.5], [full_name_dict[s]])
  plt.xlim([x[0]-0.5, x[-1]+0.5]), plt.ylim([0, 1])
plt.subplots_adjust(hspace=0.05)
plt.show()
```

<br/>

### Summary

These plots show that almost all genera are present in saliva samples, while the other body sites just have subsets of these. There is also the largest amount of consensus between saliva samples, and the use of different databases to determine presence/absence. I will now include the genera that are present in 3 of 4 databases for further analyses.

</br>

## Differences between body sites based on presence/absence {.tabset}

These plots are looking at the differences between body sites based on presence/absence (using the taxa that are present in at least 3 of 4 databases). I have also removed the HuRef blood before carrying out any comparisons.

```{python, presabs_sig_1, results='hide', fig.keep='all'}
new_db = []
for s in full_name_dict:
  if s == 'SRR8595488': continue
  nums, presence = [], [s]
  for db in conf_bracken_overall:
    nums.append(list(db.loc[:, s].values))
  for a in range(len(nums[0])):
    this_num = [nums[0][a], nums[1][a], nums[2][a], nums[3][a]]
    if sum(this_num) >= 3: presence.append(1)
    else: presence.append(0)
  new_db.append(presence)
new_db = pd.DataFrame(new_db, columns=['Samples']+list(conf_bracken_overall[0].index.values))
new_db.set_index('Samples', inplace=True)
new_db = new_db.transpose()
new_db.to_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera_in_common/combined_db_3of4.csv')

list_comparisons = [['Blood', 'Saliva'], ['Blood', 'Buccal'], ['Saliva', 'Buccal'], ['Saliva', 'Saliva_euk'], ['Buccal', 'Buccal_euk'], ['Blood', 'Saliva_euk'], ['Blood', 'Buccal_euk'], ['Saliva_euk', 'Buccal_euk']]
samples = list(new_db.columns)
comparisons, metadata, comp_len = [], [], []
for a in range(len(list_comparisons)):
    keeping = []
    this_md = []
    for b in range(len(samples)):
        if site_dict[samples[b]] in list_comparisons[a]:
            keeping.append(True)
            this_md.append([samples[b], site_dict[samples[b]]])
        else:
            keeping.append(False)
    this_comp = new_db.loc[:, keeping]
    comparisons.append(this_comp)
    this_md = pd.DataFrame(this_md, columns=['Samples', 'Groups'])
    #this_md.set_index('Samples', inplace=True)
    metadata.append(this_md)
    comp_len.append(this_comp.shape[0])
    
c1, c2, c3, c4, c5, c6, c7, c8 = comparisons[0], comparisons[1], comparisons[2], comparisons[3], comparisons[4], comparisons[5], comparisons[6], comparisons[7]
md1, md2, md3, md4, md5, md6, md7, md8 = metadata[0], metadata[1], metadata[2], metadata[3], metadata[4], metadata[5], metadata[6], metadata[7]
```

```{r, presabs_sig_2, results='hide', fig.keep='all', cache=TRUE}
ft = list(py$c1, py$c2, py$c3, py$c4, py$c5, py$c6, py$c7, py$c8)
md = list(py$md1, py$md2, py$md3, py$md4, py$md5, py$md6, py$md7, py$md8)
all_ancom = get_ancom(ft, md)
```

```{python, presabs_sig_3, results='hide', fig.keep='all', cache=TRUE}
ancom = sort_ancom_results(r.all_ancom)
ancom_df = []
comparison_names = [r'Blood vs saliva', r'Blood vs buccal', r'Saliva vs buccal', r'Saliva vs saliva_euk', r'Buccal vs buccal_euk', r'Blood vs saliva_euk', r'Blood vs buccal_euk', r'Saliva_euk vs buccal_euk']
for c in range(len(comparison_names)):
    this_row = [comparison_names[c], comp_len[c], len(ancom[3][c]), len(ancom[2][c]), len(ancom[1][c]), len(ancom[0][c])]
    ancom_df.append(this_row)
ancom_df = pd.DataFrame(ancom_df, columns=['Comparison', 'Shared genera', 'ANCOM 0.6', 'ANCOM 0.7', 'ANCOM 0.8', 'ANCOM 0.9'])
```
<br/>

### Tables of presence/absence

This is the table of presence/absence used for the ANCOM tests.

```{python, presabs_sig_table, results='hide', fig.keep='all', cache=TRUE}
pres_abs = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kraken2/db_genera_in_common/combined_db_3of4.csv', header=0, index_col=0)
pres_abs.rename(columns=full_name_dict, inplace=True)
```

```{R, presabs_sig_4, cache=TRUE}
py$pres_abs %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### ANCOM summary

This table shows the number of differentially abundant taxa between body sites as determined by ANCOM tests with different significance thresholds.

```{R, presabs_sig_4b, cache=TRUE}
py$ancom_df %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```
<br/>

### ANCOM heatmap

This is plotted using the ANCOM 0.7 results. I have removed all taxa that are present in all body sites. Yellow means present in all samples for that body site while purple means absent in all samples for that body site. 

Note that the differences that we saw before with Klebsiella, Streptococcus and Prevotella are not present here as all of these are present in all samples. 

```{python, presabs_sig_5, results='hide', fig.keep='all', cache=TRUE}
new_new_db = new_db.rename(columns=site_dict)
keeping = []
taxa_new = list(new_new_db.index.values)
for tax in taxa_new:
  this_row = new_new_db.loc[tax, :].values
  if len(list(set(this_row))) == 1: keeping.append(False)
  else: keeping.append(True)
new_new_db = new_new_db.loc[keeping, :]
plot_heatmap(new_new_db, ancom[2])
```


<br/>

# Functional profiling

Function is profiled using:<br/>
1. [HUMAnN2](https://huttenhower.sph.harvard.edu/humann)<br/>
    - Uniref90 database<br/>
    - Uniref50 database<br/>
2. [HUMAnN3](https://github.com/biobakery/biobakery/wiki/humann3)<br/>
    - Uniref90 and Uniref50<br/>

## HUMAnN unaligned reads

We can see here that it's a bit variable as to whether HUMAnN3 classifies more or less reads than either of the databases with HUMAnN2 (HUMAnN3 uses both Uniref50 and Uniref90). Also, with HUMAnN3, all of the samples with no taxonomic classifications with MetaPhlAn3 have 48-49% unaligned, which is much less than any of the other samples.

```{python, func_humann2_1, results='hide', fig.keep='all', cache=TRUE}
plt.figure(figsize=(8,6))
ax1 = plt.subplot(111)

x1 = [x for x in range(21)]
x2 = [x+0.3 for x in range(21)]
x3 = [x+0.6 for x in range(21)]
#x = [x+0.2 for x in range(21)]

ax1.bar(x1, reads.loc[:, 'unaligned_after_translation_uniref_90'].values, color=colors, edgecolor='k', width=0.3)
ax1.bar(x2, reads.loc[:, 'unaligned_after_translation_uniref_50'].values, color=colors, edgecolor='k', width=0.3, alpha=0.7)
ax1.bar(x3, reads.loc[:, 'unaligned_after_translation_humann3'].values, color=colors, edgecolor='k', width=0.3, alpha=0.4)
plt.xticks(x2, sample_names, rotation=90)
plt.ylabel('Unaligned reads after translation (%)')
plt.xlim([-0.5,21])
handles = [Patch(facecolor='k', edgecolor='k', label='HUMAnN2 Uniref90'), Patch(facecolor='k', edgecolor='k', alpha=0.7, label='HUMAnN2 Uniref50'), Patch(facecolor='k', edgecolor='k', alpha=0.4, label='HUMAnN3')]
ax1.legend(handles=handles, loc='upper left', bbox_to_anchor=(1,1.02))

plt.tight_layout()
plt.show()
```
<br/>

## HUMAnN richness {.tabset}

After running HUMAnN2 and HUMAnN3, the pathabundance, pathcoverage and genefamilies tables are joined, and the pathabundance is renormalised (relative abundance calculated) and split to tables that are stratified by species or not. Here we use the unstratified pathabundance file (the MetaPhlAn2/MetaPhlAn3 species results don't seem to have been particularly accurate). The genefamilies file is stratified by default so we take only the overall values for the gene family (removing all species results) and re-calculate these values as relative abundances. 

```{python, func_humann2_2, results='hide', fig.keep='all'}

pathways_90 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_90/humann2_pathabundance_relab_unstratified.tsv', header=0, index_col=0, sep='\t')
pathways_50 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_50/humann2_pathabundance_relab_unstratified.tsv', header=0, index_col=0, sep='\t')
pathways_h3 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/humann3_final_out/humann3_pathabundance_relab_unstratified.tsv', header=0, index_col=0, sep='\t')

def get_richness(df):
  snames = list(df.columns)
  num_genes = []
  for sn in snames:
    one_sample = pd.DataFrame(df.loc[:, sn])
    one_sample = one_sample[one_sample.max(axis=1) > 0]
    num_genes.append(one_sample.shape[0])
  return snames, num_genes
  
def genes_filter(genes):
  genes.drop('UNMAPPED', axis=0, inplace=True)
  gene_names = list(genes.index.values)
  keeping = []
  for gn in gene_names:
      new_gn = gn.split('|')[0]
      if gn == new_gn: keeping.append(True)
      else: keeping.append(False)
  genes_abs = genes.loc[keeping, :]
  genes = genes_abs.divide(genes_abs.sum(axis=0)).multiply(100)
  return genes_abs, genes

genes_90_not_cpm = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_90/humann2_genefamilies.tsv', header=0, index_col=0, sep='\t')
genes_90 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_90/humann2_genefamilies_cpm.tsv', header=0, index_col=0, sep='\t')
genes_90_ko = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_90/humann2_genefamilies_cpm_ko_90.tsv', header=0, index_col=0, sep='\t')
genes_90_not_cpm = genes_filter(genes_90_not_cpm)[1]
genes_90 = genes_filter(genes_90)[1]
genes_90_ko = genes_filter(genes_90_ko)[1]

snames_genes_90, genes_90_richness = get_richness(genes_90)
snames_genes_90_ko, genes_90_richness_ko = get_richness(genes_90_ko)
snames_pways_90, pways_90_richness = get_richness(pathways_90)


genes_50_not_cpm = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_50/humann2_genefamilies.tsv', header=0, index_col=0, sep='\t')
genes_50 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_50/humann2_genefamilies_cpm.tsv', header=0, index_col=0, sep='\t')
genes_50_ko = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_50/humann2_genefamilies_cpm_ko_50.tsv', header=0, index_col=0, sep='\t')
genes_50_not_cpm = genes_filter(genes_50_not_cpm)[1]
genes_50 = genes_filter(genes_50)[1]
genes_50_ko = genes_filter(genes_50_ko)[1]

snames_genes_50, genes_50_richness = get_richness(genes_50)
snames_genes_50_ko, genes_50_richness_ko = get_richness(genes_50_ko)
snames_pways_50, pways_50_richness = get_richness(pathways_50)

genes_h3_not_cpm = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/humann3_final_out/humann3_genefamilies.tsv', header=0, index_col=0, sep='\t')
genes_h3 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/humann3_final_out/humann3_genefamilies_cpm.tsv', header=0, index_col=0, sep='\t')
genes_h3_ko_50 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/humann3_final_out/humann3_genefamilies_cpm_ko_50.tsv', header=0, index_col=0, sep='\t')
genes_h3_ko_90 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/humann3_final_out/humann3_genefamilies_cpm_ko_90.tsv', header=0, index_col=0, sep='\t')
genes_h3_not_cpm = genes_filter(genes_h3_not_cpm)[1]
genes_h3 = genes_filter(genes_h3)[1]
genes_h3_ko_50 = genes_filter(genes_h3_ko_50)[1]
genes_h3_ko_90 = genes_filter(genes_h3_ko_90)[1]

snames_genes_h3, genes_h3_richness = get_richness(genes_h3)
snames_genes_h3_ko_50, genes_h3_richness_ko_50 = get_richness(genes_h3_ko_50)
snames_genes_h3_ko_90, genes_h3_richness_ko_90 = get_richness(genes_h3_ko_90)
snames_pways_h3, pways_h3_richness = get_richness(pathways_h3)

```

### Ungrouped KEGG orthologs

Here I've used the built-in HUMAnN databases to map the Uniref gene families to KEGG orthologs, and this is a comparison of the proportion of reads that are ungrouped by using the mapping tables on the HUMAnN2 Uniref90 and Uniref50 databases and then the HUMAnN2 mapping tables separately on the same output for the Uniref90 and UNiref50 (even though these are combined in the HUMAnN2 output).

```{python, func_humann2_2b, results='hide', fig.keep='all', cache=TRUE}
plt.figure(figsize=(8,6))
ax1 = plt.subplot(211)
ax2 = plt.subplot(212)

x = [a for a in range(len(genes_50_ko.columns))]
x1 = [a+0.4 for a in x]

ax1.bar(x, genes_90_ko.loc['UNGROUPED', :].values, width=0.4, color=colors, edgecolor='k')
ax1.bar(x1, genes_50_ko.loc['UNGROUPED', :].values, width=0.4, color=colors, edgecolor='k', alpha=0.6)

ax2.bar(x, genes_h3_ko_90.loc['UNGROUPED', :].values, width=0.4, color=colors, edgecolor='k')
ax2.bar(x1, genes_h3_ko_50.loc['UNGROUPED', :].values, width=0.4, color=colors, edgecolor='k', alpha=0.6)

plt.sca(ax1)
plt.xlim([-0.5,21]), plt.xticks([])
handles = [Patch(facecolor='k', edgecolor='k', label='Uniref90'), Patch(facecolor='k', edgecolor='k', alpha=0.6, label='Uniref50')]
ax1.legend(handles=handles, loc='upper left', bbox_to_anchor=(1,1.02))
plt.ylabel('Ungrouped\nKEGG orthologs (%)')
plt.ylim([0,100])

plt.sca(ax2)
plt.xlim([-0.5,21])
plt.ylim([0,100])
plt.ylabel('Ungrouped\nKEGG orthologs (%)')
plt.xticks(x1, sample_names, rotation=90)
plt.tight_layout()
plt.show()
```
Given how small the proportion of the Uniref gene families that have mapped to KEGG orthologs is, we will continue with the Uniref gene families, and we use the tables that have been normalised within HUMAnN to CPM.


```{python, older_code}
# genes_50 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_50/humann2_genefamilies.tsv', header=0, index_col=0, sep='\t')
# genes_50.drop('UNMAPPED', axis=0, inplace=True)
# gene_names = list(genes_50.index.values)
# keeping = []
# for gn in gene_names:
#     new_gn = gn.split('|')[0]
#     if gn == new_gn: keeping.append(True)
#     else: keeping.append(False)
# genes_50_abs = genes_50.loc[keeping, :]
# genes_50 = genes_50_abs.divide(genes_50_abs.sum(axis=0)).multiply(100)

# 
# snames_genes_50, genes_50_richness = get_richness(genes_50)
# snames_pways_50, pways_50_richness = get_richness(pathways_50)
# 
# genes_h3 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/humann3_final_out/humann3_genefamilies_cpm.tsv', header=0, index_col=0, sep='\t')
# genes_h3.drop('UNMAPPED', axis=0, inplace=True)
# gene_names = list(genes_h3.index.values)
# keeping = []
# for gn in gene_names:
#     new_gn = gn.split('|')[0]
#     if gn == new_gn: keeping.append(True)
#     else: keeping.append(False)
# genes_abs = genes_h3.loc[keeping, :]
# genes_h3 = genes_abs.divide(genes_abs.sum(axis=0)).multiply(100)
# 
# snames_genes, genes_richness = get_richness(genes_h3)
# snames_pways, pways_richness = get_richness(pathways)

# genes_90 = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/humann2_final_out_90/humann2_genefamilies.tsv', header=0, index_col=0, sep='\t')
# genes_90.drop('UNMAPPED', axis=0, inplace=True)
# gene_names = list(genes_90.index.values)
# keeping = []
# for gn in gene_names:
#     new_gn = gn.split('|')[0]
#     if gn == new_gn: keeping.append(True)
#     else: keeping.append(False)
# genes_90_abs = genes_90.loc[keeping, :]
# genes_90 = genes_90_abs.divide(genes_90_abs.sum(axis=0)).multiply(100)
```

### Gene family richness

```{python, func_humann2_3, results='hide', fig.keep='all', cache=TRUE}

plt.figure(figsize=(8,6))
ax1 = plt.subplot(211)
ax2 = plt.subplot(212)

x = [a for a in range(len(genes_90.columns))]
x1 = [a+0.3 for a in x]
x2 = [a+0.6 for a in x]

ax1.bar(x, genes_90_richness, width=0.3, color=colors, edgecolor='k')
ax2.bar(x, genes_90_richness, width=0.3, color=colors, edgecolor='k')

ax1.bar(x1[:len(genes_50_richness)], genes_50_richness, width=0.3, color=colors, edgecolor='k', alpha=0.7)
ax2.bar(x1[:len(genes_50_richness)], genes_50_richness, width=0.3, color=colors, edgecolor='k', alpha=0.7)

ax1.bar(x2, genes_h3_richness, width=0.3, color=colors, alpha=0.4, edgecolor='k')
ax2.bar(x2, genes_h3_richness, width=0.3, color=colors, alpha=0.4, edgecolor='k')

plt.sca(ax1)
plt.xlim([-0.5,21]), plt.xticks([])
handles = [Patch(facecolor='k', edgecolor='k', label='HUMAnN2 Uniref90'), Patch(facecolor='k', edgecolor='k', alpha=0.7, label='HUMAnN2 Uniref50'), Patch(facecolor='k', edgecolor='k', alpha=0.4, label='HUMAnN3')]
ax1.legend(handles=handles, loc='upper left', bbox_to_anchor=(1,1.02))
plt.ylabel('Gene family\nrichness')

plt.sca(ax2)
plt.semilogy()
plt.xlim([-0.5,21])
plt.ylabel('Gene family\nrichness (log scale)')
plt.xticks(x1, sample_names, rotation=90)
plt.tight_layout()
plt.show()
```

We can see here that the number of gene families is lower for HUMAnN3 than for either of the HUMAnN2 runs, aside from for the blood samples where we have higher richness for HUMAnN3 than Uniref90, but lower than Uniref50.  

### Pathway richness

```{python, func_humann2_4, results='hide', fig.keep='all', cache=TRUE}

plt.figure(figsize=(8,6))
ax1 = plt.subplot(211)
ax2 = plt.subplot(212)

x = [a for a in range(len(genes_90.columns))]
x1 = [a+0.3 for a in x]
x2 = [a+0.6 for a in x]

ax1.bar(x, pways_90_richness, width=0.3, color=colors, edgecolor='k')
ax1.bar(x1, pways_50_richness, width=0.3, color=colors, edgecolor='k', alpha=0.7)

ax2.bar(x, pways_90_richness, width=0.3, color=colors, edgecolor='k')
ax2.bar(x1, pways_50_richness, width=0.3, color=colors, edgecolor='k', alpha=0.7)

ax1.bar(x2, pways_h3_richness, width=0.3, color=colors, alpha=0.4, edgecolor='k')
ax2.bar(x2, pways_h3_richness, width=0.3, color=colors, alpha=0.4, edgecolor='k')

plt.sca(ax1)
plt.xticks([])
plt.ylabel('Pathway richness')
handles = [Patch(facecolor='k', edgecolor='k', label='HUMAnN2 Uniref90'), Patch(facecolor='k', edgecolor='k', alpha=0.7, label='HUMAnN2 Uniref50'), Patch(facecolor='k', edgecolor='k', alpha=0.4, label='HUMAnN3')]
ax1.legend(handles=handles, bbox_to_anchor=(1,1))
plt.xlim([-0.5,21])

plt.sca(ax2)
plt.semilogy()
plt.xlim([-0.5,21]), plt.xticks(x1, sample_names, rotation=90)
plt.ylabel('Pathway richness (log scale)')
plt.tight_layout()
plt.show()

```
Here we can see that HUMAnN3 gives us no pathways for the blood samples, and - as with the gene families - HUMAnN3 has lower numbers of pathways than either of the HUMAnN2 runs for all other samples.

## HUMAnN nMDS plots {.tabset}

### Gene families

```{python, humann_nmds_2, results='hide', fig.keep='all', cache=TRUE}
plt.figure(figsize=(8,6))
ax1, ax2, ax3 = plt.subplot(221), plt.subplot(222), plt.subplot(223)
genes_50_t = genes_50.transpose()
pos_50, npos_50, nmds_stress_50 = transform_for_NMDS(genes_50_t)
genes_90_t = genes_90.transpose()
pos_90, npos_90, nmds_stress_90 = transform_for_NMDS(genes_90_t)
genes_h3_t = genes_h3.transpose()
pos_h3, npos_h3, nmds_stress_h3 = transform_for_NMDS(genes_h3_t)
for a in range(len(npos_50)):
   ax1.scatter(npos_50[a,0], npos_50[a,1], marker=shapes[a], color=colors[a], s=100)
   ax2.scatter(npos_90[a,0], npos_90[a,1], marker=shapes[a], color=colors[a], s=100)
   ax3.scatter(npos_h3[a,0], npos_h3[a,1], marker=shapes[a], color=colors[a], s=100)
ax1.set_xlabel('nMDS1'), ax2.set_xlabel('nMDS1'), ax3.set_xlabel('nMDS1')
ax1.set_ylabel('nMDS2'), ax2.set_ylabel('nMDS2'), ax3.set_ylabel('nMDS2')
ax1.set_title('HUMAnN2 Uniref50'), ax2.set_title('HUMAnN2 Uniref90'), ax3.set_title('HUMAnN3')
handles1 = [Patch(facecolor=colors_dict[color], edgecolor='k', label=color) for color in colors_dict]
handles2 = [Line2D([0], [0], marker=shapes_dict[shape], color='w', label=shape, markerfacecolor='k', markersize=15) for shape in shapes_dict]
plt.tight_layout()
ax3.legend(handles=handles1+handles2, loc='upper left', bbox_to_anchor=(1,1.02))
plt.show()
```

### KEGG orthologs

```{python, humann_nmds_1, results='hide', fig.keep='all', cache=TRUE}
plt.figure(figsize=(8,6))
ax1, ax2, ax3, ax4 = plt.subplot(221), plt.subplot(222), plt.subplot(223), plt.subplot(224)
genes_50_ko_t = genes_50_ko.transpose()
pos_50_ko, npos_50_ko, nmds_stress_50_ko = transform_for_NMDS(genes_50_ko_t)
genes_90_ko_t = genes_90_ko.transpose()
pos_90_ko, npos_90_ko, nmds_stress_90_ko = transform_for_NMDS(genes_90_ko_t)
genes_h3_ko_50_t = genes_h3_ko_50.transpose()
pos_h3_ko_50, npos_h3_ko_50, nmds_stress_h3_ko_50 = transform_for_NMDS(genes_h3_ko_50_t)
genes_h3_ko_90_t = genes_h3_ko_90.transpose()
pos_h3_ko_90, npos_h3_ko_90, nmds_stress_h3_ko_90 = transform_for_NMDS(genes_h3_ko_90_t)
for a in range(len(npos_50)):
   ax1.scatter(npos_50_ko[a,0], npos_50_ko[a,1], marker=shapes[a], color=colors[a], s=100)
   ax2.scatter(npos_90_ko[a,0], npos_90_ko[a,1], marker=shapes[a], color=colors[a], s=100)
   ax3.scatter(npos_h3_ko_50[a,0], npos_h3_ko_50[a,1], marker=shapes[a], color=colors[a], s=100)
   ax4.scatter(npos_h3_ko_90[a,0], npos_h3_ko_90[a,1], marker=shapes[a], color=colors[a], s=100)
ax1.set_xlabel('nMDS1'), ax2.set_xlabel('nMDS1'), ax3.set_xlabel('nMDS1'), ax4.set_xlabel('nMDS1')
ax1.set_ylabel('nMDS2'), ax2.set_ylabel('nMDS2'), ax3.set_ylabel('nMDS2'), ax4.set_ylabel('nMDS2')
ax1.set_title('HUMAnN2 Uniref50'), ax2.set_title('HUMAnN2 Uniref90'), ax3.set_title('HUMAnN3 Uniref50'), ax4.set_title('HUMAnN3 Uniref90'), 
handles1 = [Patch(facecolor=colors_dict[color], edgecolor='k', label=color) for color in colors_dict]
handles2 = [Line2D([0], [0], marker=shapes_dict[shape], color='w', label=shape, markerfacecolor='k', markersize=15) for shape in shapes_dict]
ax4.legend(handles=handles1+handles2, loc='upper left', bbox_to_anchor=(1,1.02))
plt.tight_layout()
plt.show()
```

## HUMAnN2 differential abundance between body sites {.tabset}

Now we are looking at the gene families, pathways and KEGG orthologs that are differentially abundant between body sites, using the lists that are output by HUMAnN2.

```{python, func_humann2_5, results='hide', fig.keep='all'}
def get_diff_abundant(genes, name_csv, fp= '/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/'):
    col_names = list(genes.columns)
    col_names_dict = {}
    for cn in col_names:
        if 'SRR8595488' in cn:
          genes.drop([cn], axis=1, inplace=True)
        cn_new = cn.split('_')[0]
        col_names_dict[cn] = site_dict[cn_new]
    new_genes = genes.rename(columns=col_names_dict)
    comparisons = [['Saliva', 'Blood'], ['Buccal', 'Blood'], ['Saliva', 'Buccal'], ['Saliva', 'Saliva_euk'], ['Buccal', 'Buccal_euk'], ['Blood', 'Saliva_euk'], ['Blood', 'Buccal_euk']]
    fig = plt.figure(figsize=(15,8))
    ax = [plt.subplot2grid((2,8), (0,1), colspan=2), plt.subplot2grid((2,8), (0,3), colspan=2), plt.subplot2grid((2,8), (0,5), colspan=2), plt.subplot2grid((2,8), (1,0), colspan=2), plt.subplot2grid((2,8), (1,2), colspan=2), plt.subplot2grid((2,8), (1,4), colspan=2), plt.subplot2grid((2,8), (1,6), colspan=2)]
    for a in range(len(comparisons)):
        ax[a].set_title(comparisons[a][0]+r' $vs$ '+comparisons[a][1])
        this_comparison = new_genes.loc[:, comparisons[a]]
        this_comparison = this_comparison[this_comparison.max(axis=1) > 0]
        genes_comp = list(this_comparison.index.values)
        fc, pval, colors_p, this_fc = [], [], [], []
        for b in range(len(genes_comp)):
            c1, c2 = list(this_comparison.loc[genes_comp[b], comparisons[a][0]]), list(this_comparison.loc[genes_comp[b], comparisons[a][1]])
            
            c1, c2 = [0.000001 if x == 0 else x for x in c1], [0.000001 if x == 0 else x for x in c2]
            c1, c2 = [math.log2(c1[c]) for c in range(len(c1))], [math.log2(c2[c]) for c in range(len(c2))]
            t, p = stats.ttest_ind(c1, c2)
            c1, c2 = np.median(c1), np.median(c2)
            diff = c2-c1
            if diff < 1 and diff > 0: diff = -(1/diff)
            fc.append(diff), pval.append(p)
            this_fc.append([genes_comp[b], diff, p])
        colors_p_all = [colors_dict[comparisons[a][0]], colors_dict[comparisons[a][1]]]
        for c in range(len(fc)):
            if fc[c] >= 2 and pval[c] <= 0.01: colors_p.append(colors_p_all[1])
            elif fc[c] <= -2 and pval[c] <= 0.01: colors_p.append(colors_p_all[0])
            else: colors_p.append('#B1B1B1')
            pval[c] = -math.log10(pval[c])
        com = comparisons[a][0]+'_'+comparisons[a][1]
        if a == 0:
          df_pval = pd.DataFrame(this_fc, columns=['Genes/pathways', com+'_FC', com+'_p'])
          df_pval.set_index('Genes/pathways', inplace=True)
        else:
          new_df = pd.DataFrame(this_fc, columns=['Genes/pathways', com+'_FC', com+'_p'])
          new_df.set_index('Genes/pathways', inplace=True)
          df_pval = pd.concat([df_pval, new_df], axis=1, join='outer')
        ma = max([max(fc), abs(min(fc))])+0.5
        if ma > 10: ma = 10.5
        ax[a].set_xlim([-ma, ma])
        ax[a].scatter(fc, pval, marker='o', c=colors_p, s=10)
        ax[a].set_xlabel(r'Log$_2$ fold change')
        if a == 0 or a == 3:
          ax[a].set_ylabel('-log(p value)')
        plt.sca(ax[a])
    df_pval.to_csv(fp+name_csv+'.csv')
    handles = [Patch(facecolor=colors_dict[color], edgecolor='k', label=color) for color in colors_dict]
    ax[2].legend(handles=handles, bbox_to_anchor=(1.04,1), loc='upper left')
    return
```

### Number of differentially abundant pathways between body sites (Uniref90)

Volcano plot showing the number of differentially abundant pathways between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in. 
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this. 

```{python, func_humann2_6, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(pathways_90)
get_diff_abundant(pways, 'pathways_90')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

### Number of differentially abundant pathways between body sites (Uniref50)

Volcano plot showing the number of differentially abundant pathways between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in. 
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this.

```{python, func_humann2_7, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(pathways_50)
get_diff_abundant(pways, 'pathways_50')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

### Number of differentially abundant gene families between body sites (Uniref90)

Volcano plot showing the number of differentially abundant gene families between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in. 
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this.

```{python, func_humann2_8, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(genes_90)
get_diff_abundant(pways, 'genes_90')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

### Number of differentially abundant gene families between body sites (Uniref50)

Volcano plot showing the number of differentially abundant gene families between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in.
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this.

```{python, func_humann2_9, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(genes_50)
get_diff_abundant(pways, 'genes_50')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

### Number of differentially abundant KEGG orthologs between body sites (Uniref90)

Volcano plot showing the number of differentially abundant gene families between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in.
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this.

```{python, func_humann2_10a, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(genes_90_ko)
get_diff_abundant(pways, 'genes_90_ko')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

### Number of differentially abundant KEGG orthologs between body sites (Uniref50)

Volcano plot showing the number of differentially abundant KEGG orthologs between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in.
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this.

```{python, func_humann2_10, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(genes_50_ko)
get_diff_abundant(pways, 'genes_50_ko')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```


## HUMAnN3 differential abundance between body sites {.tabset}

Now we are looking at the gene families and pathways that are differentially abundant between body sites, using the lists that are output by HUMAnN3. 

### Number of differentially abundant pathways between body sites

Volcano plot showing the number of differentially abundant pathways between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in. 
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this. 

```{python, func_humann3_1, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(pathways_h3)
get_diff_abundant(pways, 'pathways', fp = '/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

### Number of differentially abundant gene families between body sites

Volcano plot showing the number of differentially abundant pathways between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in. 
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this.

```{python, func_humann3_2, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(genes_h3)
get_diff_abundant(pways, 'genes', fp = '/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

### Number of differentially abundant KEGG orthologs between body sites (Uniref90)

Volcano plot showing the number of differentially abundant pathways between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in. 
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this.

Note that now we have a separation of Uniref90 and Uniref50 because this is what the mapping files do. I'm not sure yet if it's appropriate to just combine these two tables or now. 

```{python, func_humann3_ko_1, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(genes_h3_ko_90)
get_diff_abundant(pways, 'genes_ko_90', fp = '/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

### Number of differentially abundant KEGG orthologs between body sites (Uniref50)

Volcano plot showing the number of differentially abundant pathways between body sites. 
Colored points denote genes that are significantly differentially abundant (T-test p<0.01 and log2 (median) FC > 2 or < -2) and they are colored by which body site they are highest in abundance in. 
Note that very large fold changes are not shown. Initial results showed that there were very few genes with > 10 or < -10, so the maximum plotted is this.

Note that now we have a separation of Uniref90 and Uniref50 because this is what the mapping files do. I'm not sure yet if it's appropriate to just combine these two tables or now. 

```{python, func_humann3_ko_2, results='hide', fig.keep='all', cache=TRUE}
pways = pd.DataFrame(genes_h3_ko_50)
get_diff_abundant(pways, 'genes_ko_50', fp = '/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/')
plt.subplots_adjust(hspace=0.5, wspace=0.5)
plt.show()
```

## HUMAnN most differentially abundant pathways

For each of these, I am initially filtering out to only P values that are below 0.01, and then am taking only the most up- or down-regulated pathways (50 of each) and printing them to a table.

## HUMAnN2 {.tabset}

For HUMAnN2 I am now looking at the pathways assigned to the Uniref50 database.

```{python, diff_path_1a, results='hide', fig.keep='all'}
pathways = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/pathways_50.csv', header=0, index_col=0)
```

### Blood vs saliva

Negative fold changes are higher in saliva samples while positive are higher in blood samples.

```{python, diff_path_1pa, results='hide', fig.keep='all', cache=TRUE}
bs = pathways.loc[:, ['Saliva_Blood_FC', 'Saliva_Blood_p']]
bs = bs[bs['Saliva_Blood_p'] < 0.01]
bs = bs.sort_values(by=['Saliva_Blood_FC'], ascending=False).rename(columns={'Saliva_Blood_FC':'Log2 fold change', 'Saliva_Blood_p':'p'})
bs = pd.concat([bs.head(n=50), bs.tail(n=50)])
bs.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_2, cache=TRUE}
py$bs %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal

Negative fold changes are higher in buccal samples while positive are higher in blood samples. 

```{python, diff_path_3, results='hide', fig.keep='all', cache=TRUE}
bb = pathways.loc[:, ['Buccal_Blood_FC', 'Buccal_Blood_p']]
bb = bb[bb['Buccal_Blood_p'] < 0.01]
bb= bb.sort_values(by=['Buccal_Blood_FC'], ascending=False).rename(columns={'Buccal_Blood_FC':'Log2 fold change', 'Buccal_Blood_p':'p'})
bb = pd.concat([bb.head(n=50), bb.tail(n=50)])
bb.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_4, cache=TRUE}
py$bb %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs buccal

Negative fold changes are higher in saliva samples while positive are higher in buccal samples.

```{python, diff_path_5, results='hide', fig.keep='all', cache=TRUE}
sb = pathways.loc[:, ['Saliva_Buccal_FC', 'Saliva_Buccal_p']]
sb = sb[sb['Saliva_Buccal_p'] < 0.01]
sb= sb.sort_values(by=['Saliva_Buccal_FC'], ascending=False).rename(columns={'Saliva_Buccal_FC':'Log2 fold change', 'Saliva_Buccal_p':'p'})
sb = pd.concat([sb.head(n=50), sb.tail(n=50)])
sb.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_6, cache=TRUE}
py$sb %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs saliva_euk

Negative fold changes are higher in saliva samples while positive are higher in saliva_euk samples. 

```{python, diff_path_7, results='hide', fig.keep='all', cache=TRUE}
ss = pathways.loc[:, ['Saliva_Saliva_euk_FC', 'Saliva_Saliva_euk_p']]
ss = ss[ss['Saliva_Saliva_euk_p'] < 0.01]
ss = ss.sort_values(by=['Saliva_Saliva_euk_FC'], ascending=False).rename(columns={'Saliva_Saliva_euk_FC':'Log2 fold change', 'Saliva_Saliva_euk_p':'p'})
ss = pd.concat([ss.head(n=50), ss.tail(n=50)])
ss = ss.groupby(by=ss.index, axis=0).mean()
ss.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_8, cache=TRUE}
py$ss %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Buccal vs buccal_euk

Negative fold changes are higher in buccal samples while positive are higher in buccal_euk samples.

```{python, diff_path_9, results='hide', fig.keep='all', cache=TRUE}
bbe = pathways.loc[:, ['Buccal_Buccal_euk_FC', 'Buccal_Buccal_euk_p']]
bbe = bbe[bbe['Buccal_Buccal_euk_p'] < 0.01]
bbe = bbe.sort_values(by=['Buccal_Buccal_euk_FC'], ascending=False).rename(columns={'Buccal_Buccal_euk_FC':'Log2 fold change', 'Buccal_Buccal_euk_p':'p'})
bbe = pd.concat([bbe.head(n=50), bbe.tail(n=50)])
bbe = bbe.groupby(by=bbe.index, axis=0).mean()
bbe.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_10, cache=TRUE}
py$bbe %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs saliva_euk

Negative fold changes are higher in blood samples while positive are higher in saliva_euk samples. 

```{python, diff_path_11, results='hide', fig.keep='all', cache=TRUE}
blbe = pathways.loc[:, ['Blood_Saliva_euk_FC', 'Blood_Saliva_euk_p']]
blbe = blbe[blbe['Blood_Saliva_euk_p'] < 0.01]
blbe = blbe.sort_values(by=['Blood_Saliva_euk_FC'], ascending=False).rename(columns={'Blood_Saliva_euk_FC':'Log2 fold change', 'Blood_Saliva_euk_p':'p'})
blbe = pd.concat([blbe.head(n=50), blbe.tail(n=50)])
blbe = blbe.groupby(by=blbe.index, axis=0).mean()
blbe.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_12, cache=TRUE}
py$blbe %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal_euk

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_path_13, results='hide', fig.keep='all'}
blse = pathways.loc[:, ['Blood_Buccal_euk_FC', 'Blood_Buccal_euk_p']]
blse = blse[blse['Blood_Buccal_euk_p'] < 0.01]
blse = blse.sort_values(by=['Blood_Buccal_euk_FC'], ascending=False).rename(columns={'Blood_Buccal_euk_FC':'Log2 fold change', 'Blood_Buccal_euk_p':'p'})
blse = pd.concat([blse.head(n=50), blse.tail(n=50)])
blse = blse.groupby(by=blse.index, axis=0).mean()
blse.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_14, cache=TRUE}
py$blse %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

## HUMAnN3 {.tabset}

```{python, diff_path_h3_1, results='hide', fig.keep='all'}
pathways = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/pathways.csv', header=0, index_col=0)
```

### Blood vs saliva

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_path_h3_2, results='hide', fig.keep='all', cache=TRUE}
bs3 = pathways.loc[:, ['Saliva_Blood_FC', 'Saliva_Blood_p']]
bs3 = bs3[bs3['Saliva_Blood_p'] < 0.01]
bs3 = bs3.sort_values(by=['Saliva_Blood_FC'], ascending=False).rename(columns={'Saliva_Blood_FC':'Log2 fold change', 'Saliva_Blood_p':'p'})
bs3 = pd.concat([bs3.head(n=50), bs3.tail(n=50)])
bs3 = bs3.groupby(by=bs3.index, axis=0).mean()
bs3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_h3_3a, cache=TRUE}
py$bs3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal

Negative fold changes are higher in buccal samples while positive are higher in blood samples.

```{python, diff_path_h3_3, results='hide', fig.keep='all', cache=TRUE}
bb3 = pathways.loc[:, ['Buccal_Blood_FC', 'Buccal_Blood_p']]
bb3 = bb3[bb3['Buccal_Blood_p'] < 0.01]
bb3 = bb3.sort_values(by=['Buccal_Blood_FC'], ascending=False).rename(columns={'Buccal_Blood_FC':'Log2 fold change', 'Buccal_Blood_p':'p'})
bb3 = pd.concat([bb3.head(n=50), bb3.tail(n=50)])
bb3 = bb3.groupby(by=bb3.index, axis=0).mean()
bb3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_h3_4, cache=TRUE}
py$bb3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs buccal

Negative fold changes are higher in saliva samples while positive are higher in buccal samples. 

```{python, diff_path_h3_5, results='hide', fig.keep='all', cache=TRUE}
sb3 = pathways.loc[:, ['Saliva_Buccal_FC', 'Saliva_Buccal_p']]
sb3 = sb3[sb3['Saliva_Buccal_p'] < 0.01]
sb3 = sb3.sort_values(by=['Saliva_Buccal_FC'], ascending=False).rename(columns={'Saliva_Buccal_FC':'Log2 fold change', 'Saliva_Buccal_p':'p'})
sb3 = pd.concat([sb3.head(n=50), sb3.tail(n=50)])
sb3 = sb3.groupby(by=sb3.index, axis=0).mean()
sb3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_h3_6, cache=TRUE}
py$sb3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs saliva_euk

Negative fold changes are higher in saliva samples while positive are higher in saliva_euk samples. 

```{python, diff_path_h3_7, results='hide', fig.keep='all'}
ss3 = pathways.loc[:, ['Saliva_Saliva_euk_FC', 'Saliva_Saliva_euk_p']]
ss3 = ss3[ss3['Saliva_Saliva_euk_p'] < 0.01]
ss3 = ss3.sort_values(by=['Saliva_Saliva_euk_FC'], ascending=False).rename(columns={'Saliva_Saliva_euk_FC':'Log2 fold change', 'Saliva_Saliva_euk_p':'p'})
ss3 = pd.concat([ss3.head(n=50), ss3.tail(n=50)])
ss3 = ss3.groupby(by=ss3.index, axis=0).mean()
ss3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_h3_8, cache=TRUE}
py$ss3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Buccal vs buccal_euk

Negative fold changes are higher in buccal samples while positive are higher in buccal_euk samples. 

```{python, diff_path_h3_9, results='hide', fig.keep='all'}
bbe3 = pathways.loc[:, ['Buccal_Buccal_euk_FC', 'Buccal_Buccal_euk_p']]
bbe3 = bbe3[bbe3['Buccal_Buccal_euk_p'] < 0.01]
bbe3 = bbe3.sort_values(by=['Buccal_Buccal_euk_FC'], ascending=False).rename(columns={'Buccal_Buccal_euk_FC':'Log2 fold change', 'Buccal_Buccal_euk_p':'p'})
bbe3 = pd.concat([bbe3.head(n=50), bbe3.tail(n=50)])
bbe3 = bbe3.groupby(by=bbe3.index, axis=0).mean()
bbe3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_h3_10, cache=TRUE}
py$bbe3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs saliva_euk

Negative fold changes are higher in blood samples while positive are higher in saliva_euk samples.

```{python, diff_path_h3_11, results='hide', fig.keep='all'}
blbe3 = pathways.loc[:, ['Blood_Saliva_euk_FC', 'Blood_Saliva_euk_p']]
blbe3 = blbe3[blbe3['Blood_Saliva_euk_p'] < 0.01]
blbe3 = blbe3.sort_values(by=['Blood_Saliva_euk_FC'], ascending=False).rename(columns={'Blood_Saliva_euk_FC':'Log2 fold change', 'Blood_Saliva_euk_p':'p'})
blbe3 = pd.concat([blbe3.head(n=50), blbe3.tail(n=50)])
blbe3 = blbe3.groupby(by=blbe3.index, axis=0).mean()
blbe3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_h3_12, cache=TRUE}
py$blbe3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal_euk

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_path_h3_13, results='hide', fig.keep='all'}
blse3 = pathways.loc[:, ['Blood_Buccal_euk_FC', 'Blood_Buccal_euk_p']]
blse3 = blse3[blse3['Blood_Buccal_euk_p'] < 0.01]
blse3 = blse3.sort_values(by=['Blood_Buccal_euk_FC'], ascending=False).rename(columns={'Blood_Buccal_euk_FC':'Log2 fold change', 'Blood_Buccal_euk_p':'p'})
blse3 = pd.concat([blse3.head(n=50), blse.tail(n=50)])
blse3 = blse3.groupby(by=blse3.index, axis=0).mean()
blse3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_path_h3_14, cache=TRUE}
py$blse3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

## HUMAnN most differentially abundant gene families

For each of these, I am initially filtering out to only P values that are below 0.01, and then am taking only the most up- or down-regulated gene families and printing them to a table.

## HUMAnN2 {.tabset}

For HUMAnN2 I am now looking at the gene families assigned to the Uniref50 database.

```{python, diff_path_1gf, results='hide', fig.keep='all'}
genes = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/genes_50.csv', header=0, index_col=0)
```

### Blood vs saliva

Negative fold changes are higher in saliva samples while positive are higher in blood samples.

```{python, diff_genes_1, results='hide', fig.keep='all', cache=TRUE}
bs = genes.loc[:, ['Saliva_Blood_FC', 'Saliva_Blood_p']]
bs = bs[bs['Saliva_Blood_p'] < 0.01]
bs = bs.sort_values(by=['Saliva_Blood_FC'], ascending=False).rename(columns={'Saliva_Blood_FC':'Log2 fold change', 'Saliva_Blood_p':'p'})
bs = pd.concat([bs.head(n=50), bs.tail(n=50)])
bs.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_2, cache=TRUE}
py$bs %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal

Negative fold changes are higher in buccal samples while positive are higher in blood samples. 

```{python, diff_genes_3, results='hide', fig.keep='all', cache=TRUE}
bb = genes.loc[:, ['Buccal_Blood_FC', 'Buccal_Blood_p']]
bb = bb[bb['Buccal_Blood_p'] < 0.01]
bb= bb.sort_values(by=['Buccal_Blood_FC'], ascending=False).rename(columns={'Buccal_Blood_FC':'Log2 fold change', 'Buccal_Blood_p':'p'})
bb = pd.concat([bb.head(n=50), bb.tail(n=50)])
bb.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_4, cache=TRUE}
py$bb %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs buccal

Negative fold changes are higher in saliva samples while positive are higher in buccal samples.

```{python, diff_genes_5, results='hide', fig.keep='all', cache=TRUE}
sb = genes.loc[:, ['Saliva_Buccal_FC', 'Saliva_Buccal_p']]
sb = sb[sb['Saliva_Buccal_p'] < 0.01]
sb= sb.sort_values(by=['Saliva_Buccal_FC'], ascending=False).rename(columns={'Saliva_Buccal_FC':'Log2 fold change', 'Saliva_Buccal_p':'p'})
sb = pd.concat([sb.head(n=50), sb.tail(n=50)])
sb.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_6, cache=TRUE}
py$sb %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs saliva_euk

Negative fold changes are higher in saliva samples while positive are higher in saliva_euk samples. 

```{python, diff_genes_7, results='hide', fig.keep='all', cache=TRUE}
ss = genes.loc[:, ['Saliva_Saliva_euk_FC', 'Saliva_Saliva_euk_p']]
ss = ss[ss['Saliva_Saliva_euk_p'] < 0.01]
ss = ss.sort_values(by=['Saliva_Saliva_euk_FC'], ascending=False).rename(columns={'Saliva_Saliva_euk_FC':'Log2 fold change', 'Saliva_Saliva_euk_p':'p'})
ss = pd.concat([ss.head(n=50), ss.tail(n=50)])
ss.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_8, cache=TRUE}
py$ss %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Buccal vs buccal_euk

Negative fold changes are higher in buccal samples while positive are higher in buccal_euk samples.

```{python, diff_genes_9, results='hide', fig.keep='all', cache=TRUE}
bbe = genes.loc[:, ['Buccal_Buccal_euk_FC', 'Buccal_Buccal_euk_p']]
bbe = bbe[bbe['Buccal_Buccal_euk_p'] < 0.01]
bbe = bbe.sort_values(by=['Buccal_Buccal_euk_FC'], ascending=False).rename(columns={'Buccal_Buccal_euk_FC':'Log2 fold change', 'Buccal_Buccal_euk_p':'p'})
bbe = pd.concat([bbe.head(n=50), bbe.tail(n=50)])
bbe = bbe.groupby(by=bbe.index, axis=0).mean()
bbe.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_10, cache=TRUE}
py$bbe %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs saliva_euk

Negative fold changes are higher in blood samples while positive are higher in saliva_euk samples. 

```{python, diff_genes_11, results='hide', fig.keep='all', cache=TRUE}
blbe = genes.loc[:, ['Blood_Saliva_euk_FC', 'Blood_Saliva_euk_p']]
blbe = blbe[blbe['Blood_Saliva_euk_p'] < 0.01]
blbe = blbe.sort_values(by=['Blood_Saliva_euk_FC'], ascending=False).rename(columns={'Blood_Saliva_euk_FC':'Log2 fold change', 'Blood_Saliva_euk_p':'p'})
blbe = pd.concat([blbe.head(n=50), blbe.tail(n=50)])
blbe = blbe.groupby(by=blbe.index, axis=0).mean()
blbe.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_12, cache=TRUE}
py$blbe %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal_euk

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_genes_13, results='hide', fig.keep='all'}
blse = genes.loc[:, ['Blood_Buccal_euk_FC', 'Blood_Buccal_euk_p']]
blse = blse[blse['Blood_Buccal_euk_p'] < 0.01]
blse = blse.sort_values(by=['Blood_Buccal_euk_FC'], ascending=False).rename(columns={'Blood_Buccal_euk_FC':'Log2 fold change', 'Blood_Buccal_euk_p':'p'})
blse = pd.concat([blse.head(n=50), blse.tail(n=50)])
blse = blse.groupby(by=blse.index, axis=0).mean()
blse.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_14, cache=TRUE}
py$blse %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

## HUMAnN3 {.tabset}

```{python, diff_genes_h3_1, results='hide', fig.keep='all'}
genes = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/genes.csv', header=0, index_col=0)
```

### Blood vs saliva

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_genes_h3_2, results='hide', fig.keep='all', cache=TRUE}
bs3 = genes.loc[:, ['Saliva_Blood_FC', 'Saliva_Blood_p']]
bs3 = bs3[bs3['Saliva_Blood_p'] < 0.01]
bs3 = bs3.sort_values(by=['Saliva_Blood_FC'], ascending=False).rename(columns={'Saliva_Blood_FC':'Log2 fold change', 'Saliva_Blood_p':'p'})
bs3 = pd.concat([bs3.head(n=50), bs3.tail(n=50)])
bs3 = bs3.groupby(by=bs3.index, axis=0).mean()
bs3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_h3_3a, cache=TRUE}
py$bs3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal

Negative fold changes are higher in buccal samples while positive are higher in blood samples.

```{python, diff_genes_h3_3, results='hide', fig.keep='all', cache=TRUE}
bb3 = genes.loc[:, ['Buccal_Blood_FC', 'Buccal_Blood_p']]
bb3 = bb3[bb3['Buccal_Blood_p'] < 0.01]
bb3 = bb3.sort_values(by=['Buccal_Blood_FC'], ascending=False).rename(columns={'Buccal_Blood_FC':'Log2 fold change', 'Buccal_Blood_p':'p'})
bb3 = pd.concat([bb3.head(n=50), bb3.tail(n=50)])
bb3 = bb3.groupby(by=bb3.index, axis=0).mean()
bb3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_h3_4, cache=TRUE}
py$bb3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs buccal

Negative fold changes are higher in saliva samples while positive are higher in buccal samples. 

```{python, diff_genes_h3_5, results='hide', fig.keep='all', cache=TRUE}
sb3 = genes.loc[:, ['Saliva_Buccal_FC', 'Saliva_Buccal_p']]
sb3 = sb3[sb3['Saliva_Buccal_p'] < 0.01]
sb3 = sb3.sort_values(by=['Saliva_Buccal_FC'], ascending=False).rename(columns={'Saliva_Buccal_FC':'Log2 fold change', 'Saliva_Buccal_p':'p'})
sb3 = pd.concat([sb3.head(n=50), sb3.tail(n=50)])
sb3 = sb3.groupby(by=sb3.index, axis=0).mean()
sb3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_h3_6, cache=TRUE}
py$sb3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs saliva_euk

Negative fold changes are higher in saliva samples while positive are higher in saliva_euk samples. 

```{python, diff_genes_h3_7, results='hide', fig.keep='all'}
ss3 = genes.loc[:, ['Saliva_Saliva_euk_FC', 'Saliva_Saliva_euk_p']]
ss3 = ss3[ss3['Saliva_Saliva_euk_p'] < 0.01]
ss3 = ss3.sort_values(by=['Saliva_Saliva_euk_FC'], ascending=False).rename(columns={'Saliva_Saliva_euk_FC':'Log2 fold change', 'Saliva_Saliva_euk_p':'p'})
ss3 = pd.concat([ss3.head(n=50), ss3.tail(n=50)])
ss3 = ss3.groupby(by=ss3.index, axis=0).mean()
ss3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_h3_8, cache=TRUE}
py$ss3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Buccal vs buccal_euk

Negative fold changes are higher in buccal samples while positive are higher in buccal_euk samples. 

```{python, diff_genes_h3_9, results='hide', fig.keep='all'}
bbe3 = genes.loc[:, ['Buccal_Buccal_euk_FC', 'Buccal_Buccal_euk_p']]
bbe3 = bbe3[bbe3['Buccal_Buccal_euk_p'] < 0.01]
bbe3 = bbe3.sort_values(by=['Buccal_Buccal_euk_FC'], ascending=False).rename(columns={'Buccal_Buccal_euk_FC':'Log2 fold change', 'Buccal_Buccal_euk_p':'p'})
bbe3 = pd.concat([bbe3.head(n=50), bbe3.tail(n=50)])
bbe3 = bbe3.groupby(by=bbe3.index, axis=0).mean()
bbe3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_h3_10, cache=TRUE}
py$bbe3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs saliva_euk

Negative fold changes are higher in blood samples while positive are higher in saliva_euk samples.

```{python, diff_genes_h3_11, results='hide', fig.keep='all'}
blbe3 = genes.loc[:, ['Blood_Saliva_euk_FC', 'Blood_Saliva_euk_p']]
blbe3 = blbe3[blbe3['Blood_Saliva_euk_p'] < 0.01]
blbe3 = blbe3.sort_values(by=['Blood_Saliva_euk_FC'], ascending=False).rename(columns={'Blood_Saliva_euk_FC':'Log2 fold change', 'Blood_Saliva_euk_p':'p'})
blbe3 = pd.concat([blbe3.head(n=50), blbe3.tail(n=50)])
blbe3 = blbe3.groupby(by=blbe3.index, axis=0).mean()
blbe3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_h3_12, cache=TRUE}
py$blbe3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal_euk

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_genes_h3_13, results='hide', fig.keep='all'}
blse3 = genes.loc[:, ['Blood_Buccal_euk_FC', 'Blood_Buccal_euk_p']]
blse3 = blse3[blse3['Blood_Buccal_euk_p'] < 0.01]
blse3 = blse3.sort_values(by=['Blood_Buccal_euk_FC'], ascending=False).rename(columns={'Blood_Buccal_euk_FC':'Log2 fold change', 'Blood_Buccal_euk_p':'p'})
blse3 = pd.concat([blse3.head(n=50), blse.tail(n=50)])
blse3 = blse3.groupby(by=blse3.index, axis=0).mean()
blse3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_genes_h3_14, cache=TRUE}
py$blse3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

## HUMAnN most differentially abundant KEGG orthologs

For each of these, I am initially filtering out to only P values that are below 0.01, and then am taking only the most up- or down-regulated KEGG orthologs and printing them to a table.

## HUMAnN2 {.tabset}

For HUMAnN2 I am now looking at the KEGG orthologs assigned to the Uniref50 database, as above for the gene families.

```{python, diff_path_1ko, results='hide', fig.keep='all'}
kos = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann/processing/genes_50_ko.csv', header=0, index_col=0)
kegg_list = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kegg_list.csv', header=0, index_col=0)
kegg_list.drop_duplicates(inplace=True)
kegg_dict = {}
kl = list(kegg_list.index.values)
for ko in kl:
  kegg_dict[ko] = ko+' '+kegg_list.loc[ko, 'Product']
kos = kos.rename(index=kegg_dict)
```

### Blood vs saliva

Negative fold changes are higher in saliva samples while positive are higher in blood samples.

```{python, diff_kos_1, results='hide', fig.keep='all', cache=TRUE}
bs = kos.loc[:, ['Saliva_Blood_FC', 'Saliva_Blood_p']]
bs = bs[bs['Saliva_Blood_p'] < 0.01]
bs = bs.sort_values(by=['Saliva_Blood_FC'], ascending=False).rename(columns={'Saliva_Blood_FC':'Log2 fold change', 'Saliva_Blood_p':'p'})
bs = pd.concat([bs.head(n=50), bs.tail(n=50)])
bs.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_2, cache=TRUE}
py$bs %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal

Negative fold changes are higher in buccal samples while positive are higher in blood samples. 

```{python, diff_kos_3, results='hide', fig.keep='all', cache=TRUE}
bb = kos.loc[:, ['Buccal_Blood_FC', 'Buccal_Blood_p']]
bb = bb[bb['Buccal_Blood_p'] < 0.01]
bb= bb.sort_values(by=['Buccal_Blood_FC'], ascending=False).rename(columns={'Buccal_Blood_FC':'Log2 fold change', 'Buccal_Blood_p':'p'})
bb = pd.concat([bb.head(n=50), bb.tail(n=50)])
bb.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_4, cache=TRUE}
py$bb %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs buccal

Negative fold changes are higher in saliva samples while positive are higher in buccal samples.

```{python, diff_kos_5, results='hide', fig.keep='all', cache=TRUE}
sb = kos.loc[:, ['Saliva_Buccal_FC', 'Saliva_Buccal_p']]
sb = sb[sb['Saliva_Buccal_p'] < 0.01]
sb= sb.sort_values(by=['Saliva_Buccal_FC'], ascending=False).rename(columns={'Saliva_Buccal_FC':'Log2 fold change', 'Saliva_Buccal_p':'p'})
sb = pd.concat([sb.head(n=50), sb.tail(n=50)])
sb.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_6, cache=TRUE}
py$sb %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs saliva_euk

Negative fold changes are higher in saliva samples while positive are higher in saliva_euk samples. 

```{python, diff_kos_7, results='hide', fig.keep='all', cache=TRUE}
ss = kos.loc[:, ['Saliva_Saliva_euk_FC', 'Saliva_Saliva_euk_p']]
ss = ss[ss['Saliva_Saliva_euk_p'] < 0.01]
ss = ss.sort_values(by=['Saliva_Saliva_euk_FC'], ascending=False).rename(columns={'Saliva_Saliva_euk_FC':'Log2 fold change', 'Saliva_Saliva_euk_p':'p'})
ss = pd.concat([ss.head(n=50), ss.tail(n=50)])
ss.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_8, cache=TRUE}
py$ss %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Buccal vs buccal_euk

Negative fold changes are higher in buccal samples while positive are higher in buccal_euk samples.

```{python, diff_kos_9, results='hide', fig.keep='all', cache=TRUE}
bbe = kos.loc[:, ['Buccal_Buccal_euk_FC', 'Buccal_Buccal_euk_p']]
bbe = bbe[bbe['Buccal_Buccal_euk_p'] < 0.01]
bbe = bbe.sort_values(by=['Buccal_Buccal_euk_FC'], ascending=False).rename(columns={'Buccal_Buccal_euk_FC':'Log2 fold change', 'Buccal_Buccal_euk_p':'p'})
bbe = pd.concat([bbe.head(n=50), bbe.tail(n=50)])
bbe = bbe.groupby(by=bbe.index, axis=0).mean()
bbe.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_10, cache=TRUE}
py$bbe %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs saliva_euk

Negative fold changes are higher in blood samples while positive are higher in saliva_euk samples. 

```{python, diff_kos_11, results='hide', fig.keep='all', cache=TRUE}
blbe = kos.loc[:, ['Blood_Saliva_euk_FC', 'Blood_Saliva_euk_p']]
blbe = blbe[blbe['Blood_Saliva_euk_p'] < 0.01]
blbe = blbe.sort_values(by=['Blood_Saliva_euk_FC'], ascending=False).rename(columns={'Blood_Saliva_euk_FC':'Log2 fold change', 'Blood_Saliva_euk_p':'p'})
blbe = pd.concat([blbe.head(n=50), blbe.tail(n=50)])
blbe = blbe.groupby(by=blbe.index, axis=0).mean()
blbe.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_12, cache=TRUE}
py$blbe %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal_euk

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_kos_13, results='hide', fig.keep='all'}
blse = kos.loc[:, ['Blood_Buccal_euk_FC', 'Blood_Buccal_euk_p']]
blse = blse[blse['Blood_Buccal_euk_p'] < 0.01]
blse = blse.sort_values(by=['Blood_Buccal_euk_FC'], ascending=False).rename(columns={'Blood_Buccal_euk_FC':'Log2 fold change', 'Blood_Buccal_euk_p':'p'})
blse = pd.concat([blse.head(n=50), blse.tail(n=50)])
blse = blse.groupby(by=blse.index, axis=0).mean()
blse.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_14, cache=TRUE}
py$blse %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

## HUMAnN3 {.tabset}

As for HUMAnN2 I am now looking at the KEGG orthologs assigned to the Uniref50 database.

```{python, diff_kos_h3_1, results='hide', fig.keep='all'}
kos = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/metaphlan_humann3/genes_ko_50.csv', header=0, index_col=0)
kegg_list = pd.read_csv('/Users/robynwright/Documents/OneDrive/Github/Human_metagenome/kegg_list.csv', header=0, index_col=0)
kegg_list.drop_duplicates(inplace=True)
kegg_dict = {}
kl = list(kegg_list.index.values)
for ko in kl:
  kegg_dict[ko] = ko+' '+kegg_list.loc[ko, 'Product']
kos = kos.rename(index=kegg_dict)
```

### Blood vs saliva

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_kos_h3_2, results='hide', fig.keep='all', cache=TRUE}
bs3 = kos.loc[:, ['Saliva_Blood_FC', 'Saliva_Blood_p']]
bs3 = bs3[bs3['Saliva_Blood_p'] < 0.01]
bs3 = bs3.sort_values(by=['Saliva_Blood_FC'], ascending=False).rename(columns={'Saliva_Blood_FC':'Log2 fold change', 'Saliva_Blood_p':'p'})
bs3 = pd.concat([bs3.head(n=50), bs3.tail(n=50)])
bs3 = bs3.groupby(by=bs3.index, axis=0).mean()
bs3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_h3_3a, cache=TRUE}
py$bs3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal

Negative fold changes are higher in buccal samples while positive are higher in blood samples.

```{python, diff_kos_h3_3, results='hide', fig.keep='all', cache=TRUE}
bb3 = kos.loc[:, ['Buccal_Blood_FC', 'Buccal_Blood_p']]
bb3 = bb3[bb3['Buccal_Blood_p'] < 0.01]
bb3 = bb3.sort_values(by=['Buccal_Blood_FC'], ascending=False).rename(columns={'Buccal_Blood_FC':'Log2 fold change', 'Buccal_Blood_p':'p'})
bb3 = pd.concat([bb3.head(n=50), bb3.tail(n=50)])
bb3 = bb3.groupby(by=bb3.index, axis=0).mean()
bb3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_h3_4, cache=TRUE}
py$bb3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs buccal

Negative fold changes are higher in saliva samples while positive are higher in buccal samples. 

```{python, diff_kos_h3_5, results='hide', fig.keep='all', cache=TRUE}
sb3 = kos.loc[:, ['Saliva_Buccal_FC', 'Saliva_Buccal_p']]
sb3 = sb3[sb3['Saliva_Buccal_p'] < 0.01]
sb3 = sb3.sort_values(by=['Saliva_Buccal_FC'], ascending=False).rename(columns={'Saliva_Buccal_FC':'Log2 fold change', 'Saliva_Buccal_p':'p'})
sb3 = pd.concat([sb3.head(n=50), sb3.tail(n=50)])
sb3 = sb3.groupby(by=sb3.index, axis=0).mean()
sb3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_h3_6, cache=TRUE}
py$sb3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Saliva vs saliva_euk

Negative fold changes are higher in saliva samples while positive are higher in saliva_euk samples. 

```{python, diff_kos_h3_7, results='hide', fig.keep='all'}
ss3 = kos.loc[:, ['Saliva_Saliva_euk_FC', 'Saliva_Saliva_euk_p']]
ss3 = ss3[ss3['Saliva_Saliva_euk_p'] < 0.01]
ss3 = ss3.sort_values(by=['Saliva_Saliva_euk_FC'], ascending=False).rename(columns={'Saliva_Saliva_euk_FC':'Log2 fold change', 'Saliva_Saliva_euk_p':'p'})
ss3 = pd.concat([ss3.head(n=50), ss3.tail(n=50)])
ss3 = ss3.groupby(by=ss3.index, axis=0).mean()
ss3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_h3_8, cache=TRUE}
py$ss3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Buccal vs buccal_euk

Negative fold changes are higher in buccal samples while positive are higher in buccal_euk samples. 

```{python, diff_kos_h3_9, results='hide', fig.keep='all'}
bbe3 = kos.loc[:, ['Buccal_Buccal_euk_FC', 'Buccal_Buccal_euk_p']]
bbe3 = bbe3[bbe3['Buccal_Buccal_euk_p'] < 0.01]
bbe3 = bbe3.sort_values(by=['Buccal_Buccal_euk_FC'], ascending=False).rename(columns={'Buccal_Buccal_euk_FC':'Log2 fold change', 'Buccal_Buccal_euk_p':'p'})
bbe3 = pd.concat([bbe3.head(n=50), bbe3.tail(n=50)])
bbe3 = bbe3.groupby(by=bbe3.index, axis=0).mean()
bbe3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_h3_10, cache=TRUE}
py$bbe3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs saliva_euk

Negative fold changes are higher in blood samples while positive are higher in saliva_euk samples.

```{python, diff_kos_h3_11, results='hide', fig.keep='all'}
blbe3 = kos.loc[:, ['Blood_Saliva_euk_FC', 'Blood_Saliva_euk_p']]
blbe3 = blbe3[blbe3['Blood_Saliva_euk_p'] < 0.01]
blbe3 = blbe3.sort_values(by=['Blood_Saliva_euk_FC'], ascending=False).rename(columns={'Blood_Saliva_euk_FC':'Log2 fold change', 'Blood_Saliva_euk_p':'p'})
blbe3 = pd.concat([blbe3.head(n=50), blbe3.tail(n=50)])
blbe3 = blbe3.groupby(by=blbe3.index, axis=0).mean()
blbe3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_h3_12, cache=TRUE}
py$blbe3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```

### Blood vs buccal_euk

Negative fold changes are higher in blood samples while positive are higher in buccal_euk samples. 

```{python, diff_kos_h3_13, results='hide', fig.keep='all'}
blse3 = kos.loc[:, ['Blood_Buccal_euk_FC', 'Blood_Buccal_euk_p']]
blse3 = blse3[blse3['Blood_Buccal_euk_p'] < 0.01]
blse3 = blse3.sort_values(by=['Blood_Buccal_euk_FC'], ascending=False).rename(columns={'Blood_Buccal_euk_FC':'Log2 fold change', 'Blood_Buccal_euk_p':'p'})
blse3 = pd.concat([blse3.head(n=50), blse.tail(n=50)])
blse3 = blse3.groupby(by=blse3.index, axis=0).mean()
blse3.sort_values(by=['Log2 fold change'], ascending=False, inplace=True)
```

```{r, diff_kos_h3_14, cache=TRUE}
py$blse3 %>%
  kable() %>%
  kable_styling() %>%
  scroll_box(width = "600px", height = "400px")
```


