---
title: "Repair gene copies in the genomes of unicellular organisms"
output:
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
---

```{R setup_r, include=FALSE}
library(reticulate)
library(kableExtra)
library(knitr)
library(exactRankTests)
library(nlme)
library(dplyr)
library(ggplot2)
library(compositions)
library(vegan)

opts_knit$set(root.dir = '/Users/robynwright/Documents/Github/')
print(reticulate::py_config())
```

```{python, results='hide'}
import os
import pandas as pd
import csv
import numpy as np
import math
from matplotlib.lines import Line2D
from mpl_toolkits.axes_grid1 import make_axes_locatable
import matplotlib as mpl
from matplotlib.patches import Patch
import matplotlib.pyplot as plt
import random
from scipy.spatial import distance
from scipy import stats
from sklearn import manifold
from sklearn.decomposition import PCA
import statsmodels.stats.multitest as sm
from Bio import SeqIO
```

# Summary
This file contains an exploration of:<br/>
  - The genomes of 20,000 isolates (using the annotations obtained from the IMG/JGI genome database)<br/>
  - The genomes of 10 model prokaryotes<br/>

The genes that we are looking at are:
```{python, results='hide'}
names = ['Genes', 'KEGG orthologs', 'Other names']
genes = [['Hsp70', 'K04043, K03283', 'dnaK, ssa1-4, HSP-1, HSP-70, HSPA8, HSPA1A'], ['J-protein', 'K03686, K05516, K09503, K09507', 'dnaJ, ydj1, DNJ-12, DNJ-19, DNJ-13, DNAJA2, DNAJB1'], ['Hsp110', 'K09485, K09489', 'sse1/2, HSPH1/2'], ['Hsp100 disaggregase', 'K03695', 'clpB, Hsp104'], ['sHsps', 'K04080, K04081, K13993, K19766, K13130, K04455, K09544, K09541', 'ibpA, ibpB, Hsp26, Hsp42, HSP-16.2, HSP-16.1, SIP-1, HSPB1, HSPB3, HSPB4, HSPB5'], ['Hsp60', 'K04077', 'HSPD1']]
df = pd.DataFrame(genes, columns=names)
df = df.set_index('Genes')
genes_dict = {'hsp70':['K04043', 'K03283', 'dnak'], 'dnaj':['K03686', 'K05516', 'K09503', 'K09507'], 'hsp110':['K09485', 'K09489'], 'hsp110dis':['K03695'], 'shsps':['K04080', 'K04081', 'K13993', 'K19766', 'K13130', 'K04455', 'K09544', 'K09541'], 'hsp60':['K04077']}
```

```{r}
#8
py$df %>%
  kable() %>%
  kable_styling()
```

# Look in the JGI genomes

Now we open and look at the file containing the KEGG orthologs in each of the 20,000 genomes:
```{python}
genomes_kegg = pd.read_csv('/Users/robynwright/Documents/OneDrive/Papers_writing/Aging of biofilms/Review 2/New simulations/search_for_genes/ko_all.txt', sep='\t', header=0, index_col=0)
print(genomes_kegg)
```

## Filter this to contain only the KEGG orthologs that we are interested in
```{python}
ko, ko_all = [], []
genomes_cols = list(genomes_kegg.columns)
for gene in genes_dict:
  this_gene = genes_dict[gene]
  ko_all += this_gene
  for g in this_gene:
    if g in genomes_cols:
      ko.append(g)
genomes_kegg = genomes_kegg.loc[:, ko]
print(genomes_kegg)
```

## Generate some summary statistics for each KEGG ortholog
```{python, results='hide'}
mean = pd.DataFrame(genomes_kegg.mean(axis=0)).rename(columns={0:'Abundance'})
prevalence = []
for k in ko:
  this_k = list(genomes_kegg.loc[:, k].values)
  zeroes = this_k.count(0)
  prev = zeroes/20000
  prevalence.append(1-prev)
mean['Prevalence'] = prevalence
mean = mean.transpose()
```
```{r}
py$mean %>%
  kable() %>%
  kable_styling()
```
## Check whether there are genomes with no repair genes
```{python, results='hide'}
above_zero = genomes_kegg[genomes_kegg.max(axis=1) > 0]
print(above_zero)
```
So we can see from this that there are apparently 17 genomes without any of the genes that we are looking for. 

```{python, results='hide'}
zero = genomes_kegg[genomes_kegg.sum(axis=1) == 0]
print(zero)
```
```{r}
py$zero %>%
  kable() %>%
  kable_styling()
```

If we now look at these, BLASTing the 16S sequences from these gives (top hit only, percent identity):<br/>
- 2236876029: Kaistia granuli (84.88%)<br/>
- 2510065005: Thermococcus litoralis (96.87%)<br/>
- 2537562134: Planococcus halocryophilus (99.67%)<br/>
- 2600255021: Glaesserella parasuis (97.54%)<br/>
- 2617271188: Snodgrassella alvi (99.68%)<br/>
- 2639762705: Bathymodiolus heckerae (93.39%)<br/>
- 2681812881: Thioalkalivibrio sulfidiphilus (88.92%)<br/>
- 2684622808: Staphylococcus argenteus (99.78%)<br/>
- 2700988635: Candidatus Vesicomyosocius okutanii (93.24%)<br/>
- 2706794582: Enterococcus casseliflavus (98.85%)<br/>
- 2716884108: Bacteriovorax stolpii (89.80%)<br/>
- 2728369649: Bacillus miscanthi (98.20%)<br/>
- 638154507: Methanopyrus kandleri (97.42%)<br/>
- 649990008: Streptococcus gwangjuense (98.81%)<br/>
- 2671180212: Octadecabacter ascidiaceicola (96.25%)<br/>
- 2718217655: Pyrodictium occultum (87.87%)<br/>

# Look in the genomes of model prokaryotes
```{python, results='hide'}
if mean.shape[0] != 2:
  mean = mean.transpose()
ref_genomes = '/Users/robynwright/Documents/OneDrive/Papers_writing/Aging of biofilms/Review 2/new_figshare/genomes_repair/model_prokaryote_genomes/'
list_genomes = os.listdir(ref_genomes)
all_counts = []
for gen in list_genomes:
  if '.txt' in gen:
    with open(ref_genomes+'/'+gen, 'rU') as f:
      rows, count = [], 0
      for row in csv.reader(f, delimiter='\t'):
        for r in row:
          for gene in ko_all:
            if gene in r:
              rows.append(gene)
      counts = [gen[:-4]]
      for col in list(mean.columns):
        counts.append(rows.count(col))
      all_counts.append(counts)
cols = ['Name']+list(mean.columns)
all_counts = pd.DataFrame(all_counts, columns=cols)
all_counts = all_counts.set_index('Name')
new_mean = pd.concat([mean, all_counts])
```
```{r}
py$all_counts %>%
  kable() %>%
  kable_styling()
```

## Heatmap of the abundance of these genes

```{python, results='hide', fig.keep='all'}
fig = plt.figure(figsize=(8,6))
ax1 = plt.subplot(121)
if list(new_mean.index.values)[0] == 'Abundance':
  new_mean = new_mean.iloc[::-1]

ax2 = plt.subplot(7, 2, 2, frameon=False)
ax3 = plt.subplot(7, 2, 4, frameon=False)
ax4 = plt.subplot(7, 2, 6, frameon=False)
plt.sca(ax2)
plt.xticks([]), plt.yticks([])
plt.sca(ax3)
plt.xticks([]), plt.yticks([])
plt.sca(ax4)
plt.xticks([]), plt.yticks([])

divider = make_axes_locatable(ax2)
cax = divider.append_axes('left', size="100%", frameon=True)
divider = make_axes_locatable(ax3)
cax3 = divider.append_axes('left', size="100%", frameon=True)

this_mean = new_mean.drop(['K09503', 'K09489'], axis=1)

plt.sca(ax1)
x, bottom, height = [], [], []
x_labs = list(this_mean.index.values)
y = []
for a in range(len(this_mean.columns)):
  bottom.append(0), x.append(a), height.append(1)
for a in range(12):
  if a < 10:
    colors = {0:'k', 1:'#F1B703', 2:'#2980B9'}
    this_row = []
    for b in this_mean.columns:
      this_row.append(colors[this_mean.loc[x_labs[a], b]])
    plt.bar(x, height, bottom=bottom, color=this_row, edgecolor='k', width=1)
  else:
    if a == 10:
      colormap = mpl.cm.get_cmap('Reds', 256)
      norm = mpl.colors.Normalize(vmin=min(this_mean.loc[x_labs[a], :].values), vmax=max(this_mean.loc[x_labs[a], :].values))
      for c in range(len(bottom)):
        bottom[c] += 0.5
      cb1 = mpl.colorbar.ColorbarBase(cax3, cmap=colormap, norm=norm, orientation='horizontal')
    else:
      colormap = mpl.cm.get_cmap('Blues', 256)
      norm = mpl.colors.Normalize(vmin=min(this_mean.loc[x_labs[a], :].values), vmax=max(this_mean.loc[x_labs[a], :].values))
      cb1 = mpl.colorbar.ColorbarBase(cax, cmap=colormap, norm=norm, orientation='horizontal')
    this_row = []
    m = mpl.cm.ScalarMappable(norm=norm, cmap=colormap)
    for b in this_mean.columns:
      this_row.append(m.to_rgba(this_mean.loc[x_labs[a], b]))
    plt.bar(x, height, bottom=bottom, color=this_row, edgecolor='k', width=1)
  y.append(bottom[-1]+0.5)
  for c in range(len(bottom)):
    bottom[c] += 1
labels = ['0 copies', '1 copy', '2 copies']
handles = [Patch(facecolor=colors[color], edgecolor='k', label=labels[color]) for color in colors]


ko_name_dict = {'K04043':'K04043 (dnaK)', 'K03686':'K03686 (dnaJ)', 'K05516':'K05516 (cbpA)',  'K03695':'K03695 (clpB)', 'K04080':'K04080 (sHsps)', 'K04081':'K04081 (ibpB)', 'K13993':'K13993 (hsp26\nand hsp42)', 'K04077':'K04077 (hsp60)'}

ax4.legend(handles=handles, loc='upper left', bbox_to_anchor=(0,1))
plt.xlim([x[0]-0.5, x[-1]+0.5])
plt.ylim([0, bottom[-1]])
plt.yticks(y, this_mean.index.values)
ax1.xaxis.tick_top()
this_mean.rename(columns=ko_name_dict, inplace=True)
plt.xticks(x, this_mean.columns, rotation=90)
plt.tight_layout()
plt.show()
plt.savefig('/Users/robynwright/Documents/GitHub/iDynoMiCS_1.5/biofilm_manuscript_results_analysis/repair_genes/repair_genes_abundance.png', dpi=600, bbox_inches='tight')
```