---
title: "Fungal metagenomes"
output:
  html_document: 
    toc: yes
    toc_float: yes
    code_folding: hide
---

```{R, setup_r, results='hide', include=FALSE}
library(reticulate)
library(kableExtra)
library(knitr)
library(exactRankTests)
library(nlme)
library(dplyr)
library(ggplot2)
library(compositions)
library(vegan)
library(phyloseq)
library(ape)

opts_knit$set(root.dir = '/Users/robynwright/Documents/OneDrive/Langille Lab postdoc/fungal_metagenome/')
#reticulate::py_config()
```

```{python, setup_python, results='hide', include=FALSE}
import os
from Bio import SeqIO
from Bio.SeqRecord import SeqRecord
from Bio.Seq import Seq
from Bio.Alphabet import IUPAC
import pandas as pd
import csv
import numpy as np
import math
from matplotlib.lines import Line2D
import matplotlib as mpl
from matplotlib.patches import Patch
import matplotlib.pyplot as plt
import pickle
import random
from scipy.spatial import distance
from scipy import stats
from sklearn import manifold
from sklearn.decomposition import PCA
import statsmodels.stats.multitest as sm
```

## Analysis and commands run {.tabset}

### Initial preparation

Upload samples to server:
```{bash, eval=FALSE}
rsync --partial --progress W0.tar.bz2 robyn@vulcan.pharmacology.dal.ca:/home/robyn/
rsync --partial --progress W06.tar.bz2 robyn@vulcan.pharmacology.dal.ca:/home/robyn/
rsync --partial --progress W12.tar.bz2 robyn@vulcan.pharmacology.dal.ca:/home/robyn/
```

Unzip and then gzip:
```{bash, eval=FALSE}
tar -xf W0.tar.bz2
for i in W0_samples/*.fastq ; do gzip $i ; done

tar -xf W06.tar.bz2
for i in W06_samples/*.fastq ; do gzip $i ; done

tar -xf W12.tar.bz2
for i in W12_samples/*.fastq ; do gzip $i ; done
```

### Kneaddata

**Commands run:**
```{bash, eval=FALSE}
parallel -j 1 'kneaddata -i {} -o kneaddata_out_W0/ \
-db /home/shared/bowtiedb/GRCh38_PhiX --trimmomatic /home/robyn/tools/Trimmomatic-0.39/ \
-t 12 --trimmomatic-options "SLIDINGWINDOW:4:20 MINLEN:50" \
--bowtie2-options "--sensitive --dovetail" --remove-intermediate-output' \
 ::: W0_samples/*.fastq.gz
 kneaddata_read_count_table --input kneaddata_out_W0 --output W0_kneaddata_read_counts.txt
 mkdir W0_decontaminated
 for i in kneaddata_out_W0/*kneaddata.fastq ; do mv $i W0_decontaminated ; done
 
 parallel -j 1 'kneaddata -i {} -o kneaddata_out_W06/ \
-db /home/shared/bowtiedb/GRCh38_PhiX --trimmomatic /home/robyn/tools/Trimmomatic-0.39/ \
-t 12 --trimmomatic-options "SLIDINGWINDOW:4:20 MINLEN:50" \
--bowtie2-options "--sensitive --dovetail" --remove-intermediate-output' \
 ::: W06_samples/*.fastq.gz
 kneaddata_read_count_table --input kneaddata_out_W06 --output W06_kneaddata_read_counts.txt
 for i in kneaddata_out_W06/*kneaddata.fastq ; do mv $i W06_decontaminated ; done
 
 parallel -j 1 'kneaddata -i {} -o kneaddata_out_W06/ \
-db /home/shared/bowtiedb/GRCh38_PhiX --trimmomatic /home/robyn/tools/Trimmomatic-0.39/ \
-t 12 --trimmomatic-options "SLIDINGWINDOW:4:20 MINLEN:50" \
--bowtie2-options "--sensitive --dovetail" --remove-intermediate-output' \
 ::: W12_samples/*.fastq.gz
 kneaddata_read_count_table --input kneaddata_out_W12 --output W12_kneaddata_read_counts.txt
 for i in kneaddata_out_W12/*kneaddata.fastq ; do mv $i W12_decontaminated ; done
```

**W0 percent reads kept:**
```{python, w0_kneaddata, results='hide', fig.keep='all', cache=TRUE, echo=FALSE}
reads = pd.read_csv('/Users/robynwright/Documents/OneDrive/Langille Lab postdoc/fungal_metagenome/W0_kneaddata_read_counts.txt', header=0, index_col=0, sep='\t')
perc = []
samples = list(reads.index.values)
for s in samples:
  perc.append((reads.loc[s, 'final single']/reads.loc[s, 'raw single'])*100)

plt.figure(figsize=(10,4))
ax1 = plt.subplot(111)
x = []
for a in range(len(perc)):
  ax1.bar(a, perc[a], width=0.8, color='#8E0145', edgecolor='k')
  x.append(a)
  samples[a] = samples[a].replace('-W0_kneaddata', '')

plt.xticks(x, samples, rotation=90, fontsize=8)
plt.xlim([-0.7, x[-1]+0.5])
plt.ylim([0, 100])
plt.ylabel('Reads remaining (%)')
plt.tight_layout()
plt.show()
```

**W06 percent reads kept:**
```{python, w06_kneaddata, results='hide', fig.keep='all', cache=TRUE, echo=FALSE, eval=FALSE}
reads = pd.read_csv('/Users/robynwright/Documents/OneDrive/Langille Lab postdoc/fungal_metagenome/W06_kneaddata_read_counts.txt', header=0, index_col=0, sep='\t')
perc = []
samples = list(reads.index.values)
for s in samples:
  perc.append((reads.loc[s, 'final single']/reads.loc[s, 'raw single'])*100)

plt.figure(figsize=(10,4))
ax1 = plt.subplot(111)
x = []
for a in range(len(perc)):
  ax1.bar(a, perc[a], width=0.8, color='#8E0145', edgecolor='k')
  x.append(a)
  samples[a] = samples[a].replace('-W06_kneaddata', '')

plt.xticks(x, samples, rotation=90, fontsize=8)
plt.xlim([-0.7, x[-1]+0.5])
plt.ylim([0, 100])
plt.ylabel('Reads remaining (%)')
plt.tight_layout()
plt.show()
```

**W12 percent reads kept:**
```{python, w12_kneaddata, results='hide', fig.keep='all', cache=TRUE, echo=FALSE, eval=FALSE}
reads = pd.read_csv('/Users/robynwright/Documents/OneDrive/Langille Lab postdoc/fungal_metagenome/W12_kneaddata_read_counts.txt', header=0, index_col=0, sep='\t')
perc = []
samples = list(reads.index.values)
for s in samples:
  perc.append((reads.loc[s, 'final single']/reads.loc[s, 'raw single'])*100)

plt.figure(figsize=(10,4))
ax1 = plt.subplot(111)
x = []
for a in range(len(perc)):
  ax1.bar(a, perc[a], width=0.8, color='#8E0145', edgecolor='k')
  x.append(a)
  samples[a] = samples[a].replace('-W12_kneaddata', '')

plt.xticks(x, samples, rotation=90, fontsize=8)
plt.xlim([-0.7, x[-1]+0.5])
plt.ylim([0, 100])
plt.ylabel('Reads remaining (%)')
plt.tight_layout()
plt.show()
```

### Kraken2 and Bracken

**Run with RefSeq Complete v93:**
```{bash, eval=FALSE}
sudo mount -t ramfs none /scratch/ramdisk/
sudo cp -a /home/shared/Kraken2.0.8_Bracken150mer_RefSeqCompleteV93/ /scratch/ramdisk/
mkdir W0_kraken2_out
mkdir W0_kraken2_kreport
parallel -j 1 'kraken2 --use-names --threads 12 --db /scratch/ramdisk/Kraken2.0.8_Bracken150mer_RefSeqCompleteV93/ --memory-mapping {} --output W0_kraken2_out/{/.}.kraken.txt --report W0_kraken2_kreport/{/.}.kreport --confidence 0.1' ::: W0_decontaminated/*.fastq

parallel -j 1 'bracken -d /scratch/ramdisk/Kraken2.0.8_Bracken150mer_RefSeqCompleteV93/ -i {} -l S -o {.}.bracken -r 150' ::: W0_kraken2_kreport/*.kreport
```

**Relevant files:**
- 
- 

```{python, get_kraken_bracken, results='hide', fig.keep='all', echo=FALSE}
kraken_columns = {0:'Percent fragments clade', 1:'Number fragments clade', 2:'Number fragments taxon', 3:'Level', 4:'NCBI ID', 5:'Taxon name'}
def get_kraken_bracken(fol):
    all_domains = {}
    bracken, kraken, bracken_kreport = [], [], []
    bracken_pd, kraken_pd = [], []
    for fi in sorted(os.listdir(fol)):
        if fi[-7:] == 'bracken':
            bracken.append(fi)
        elif fi[-7:] == 'kreport' and 'bracken' not in fi:
            kraken.append(fi)
        elif fi[-7:] == 'kreport':
            bracken_kreport.append(fi)
    for bk in bracken_kreport:
        with open(fol+'/'+bk, 'rU') as f:
            bk = []
            domains = {}
            this_domain, domain_name = [], ''
            for row in csv.reader(f, delimiter='\t'):
                bk.append(row)
                row[5] = row[5].lstrip()
                if row[3] == 'D':
                    if domain_name != '':
                        domains[domain_name] = this_domain
                    this_domain, domain_name = [], row[5]
                else:
                    if row[3] != 'R' and row[3] != 'U' and 'D' not in row[3]:
                        this_domain.append(row[5])
            domains[domain_name] = this_domain
            for domain in domains:
                if domain in all_domains:
                    all_domains[domain] = list(set(all_domains[domain]+domains[domain]))
                else:
                    all_domains[domain] = list(set(domains[domain]))
    for b in bracken:
        sample = pd.read_csv(fol+'/'+b, sep='\t', header=0, index_col=0)
        b = b.replace('_kneaddata', '')
        sample.drop(['taxonomy_id', 'taxonomy_lvl', 'kraken_assigned_reads', 'added_reads', 'fraction_total_reads'], axis=1, inplace=True)
        sample.rename(columns={'new_est_reads':b[:-8]}, inplace=True)
        bracken_pd.append(sample)
    for k in kraken:
        sample = pd.read_csv(fol+'/'+k, sep='\t', header=None, index_col=3)
        sample = sample.loc[['U', 'D'], :]
        sample = sample.rename(columns=kraken_columns).drop(['Number fragments taxon', 'NCBI ID'], axis=1).rename(columns={'Percent fragments clade':k[:-8]+'_percent', 'Number fragments clade':k[:-8]+'_reads'}).set_index('Taxon name')
        taxa = list(sample.index.values)
        taxa_dict = {}
        for t in taxa:
            taxa_dict[t] = t.replace(' ', '')
        sample = sample.rename(index=taxa_dict)
        kraken_pd.append(sample)
    bracken = pd.concat(bracken_pd, join='outer')
    kraken = pd.concat(kraken_pd, join='outer')
    kraken = kraken.rename(index={'d__Bacteria':'Bacteria', 'd__Archaea':'Archaea'})
    kraken = kraken.groupby(by=kraken.index, axis=0).sum()
    bracken = bracken.groupby(by=bracken.index, axis=0).sum().fillna(value=0)
    kraken_samples = list(kraken.columns)
    reads, percent, rename = [], [], {}
    for s in kraken_samples:
      if 'reads' in s:
        reads.append(True)
        percent.append(False)
        rename[s] = s.replace('_kneaddata_reads', '')
      else:
        reads.append(False)
        percent.append(True)
        rename[s] = s.replace('_kneaddata_percent', '')
    kraken_reads = kraken.loc[:, reads].rename(columns=rename)
    kraken_percent = kraken.loc[:, percent].rename(columns=rename)
    return bracken, kraken, bracken_kreport, kraken_reads, kraken_percent
```

```{python, tax_plot, echo=FALSE}
def plot_domain(kraken_reads, kraken_percent):
  plotting = ['Eukaryota', 'Bacteria', 'Archaea', 'Viruses', 'unclassified']
  colors = {'Eukaryota':'#7DCEA0', 'Bacteria':'#5499C7', 'Archaea':'#EDBB99', 'Viruses':'#F7DC6F', 'unclassified':'#CCD1D1'}
  plt.figure(figsize=(8,6))
  ax1 = plt.subplot(211)
  ax2 = plt.subplot(212)
  x = 0
  handles, all_x = [], []
  for s in kraken_reads.columns:
    c_reads, c_perc = 0, 0
    for p in plotting:
      reads, perc = kraken_reads.loc[p, s], kraken_percent.loc[p, s]
      ax1.bar(x, reads, bottom=c_reads, color=colors[p], edgecolor='k', width=0.8)
      ax2.bar(x, perc, bottom=c_perc, color=colors[p], edgecolor='k', width=0.8)
      c_reads += reads
      c_perc += perc
    all_x.append(x)
    x += 1
  for p in plotting:
    handles.append(Patch(facecolor=colors[p], edgecolor='k', label=p))
  ax1.set_xlim([-0.7, all_x[-1]]+0.5), ax2.set_xlim([-0.7, all_x[-1]]+0.5)
  ax2.set_ylim([0, 100])
  plt.sca(ax2)
  plt.xticks(x, list(kraken_reads.columns), rotation=90)
  plt.sca(ax1)
  plt.semilogy()
  plt.xticks([])
  ax1.legend(handles=handles, bbox_to_anchor=(1,1.02), loc='upper left')
  ax1.set_ylabel('Number of reads (log scale)')
  ax2.set_ylabel('Number of reads (%)')
  return
```

**W0 domain level:**
```{python, W0_domain, results='hide', fig.keep='all', cache=TRUE, echo=FALSE}
fol = '/Users/robynwright/Documents/OneDrive/Langille Lab postdoc/fungal_metagenome/W0_kraken2_kreport/'
bracken, kraken, bracken_kreport, kraken_reads, kraken_percent = get_kraken_bracken(fol)

plot_domain(kraken_reads, kraken_percent)
plt.show()
```

**W06 domain level:**
```{python}

```

**W12 domain level:**
```{python}

```

### Get fungal reads

```{bash, eval=FALSE}

```

### HUMAnN3

**Run using Uniref90 and Uniref50 databases:**
```{bash, eval=FALSE}

```

**Combine output files:**
```{bash, eval=FALSE}

```

**Renormalise output files:**
```{bash, eval=FALSE}

```

**Regroup output to other functional categories:**
```{bash, eval=FALSE}

```

## Taxonomic analysis {.tabset}

### Number of reads (domain level)

### Percentage of reads (domain level)

### NMDS plots (all taxa)

### Stacked bar plots (all taxa)

### NMDS plots (fungi only)

### Stacked bar plots (fungi only)

### Differential abundance
Do I want to do this? Between sample sets on different weeks? Between different sample sets within the same week? Probably still not enough samples for machine learning...

## Functional analysis {.tabset}

### NMDS plots

### Gene family richness

### Pathway richness

### Any pathways of interest?

### Differential abundance?