---
title: "Repeat Kraken confidence parameter testing"
output:
  html_document:
    toc: yes
    toc_float: yes
    code_folding: hide
  pdf_document:
    toc: yes
---

```{R, results='hide', fig.keep='all', message=FALSE}
library(reticulate)
#library(kableExtra)
library(knitr)
library(phyloseq)
library(microbiome)
```

```{python, results='hide', fig.keep='all', message=FALSE}
import os
import pandas as pd
import pickle
import matplotlib.pyplot as plt
import numpy as np
import matplotlib as mpl
from matplotlib.patches import Patch
from scipy.spatial import distance
from skbio.stats import ordination
from datetime import datetime
import scipy.spatial.distance as ssd
from scipy.cluster import hierarchy
from skbio.stats.composition import clr
from matplotlib.lines import Line2D

folder = '/Users/robynwright/Documents/OneDrive/Langille_Lab_postdoc/kraken_confidence_testing/all_confidence_together_kraken2/'
folder_krep = folder+'kraken2_kreport/all_together/'
folder_pickle = folder+'pickle/'
folder_pcoa = folder+'distance_dfs_and_pcoa/'
folder_blood = folder+'pickle_blood/'

def popen(name, pcoa=None, blood=None):
  if blood != None:
    name = folder_blood+name
  else:
    if pcoa != None: name = folder_pcoa+name
    else: name = folder_pickle+name
  with open(name, 'rb') as f:
    obj = pickle.load(f)
  return obj

def pdump(obj, name, blood=None):
  if blood != None:
    name = folder_blood+name
  else:
    name = folder_pickle+name
  with open(name, 'wb') as f:
    pickle.dump(obj, f)
  return

def remove_masked(df_in):
  keeping = []
  for col in df_in.columns:
    if 'masked' not in col:
      keeping.append(col)
  if list(df_in.index.values) == list(df_in.columns): return pd.DataFrame(df_in.loc[keeping, keeping])
  return pd.DataFrame(df_in.loc[:, keeping])
  
def remove_covid_others(df):
  removing = ['cDNA-N2-neg_bracken', 'cDNA-O2-neg', 'O4-pos_bracken', 'cDNA-N2-neg', 'O3-pos_bracken', 'cDNA-N5-neg_bracken', 'cDNA-O1-neg_bracken', 'cDNA-N4-pos', 'cDNA-O4-pos', 'N2-neg', 'O2-neg_bracken', 'cDNA-N4-pos_bracken', 'O2-neg', 'cDNA-N3-pos_bracken', 'O5-neg_bracken', 'O4-pos', 'N1-neg_bracken', 'N4-pos', 'N3-pos_bracken', 'cDNA-O5-neg_bracken', 'N2-pos', 'cDNA-O2-neg_bracken', 'O2-pos', 'N4-pos_bracken', 'O4-neg', 'cDNA-N1-neg_bracken', 'N4-neg', 'cDNA-O3-pos_bracken', 'N5-neg_bracken', 'cDNA-O2-pos', 'N2-neg_bracken', 'cDNA-N2-pos', 'cDNA-O4-pos_bracken', 'cDNA-N4-neg', 'cDNA-O4-neg', 'O1-neg_bracken', 'cDNA-O1-pos_bracken', 'N5-neg', 'O5-neg', 'cDNA-O1-pos', 'cDNA-N2-pos_bracken', 'O4-neg_bracken', 'O3-pos', 'N3-pos', 'O3-neg_bracken', 'cDNA-N1-pos', 'cDNA-N5-pos_bracken', 'N1-pos_bracken', 'cDNA-O5-neg', 'cDNA-N5-neg', 'N1-pos', 'O2-pos_bracken', 'cDNA-N3-pos', 'cDNA-N4-neg_bracken', 'cDNA-O3-pos', 'cDNA-N3-neg_bracken', 'O1-pos', 'O5-pos_bracken', 'cDNA-O5-pos', 'cDNA-N1-pos_bracken', 'cDNA-N5-pos', 'N1-neg', 'N3-neg_bracken', 'cDNA-O5-pos_bracken', 'cDNA-N3-neg', 'cDNA-O2-pos_bracken', 'N4-neg_bracken', 'cDNA-O3-neg', 'O1-neg', 'N5-pos', 'O5-pos', 'O1-pos_bracken', 'cDNA-O3-neg_bracken', 'cDNA-O1-neg', 'N5-pos_bracken', 'O3-neg', 'N3-neg', 'N2-pos_bracken', 'cDNA-O4-neg_bracken', 'cDNA-N1-neg']
  try:
    df.drop(removing, axis=1, inplace=True)
    try: 
      df.drop(removing, axis=0, inplace=True)
    except: 
      return df
    return df
  except:
    return df

colormap = mpl.cm.get_cmap('winter', 256)
norm = mpl.colors.Normalize(vmin=0, vmax=21)
m1 = mpl.cm.ScalarMappable(norm=norm, cmap=colormap)
colors_conf = [m1.to_rgba(a) for a in range(21)]
  
sample_names = ['ZymoMock', 'PGPC Blood', 'Human samples','Quarry soils', 'COVID-19 swabs']
names_in_files = ['ZymoMock', 'PGPC', 'HumanSample', 'QuarrySoils', 'COVID']
colors = ['gray', 'maroon', 'royalblue', 'forestgreen', 'gold']
colors_samples = {'ZymoMock':'gray', 'PGPC':'maroon', 'HumanSample':'royalblue', 'QuarrySoils':'forestgreen', 'COVID':'gold'}
databases = ['gtdbh', 'refseq']
colors_databases = {'gtdbh':'orange', 'refseq':'teal'}
confidence = [0.00, 0.05, 0.10, 0.15, 0.20, 0.25, 0.30, 0.35, 0.40, 0.45, 0.50, 0.55, 0.60, 0.65, 0.70, 0.75, 0.80, 0.85, 0.90, 0.95, 1.00]
colors_conf_dict = {}
for c in range(len(confidence)):
  colors_conf_dict[confidence[c]] = colors_conf[c]

colors_domain = ['hotpink', 'darkviolet', 'teal', 'midnightblue']
```

# Processing {.tabset}

## Make GTDB database

## Get samples {.tabset}

### 

### ZymoMock

### PGPC

### Human samples

### Quarry soils

### COVID-19 samples

## Pre-processing {.tabset}

### 

### Kneaddata

### Dustmasker

## Kraken2 runs

## Additional Kraken2 to run 04 May 2021 {.tabset}

### COVID samples

RefSeq Complete V93
```{bash, eval=FALSE}
parallel -j 1 'kraken2 --use-names --threads 20 --db /scratch/ramdisk/Kraken2.0.8_Bracken150mer_RefSeqCompleteV93/ --memory-mapping {1} --output kraken2_outraw/{1/.}_refseq_{2}.kraken.txt --report kraken2_kreport/{1/.}_refseq_{2}.kreport --confidence {2}' ::: cat_reads/*.fastq ::: 0.05 0.15 0.25 0.35 0.45 0.55 0.65 0.75 0.85 0.95
```

Bracken
```{bash, eval=FALSE}
parallel -j 1 'bracken -d /scratch/ramdisk/Kraken2.0.8_Bracken150mer_RefSeqCompleteV93/ -i {} -l S -o {.}.bracken -r 150' ::: kraken2_kreport/*refseq*.kreport
```

GTDB + human
```{bash, eval=FALSE}
parallel -j 1 'kraken2 --use-names --threads 20 --db /scratch/ramdisk/Kraken2_GTDB_human_Dec2020/ --memory-mapping {1} --output kraken2_outraw/{1/.}_gtdbh_{2}.kraken.txt --report kraken2_kreport/{1/.}_gtdbh_{2}.kreport --confidence {2}' ::: cat_reads/*.fastq ::: 0 0.05 0.10 0.15 0.20 0.25 0.30 0.35 0.40 0.45 0.50 0.55 0.60 0.65 0.70 0.75 0.80 0.85 0.90 0.95 1.00
```

Bracken
```{bash, eval=FALSE}
parallel -j 1 'bracken -d /scratch/ramdisk/Kraken2_GTDB_human_Dec2020/ -i {} -l S -o {.}.bracken -r 150' ::: kraken2_kreport/*gtdbh*.kreport
```

### Blood samples (GTDB no human)

Kraken:
```{bash, eval=FALSE}
parallel -j 1 'kraken2 --use-names --threads 20 --db /scratch/ramdisk/Kraken2_GTDB_no_human_May2021/ --memory-mapping {1} --output kraken2_outraw/{1/.}_gtdb-nohuman_{2}.kraken.txt --report kraken2_kreport/{1/.}_gtdb-nohuman_{2}.kreport --confidence {2}' ::: joined_lanes/*.fastq.gz ::: 0.00 0.05 0.10 0.15 0.20 0.25 0.30 0.35 0.40 0.45 0.50 0.55 0.60 0.65 0.70 0.75 0.80 0.85 0.90 0.95 1.00
```

Bracken
```{bash, eval=FALSE}
parallel -j 1 'bracken -d /scratch/ramdisk/Kraken2_GTDB_no_human_May2021/ -i {} -l S -o {.}.bracken -r 150' ::: kraken2_kreport/*gtdb-nohuman*.kreport
```


# Results {.tabset}

## Pre-processing and importing

Pre-processing between when I processed these reads on the server and importing them here - changes all file names to match the file format that I wanted them (and also re-naming some of the files from other projects so they aren't recognisable) in so that all files could be in the same folder:
```{python, eval=FALSE}
import os 

folder = '/Users/robynwright/Documents/OneDrive/Langille_Lab_postdoc/kraken_confidence_testing/all_confidence_together_kraken2/kraken2_kreport/refseq/'
files = sorted(os.listdir(folder))
files = [f for f in files if f != '.DS_Store']
"""
#MirallesMetaG
files = [f for f in files if 'MirallesMetaG' in f]
original_files = list(files)
rename_sample = {'N2O':'1', 'P15O':'2', 'P4O':'3', 'P9O':'4'}
for f in range(len(files)):
  new = files[f].split('_')[1].split('.', 1)
  new = 'QuarrySoils'+rename_sample[new[0]]+'_refseq_'+new[1]
  files[f] = new

#AsheMetaG
files = [f for f in files if 'AsheMetaG' in f]
original_files = list(files)
files = [f.split('.', 2) for f in files]
files = [[f[0].split('_'), f[2]] for f in files]
rename_sample = {'S1E1_S1':'1', 'S1E3_S3':'2', 'S2E3_S10':'3', 'S3E5_S17':'4'}
files = [[rename_sample[f[0][1]+'_'+f[0][2]], f[1]] for f in files]
files = ['HumanSample'+f[0]+'_refseq_'+f[1] for f in files]

#ZymoMock
files = [f for f in files if 'ZymoMock' in f]
original_files = list(files)
files = [f.split('.', 1) for f in files]
files = [f[0]+'_refseq_'+f[1] for f in files]
"""

# for a in range(len(files)):
#   cmd = 'mv '+folder+original_files[a]+' '+folder+files[a]
#   os.system(cmd)
```

Sort blood samples:
```{python, eval=FALSE}
import os 

folder = '/Users/robynwright/Documents/OneDrive/Langille_Lab_postdoc/kraken_confidence_testing/all_confidence_together_kraken2/kraken2_kreport/blood_mg/'
files = sorted(os.listdir(folder))
files = [f for f in files if f != '.DS_Store']
original_files = list(files)

files = ['PGPC'+f[7:] for f in files]
files = [f.replace('gtdb', 'gtdbh') for f in files]
for a in range(len(files)):
  f = files[a]
  if 'masked' in f:
    f = f.replace('masked_removed_', '').split('_', 2)
    f = f[0]+'_'+f[1]+'-masked_'+f[2]
    files[a] = f

for a in range(len(files)):
  cmd = 'mv '+folder+original_files[a]+' '+folder+files[a]
  os.system(cmd)
```

Sort COVID samples:
```{python, eval=FALSE}
for f in files:
  try:
    if isinstance(float(f.split('.')[0][-1]), float): keep_going = True
  except: 
    continue
  new_name = f.split('_')
  new_new_name = 'COVID-'+new_name[0]+'_refseq_'+new_name[1]
  if len(new_name) > 2: new_new_name += '_'+new_name[2]
  os.system('mv kraken2_kreport_RSC/'+f+' kraken2_kreport_RSC/'+new_new_name)
```

Format and read in results to dataframes (on server):
```{python, eval=FALSE}
import os
import pandas as pd
import pickle

folder = os.getcwd()
folder_krep = folder+'/all_together/'
folder_pickle = folder+'/pickle/'

#function to open pickle file
def popen(name):
  name = folder_pickle+name
  with open(name, 'rb') as f:
    obj = pickle.load(f)
  return obj

#function to close pickle file
def pdump(obj, name):
  name = folder_pickle+name
  with open(name, 'wb') as f:
    pickle.dump(obj, f)
  return
  
### GET TAXONOMY HIERARCHIES ###
#get file list
files = os.listdir(folder_krep)
files = [f for f in files if '.kreport' in f and 'bracken' not in f]

#set up dictionaries, including one that will be used to save the most recent higher taxonomic levels
all_taxa = {}
current = {'D':'', 'P':'', 'C':'', 'O':'', 'F':'', 'G':''}

#loop through all of the files saving the full taxonomy each time you reach a species
for f in files:
  classifications = pd.read_csv(folder_krep+f, sep='\t').transpose()
  classifications = classifications.reset_index().transpose()
  classifications.columns = ['Percentfragments', 'Number fragments root', 'Number fragments', 'Rank', 'NCBI taxid', 'Scientific name']
  rows = classifications.index.values
  for row in rows:
    rank = classifications.loc[row, 'Rank']
    name = classifications.loc[row, 'Scientific name'].lstrip()
    if rank in current:
      current[rank] = name
    if rank == 'S' or rank == 'S1':
      full_tax = current['D']+';'+current['P']+';'+current['C']+';'+current['O']+';'+current['F']+';'+current['G']+';'+name
      all_taxa[name] = full_tax

#save the file
pdump(all_taxa, 'all_taxa_kraken_full.dict')  
  
#Get tax dict without the s__, g__ etc for GTDB samples - basically make them comparable easily with refseq classifications
for tax in all_taxa:
  if '__' in tax:
    this_tax = all_taxa[tax].split(';')
    new_tax = ''
    for t in range(len(this_tax)):
      this_tax[t] = this_tax[t].split('__')[1]
      new_tax += this_tax[t]
      if t < len(this_tax)-1:
        new_tax += ';'
    new_tax = new_tax.replace('Animalia', 'Eukaryota')
    all_taxa[tax] = new_tax

#save the file
pdump(all_taxa, 'all_taxa_gtdb_kraken_fix.dict')

#Get domain for all taxa
domain_dict = {}
domains = []
count = 0
for tax in all_taxa:
  dom = all_taxa[tax].split(';')[0]
  domain_dict[tax] = dom
  if dom not in domains:
    domains.append(dom)
pdump(domain_dict, 'sp_to_domain_kraken.dict')

### Get number of reads before Kraken ###
#get file list
files = os.listdir(folder_krep)
files = [f for f in files if 'kreport' in f and 'bracken' not in f]

#only keep the files if they are confidence = 0 (we only need to look at one of them to see unclassified and classifications at root)
new_files = []
for f in files:
  if 'neg' in f or 'pos' in f:
    if 'COVID' not in f:
      continue
  num = f.split('_')[2].replace('.kreport', '')
  if num in ['0', '0.00']:
    new_files.append(f)
    
files = new_files

#loop through all the files making dataframes of only the taxa (root and unclassified) that we want, also renaming the columns to match the file name
reads = []
sample_names = []
for f in files:
  fname = f.split('_')[0]
  df = pd.read_csv(folder_krep+f, index_col=5, header=None, sep='\t')
  df = pd.DataFrame(df.loc[['root', 'unclassified'], 1]).rename(columns={1:fname})
  reads.append(df), sample_names.append(fname)

#combine the dataframes, get the sum of reads and save the files
initial_reads = pd.concat(reads).fillna(value=0)
initial_reads = initial_reads.groupby(by=initial_reads.index, axis=0).sum()
reads_sum = pd.DataFrame(initial_reads.sum(axis=0))
reads_sum = reads_sum.transpose().rename(index={0:'Sum'})
pdump(reads_sum, 'initial_reads.df')

### Get all classifications for all confidence intervals ###
#get file list
files = os.listdir(folder_krep)
files = [f for f in files if f[-7:] == 'bracken']

#read each of the files and save these dataframes in a list
all_df = []
count = 0
for f in files:
  fname = f.replace('.bracken', '')
  df = pd.read_csv(folder_krep+f, index_col=0, header=0, sep='\t')
  df = df.rename(columns={'new_est_reads':fname})
  all_df.append(pd.DataFrame(df.loc[:, fname]))
  if count % 100 == 0: print(count, fname)
  count += 1

#go through the list of dataframes, and each time you have 100 dataframes combine them
this_df = []
combined_dfs = []
for a in range(len(all_df)):
  this_df.append(all_df[a])
  if a == 0: continue
  if a % 100 == 0 or a == len(all_df)-1: 
    print(a, all_df[a].columns)
    this_df = pd.concat(this_df).fillna(value=0)
    this_df = this_df.groupby(by=this_df.index, axis=0).sum()
    combined_dfs.append(this_df)
    this_df = []

#now go through those and combine them when there's 10 together, combine them
this_df = []
combined_combined_dfs = []
for a in range(len(combined_dfs)):
  this_df.append(combined_dfs[a])
  print(a)
  if a == 0: continue
  if a % 10 == 0 or a == len(combined_dfs)-1: 
    this_df = pd.concat(this_df).fillna(value=0)
    this_df = this_df.groupby(by=this_df.index, axis=0).sum()
    combined_combined_dfs.append(this_df)
    this_df = []

#now combine this list of 9 dataframes (hopefully)
all_combined_df = pd.concat(combined_combined_dfs)
all_combined_df = all_combined_df.groupby(by=all_combined_df.index, axis=0).sum()

#save it
pdump(all_combined_df, 'all_df.df')
all_combined_df.to_csv('all_combined_df.csv')

#calculate the number of reads in each sample
sum_reads = pd.DataFrame(all_combined_df.sum(axis=0))
sum_reads = sum_reads.transpose().rename(index={0:'Sum'})
pdump(sum_reads, 'sum_reads_classified.df')

#Get the number of species in each sample
presence_absence = pd.DataFrame(all_combined_df)
presence_absence[presence_absence > 1] = 1
sum_species = pd.DataFrame(presence_absence.sum(axis=0))
sum_species = sum_species.transpose().rename(index={0:'Sum'})
pdump(sum_species, 'sum_species_classified.df')
  
### Get tables and sums for each domain ###
#Get the domain and GTDB species name for each taxon
all_combined_df = popen('all_df.df')
all_taxa = popen('all_taxa_gtdb_kraken_fix.dict')
domain_dict = popen('sp_to_domain_kraken.dict')
all_domains = []
rename_gtdb = {}
for val in all_combined_df.index.values:
  all_domains.append(domain_dict[val])
  rename_gtdb[val] = all_taxa[val].split(';')[-1]

#Add a column to the main dataframe
all_combined_df['Domain'] = all_domains
all_domains = sorted(list(set(all_domains)))

#Rename the species based on removing the s__ etc.
all_combined_gtdb_fix_df = pd.DataFrame(all_combined_df.rename(index=rename_gtdb))
all_gtdb_rsc_combined = pd.DataFrame(all_combined_gtdb_fix_df.drop(['Domain'], axis=1))
all_gtdb_rsc_combined = all_gtdb_rsc_combined.groupby(by=all_gtdb_rsc_combined.index, axis=0).sum()

#As we lose the domain by summing the columns, let's add that back
new_domain = []
for val in all_gtdb_rsc_combined.index.values:
  dom = all_combined_gtdb_fix_df.loc[val, 'Domain']
  if isinstance(dom, str): new_domain.append(dom)
  else: 
    new_domain.append(list(dom)[0])
all_gtdb_rsc_combined['Domain'] = new_domain

#And then save the file
pdump(all_gtdb_rsc_combined, 'all_df_gtdb_rsc_combined.df')
all_gtdb_rsc_combined.to_csv('all_gtdb_rsc_combined.csv')

#Save individual dataframes for each domain as well as getting the sums of reads for each domain and saving these
for dom in set(all_domains):
  this_df = pd.DataFrame(all_gtdb_rsc_combined.loc[all_gtdb_rsc_combined['Domain'] == dom]).drop(['Domain'], axis=1)
  pdump(this_df, dom+'_reads.df')
  this_df.to_csv(dom+'_reads.csv')
  pdump(pd.DataFrame(this_df.sum(axis=0)).transpose().rename(index={0:'Sum'}), dom+'_reads_sum.df')
  num_sp = pd.DataFrame(this_df)
  num_sp[num_sp > 1] = 1
  pdump(pd.DataFrame(num_sp.sum(axis=0)).transpose().rename(index={0:'Sum'}), dom+'_species_sum.df')
  
#Filter out taxa with <100 reads total
all_reads = popen('all_df_gtdb_rsc_combined.df')
all_reads_rem100 = all_reads_rem10[all_reads.max(axis=1) > 100]
pdump(all_reads_rem100, 'all_df_rem100.df')
pdump(pd.DataFrame(all_reads_rem100.sum(axis=0)).transpose().rename(index={0:'Sum'}), 'sum_reads_classified_rem100.df')

#Remove some COVID samples that accidentally slipped through into the dataframes and filter taxa with <100 total reads
dfs = ['all_df_rem100.df', 'all_df_gtdb_rsc_combined.df', 'all_df.df', 'Archaea_reads_sum.df', 'Archaea_reads.df', 'Archaea_species_sum.df', 'Bacteria_reads_sum.df', 'Bacteria_reads.df', 'Bacteria_species_sum.df', 'Eukaryota_reads_sum.df', 'Eukaryota_reads.df', 'Eukaryota_species_sum.df', 'initial_reads.df', 'sum_reads_classified.df', 'sum_species_classified.df', 'Viruses_reads_sum.df', 'Viruses_reads.df', 'Viruses_species_sum.df']

removing = ['cDNA-N2-neg_bracken', 'cDNA-O2-neg', 'O4-pos_bracken', 'cDNA-N2-neg', 'O3-pos_bracken', 'cDNA-N5-neg_bracken', 'cDNA-O1-neg_bracken', 'cDNA-N4-pos', 'cDNA-O4-pos', 'N2-neg', 'O2-neg_bracken', 'cDNA-N4-pos_bracken', 'O2-neg', 'cDNA-N3-pos_bracken', 'O5-neg_bracken', 'O4-pos', 'N1-neg_bracken', 'N4-pos', 'N3-pos_bracken', 'cDNA-O5-neg_bracken', 'N2-pos', 'cDNA-O2-neg_bracken', 'O2-pos', 'N4-pos_bracken', 'O4-neg', 'cDNA-N1-neg_bracken', 'N4-neg', 'cDNA-O3-pos_bracken', 'N5-neg_bracken', 'cDNA-O2-pos', 'N2-neg_bracken', 'cDNA-N2-pos', 'cDNA-O4-pos_bracken', 'cDNA-N4-neg', 'cDNA-O4-neg', 'O1-neg_bracken', 'cDNA-O1-pos_bracken', 'N5-neg', 'O5-neg', 'cDNA-O1-pos', 'cDNA-N2-pos_bracken', 'O4-neg_bracken', 'O3-pos', 'N3-pos', 'O3-neg_bracken', 'cDNA-N1-pos', 'cDNA-N5-pos_bracken', 'N1-pos_bracken', 'cDNA-O5-neg', 'cDNA-N5-neg', 'N1-pos', 'O2-pos_bracken', 'cDNA-N3-pos', 'cDNA-N4-neg_bracken', 'cDNA-O3-pos', 'cDNA-N3-neg_bracken', 'O1-pos', 'O5-pos_bracken', 'cDNA-O5-pos', 'cDNA-N1-pos_bracken', 'cDNA-N5-pos', 'N1-neg', 'N3-neg_bracken', 'cDNA-O5-pos_bracken', 'cDNA-N3-neg', 'cDNA-O2-pos_bracken', 'N4-neg_bracken', 'cDNA-O3-neg', 'O1-neg', 'N5-pos', 'O5-pos', 'O1-pos_bracken', 'cDNA-O3-neg_bracken', 'cDNA-O1-neg', 'N5-pos_bracken', 'O3-neg', 'N3-neg', 'N2-pos_bracken', 'cDNA-O4-neg_bracken', 'cDNA-N1-neg']

for df in dfs:
  this_df = popen(df)
  try:
    this_df.drop(removing, axis=1, inplace=True)
    pdump(this_df, df)
  except:
    continue

#Save the number of species
all_reads_rem100 = popen('all_df_rem100.df')
all_species_rem100 = pd.DataFrame(all_reads_rem100).drop(['Domain'], axis=1)
all_species_rem100[all_species_rem100 > 1] = 1
pdump(pd.DataFrame(all_species_rem100.sum(axis=0)).transpose().rename(index={0:'Sum'}), 'sum_species_classified_rem100.df')

#And save individual dataframes for each domain as well as getting the sums of reads for each domain and saving these
all_reads_rem100 = popen('all_df_rem100.df')
all_domains = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
for dom in set(all_domains):
  this_df = pd.DataFrame(all_reads_rem100.loc[all_reads_rem100['Domain'] == dom]).drop(['Domain'], axis=1)
  pdump(this_df, dom+'_reads_rem100.df')
  this_df.to_csv(dom+'_reads_rem100.csv')
  pdump(pd.DataFrame(this_df.sum(axis=0)).transpose().rename(index={0:'Sum'}), dom+'_reads_sum_rem100.df')
  num_sp = pd.DataFrame(this_df)
  num_sp[num_sp > 1] = 1
  pdump(pd.DataFrame(num_sp.sum(axis=0)).transpose().rename(index={0:'Sum'}), dom+'_species_sum_rem100.df')

#And lets get this all >100 reads at genus level (and resave individual dataframes)
all_reads_rem100 = popen('all_df_rem100.df')
all_taxa = popen('all_taxa_gtdb_kraken_fix.dict')
all_genus = {}
genus_domain = {}
count = 0
random_list = []
for tax in all_taxa:
  all_genus[all_taxa[tax].split(';')[-1]] = all_taxa[tax].split(';')[-2]
  all_genus[tax] = all_taxa[tax].split(';')[-2]
  genus_domain[all_taxa[tax].split(';')[-2]] = all_taxa[tax].split(';')[0]
  if count < 100:
    random_list.append([tax, all_taxa[tax].split(';')[-1], all_taxa[tax].split(';')[-2], all_taxa[tax].split(';')[0]])
  count += 1
all_genus_rem100 = all_reads_rem100.rename(index=all_genus)
all_genus_rem100 = all_genus_rem100.groupby(by=all_genus_rem100.index, axis=0).sum()
gen_domain = []
for val in all_genus_rem100.index.values:
  gen_domain.append(genus_domain[val])
all_genus_rem100['Domain'] = gen_domain
pdump(all_genus_rem100, 'all_genus_rem100.df')
num_gen = all_genus_rem100.drop(['Domain'], axis=1)
num_gen[num_gen > 1] = 1
pdump(pd.DataFrame(num_gen.sum(axis=0)).transpose().rename(index={0:'Sum'}), 'all_genus_genera_sum_rem100.df')

all_genus_rem100 = popen('all_genus_rem100.df')
all_domains = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
for dom in set(all_domains):
  this_df = pd.DataFrame(all_genus_rem100.loc[all_genus_rem100['Domain'] == dom]).drop(['Domain'], axis=1)
  pdump(this_df, dom+'_reads_genus_rem100.df')
  pdump(pd.DataFrame(this_df.sum(axis=0)).transpose().rename(index={0:'Sum'}), dom+'_reads_genus_sum_rem100.df')
  num_gen = pd.DataFrame(this_df)
  num_gen[num_gen > 1] = 1
  pdump(pd.DataFrame(num_gen.sum(axis=0)).transpose().rename(index={0:'Sum'}), dom+'_genus_sum_rem100.df')
```

Format and read in results for blood samples only:
```{python, eval=FALSE}
import os
import pandas as pd
import pickle

folder = os.getcwd()
folder_krep = folder+'/kraken2_kreport/'
folder_pickle = folder+'/pickle/'

#function to open pickle file
def popen(name):
  name = folder_pickle+name
  with open(name, 'rb') as f:
    obj = pickle.load(f)
  return obj

#function to close pickle file
def pdump(obj, name):
  name = folder_pickle+name
  with open(name, 'wb') as f:
    pickle.dump(obj, f)
  return

### GET TAXONOMY HIERARCHIES ###
#get file list
files = os.listdir(folder_krep)
files = [f for f in files if '.kreport' in f and 'bracken' not in f]

#set up dictionaries, including one that will be used to save the most recent higher taxonomic levels and taxonomy IDs for each database
all_taxa, all_taxid_rsc, all_taxid_gtdb = {}, {}, {}
current = {'D':'', 'P':'', 'C':'', 'O':'', 'F':'', 'G':''}

#loop through all of the files saving the full taxonomy each time you reach a species
for f in files:
  classifications = pd.read_csv(folder_krep+f, sep='\t').transpose()
  classifications = classifications.reset_index().transpose()
  classifications.columns = ['Percentfragments', 'Number fragments root', 'Number fragments', 'Rank', 'NCBI taxid', 'Scientific name']
  rows = classifications.index.values
  for row in rows:
    taxid = classifications.loc[row, 'NCBI taxid']
    rank = classifications.loc[row, 'Rank']
    name = classifications.loc[row, 'Scientific name'].lstrip()
    if rank in current:
      current[rank] = name
    if rank == 'S' or rank == 'S1':
      full_tax = current['D']+';'+current['P']+';'+current['C']+';'+current['O']+';'+current['F']+';'+current['G']+';'+name
      all_taxa[name] = full_tax
      if 'gtdb' in f: all_taxid_gtdb[name] = taxid
      else: all_taxid_rsc[name] = taxid

pdump(all_taxa, 'all_taxa_with_s__.dict')

#Get tax dict without the s__, g__ etc for GTDB samples - basically make them comparable easily with refseq classifications
for tax in all_taxa:
  if '__' in tax:
    this_tax = all_taxa[tax].split(';')
    new_tax = ''
    for t in range(len(this_tax)):
      this_tax[t] = this_tax[t].split('__')[1]
      new_tax += this_tax[t]
      if t < len(this_tax)-1:
        new_tax += ';'
    new_tax = new_tax.replace('Animalia', 'Eukaryota')
    all_taxa[tax] = new_tax

#save the file
pdump(all_taxa, 'all_taxa.dict')

#Get domain for all taxa
domain_dict = {}
domains = []
count = 0
for tax in all_taxa:
  dom = all_taxa[tax].split(';')[0]
  domain_dict[tax] = dom
  if dom not in domains:
    domains.append(dom)
pdump(domain_dict, 'sp_to_domain_kraken.dict')
pdump(all_taxid_rsc, 'sp_taxid_rsc.dict')
pdump(all_taxid_gtdb, 'sp_taxid_gtdb.dict')

### Get all classifications for all confidence intervals ###
#get file list
files = os.listdir(folder_krep)
files = sorted([f for f in files if f[-7:] == 'bracken'])

#read each of the files and save these dataframes in a list
all_df = []
count = 0
for f in files:
  fname = f.replace('.bracken', '').replace('masked_removed_', 'masked-')
  df = pd.read_csv(folder_krep+f, index_col=0, header=0, sep='\t')
  df = df.rename(columns={'new_est_reads':fname})
  all_df.append(pd.DataFrame(df.loc[:, fname]))
  if count % 100 == 0: print(count, fname)
  count += 1

#go through the list of dataframes, and each time you have 100 dataframes combine them
this_df = []
combined_dfs = []
for a in range(len(all_df)):
  this_df.append(all_df[a])
  if a == 0: continue
  if a % 100 == 0 or a == len(all_df)-1: 
    print(a, all_df[a].columns)
    this_df = pd.concat(this_df).fillna(value=0)
    this_df = this_df.groupby(by=this_df.index, axis=0).sum()
    combined_dfs.append(this_df)
    this_df = []

#now go through those and combine them when there's 10 together, combine them
this_df = []
combined_combined_dfs = []
for a in range(len(combined_dfs)):
  this_df.append(combined_dfs[a])
  print(a)
  if a == 0: continue
  if a % 10 == 0 or a == len(combined_dfs)-1: 
    this_df = pd.concat(this_df).fillna(value=0)
    this_df = this_df.groupby(by=this_df.index, axis=0).sum()
    combined_combined_dfs.append(this_df)
    this_df = []

#now combine this list of 9 dataframes (hopefully)
all_combined_df = pd.concat(combined_combined_dfs)
all_combined_df = all_combined_df.groupby(by=all_combined_df.index, axis=0).sum()

#save it
pdump(all_combined_df, 'all_PGPC_blood.df')

#calculate the number of reads in each sample
sum_reads = pd.DataFrame(all_combined_df.sum(axis=0))
sum_reads = sum_reads.transpose().rename(index={0:'Sum'})
pdump(sum_reads, 'sum_reads_classified.df')

#Get the number of species in each sample
presence_absence = pd.DataFrame(all_combined_df)
presence_absence[presence_absence > 1] = 1
sum_species = pd.DataFrame(presence_absence.sum(axis=0))
sum_species = sum_species.transpose().rename(index={0:'Sum'})
pdump(sum_species, 'sum_species_classified.df')
  
### Get tables and sums for each domain ###
#Get the domain and GTDB species name for each taxon
all_combined_df = popen('all_PGPC_blood.df')
all_taxa = popen('all_taxa.dict')
domain_dict = popen('sp_to_domain_kraken.dict')
all_domains = []
rename_gtdb = {}
for val in all_combined_df.index.values:
  all_domains.append(domain_dict[val])
  rename_gtdb[val] = all_taxa[val].split(';')[-1]

#Add a column to the main dataframe
all_combined_df['Domain'] = all_domains
all_domains = sorted(list(set(all_domains)))

#Rename the species based on removing the s__ etc.
all_combined_gtdb_fix_df = pd.DataFrame(all_combined_df.rename(index=rename_gtdb))
all_gtdb_rsc_combined = pd.DataFrame(all_combined_gtdb_fix_df.drop(['Domain'], axis=1))
all_gtdb_rsc_combined = all_gtdb_rsc_combined.groupby(by=all_gtdb_rsc_combined.index, axis=0).sum()

#As we lose the domain by summing the columns, let's add that back
new_domain = []
for val in all_gtdb_rsc_combined.index.values:
  dom = all_combined_gtdb_fix_df.loc[val, 'Domain']
  if isinstance(dom, str): new_domain.append(dom)
  else: 
    new_domain.append(list(dom)[0])
all_gtdb_rsc_combined['Domain'] = new_domain

#And then save the file
pdump(all_gtdb_rsc_combined, 'all_df_blood_combined.df')

#Save individual dataframes for each domain as well as getting the sums of reads for each domain and saving these
for dom in set(all_domains):
  this_df = pd.DataFrame(all_gtdb_rsc_combined.loc[all_gtdb_rsc_combined['Domain'] == dom]).drop(['Domain'], axis=1)
  pdump(this_df, dom+'_reads.df')
  this_df.to_csv(dom+'_reads.csv')
  pdump(pd.DataFrame(this_df.sum(axis=0)).transpose().rename(index={0:'Sum'}), dom+'_reads_sum.df')
  num_sp = pd.DataFrame(this_df)
  num_sp[num_sp > 1] = 1
  pdump(pd.DataFrame(num_sp.sum(axis=0)).transpose().rename(index={0:'Sum'}), dom+'_species_sum.df')
```

## Number of reads {.tabset}

For all of the plots in this section, one is shown with all species (49,479 total species), and another where only species with at least 100 reads (27,583 species) in one sample are kept. If I had kept only those with at least 10 reads, there were 44,463 species remaining.

### Reads after kneaddata

```{python, results='hide', fig.keep='all', cache=TRUE}
numbers = [[], [], [], [], []]
x = [1, 2, 3, 4, 5]

initial_reads = popen('initial_reads.df')

plt.figure(figsize=(8,5))
ax1 = plt.subplot(111)

for a in range(len(names_in_files)):
  for col in initial_reads.columns:
    if names_in_files[a] in col:
      numbers[a].append(initial_reads.loc['Sum', col])

for a in range(len(names_in_files)):
  pltx = np.random.normal(x[a], scale=0.05, size=len(numbers[a]))
  plt.scatter(pltx, numbers[a], color=colors[a], alpha=0.2)
  
ax1.boxplot(numbers[1:], positions=x[1:], showfliers=False, medianprops=dict(color='k'))
plt.xticks(x, sample_names)
plt.ylabel('Number of reads')
plt.semilogy()
plt.show()
#plt.savefig(folder+'/figures/Initial reads.png', dpi=600, bbox_inches='tight')
```

### Reads classified {.tabset}

#### Total reads classified

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified.df')
reads_classified = remove_masked(reads_classified)

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = []
    colors_plot = []
    for col in reads_classified.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf.append(reads_classified.loc['Sum', col])
        for sname in colors_samples:
          if sname in col: 
            colors_plot.append(colors_samples[sname])
            break
    #axes[d].boxplot([this_conf], positions=[c], showfliers=False, medianprops=dict(color='k'))
    pltx = np.random.normal(c, scale=0.05, size=len(this_conf))
    axes[d].scatter(pltx, this_conf, color=colors_plot, s=5, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**2, 10**8))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of reads classified'), plt.xlabel('Confidence threshold')

handles = [Patch(facecolor=colors_samples[sname], edgecolor='k', label=sname) for sname in colors_samples]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of reads classified.png', dpi=600, bbox_inches='tight')
```

#### Total reads classified >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified_rem100.df')
reads_classified = remove_masked(reads_classified)
reads_classified = remove_covid_others(reads_classified.drop(['Domain'], axis=1))

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = []
    colors_plot = []
    for col in reads_classified.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf.append(reads_classified.loc['Sum', col])
        for sname in colors_samples:
          if sname in col: 
            colors_plot.append(colors_samples[sname])
            break
    #axes[d].boxplot([this_conf], positions=[c], showfliers=False, medianprops=dict(color='k'))
    pltx = np.random.normal(c, scale=0.05, size=len(this_conf))
    axes[d].scatter(pltx, this_conf, color=colors_plot, s=5, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**2, 10**8))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of reads classified'), plt.xlabel('Confidence threshold')

handles = [Patch(facecolor=colors_samples[sname], edgecolor='k', label=sname) for sname in colors_samples]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of reads classified above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Proportion reads classified

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified.df')
reads_classified = remove_masked(reads_classified)
initial_reads = popen('initial_reads.df')

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = []
    colors_plot = []
    for col in reads_classified.columns:
      if db in col and float(col.split('_')[2]) == conf:
        reads = reads_classified.loc['Sum', col]
        initial = initial_reads.loc['Sum', col.split('_')[0]]
        this_conf.append((reads/initial)*100)
        for sname in colors_samples:
          if sname in col: 
            colors_plot.append(colors_samples[sname])
            break
    #axes[d].boxplot([this_conf], positions=[c], showfliers=False, medianprops=dict(color='k'))
    pltx = np.random.normal(c, scale=0.05, size=len(this_conf))
    axes[d].scatter(pltx, this_conf, color=colors_plot, s=5, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Proportion of reads classified (%)'), plt.xlabel('Confidence threshold')
  plt.ylim([-2, 90])

handles = [Patch(facecolor=colors_samples[sname], edgecolor='k', label=sname) for sname in colors_samples]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Proportion of reads classified.png', dpi=600, bbox_inches='tight')
```

#### Proportion reads classified >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified_rem100.df')
reads_classified = remove_masked(reads_classified)
reads_classified = remove_covid_others(reads_classified.drop(['Domain'], axis=1))
initial_reads = popen('initial_reads.df')

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = []
    colors_plot = []
    for col in reads_classified.columns:
      if db in col and float(col.split('_')[2]) == conf:
        reads = reads_classified.loc['Sum', col]
        initial = initial_reads.loc['Sum', col.split('_')[0]]
        this_conf.append((reads/initial)*100)
        for sname in colors_samples:
          if sname in col: 
            colors_plot.append(colors_samples[sname])
            break
    #axes[d].boxplot([this_conf], positions=[c], showfliers=False, medianprops=dict(color='k'))
    pltx = np.random.normal(c, scale=0.05, size=len(this_conf))
    axes[d].scatter(pltx, this_conf, color=colors_plot, s=5, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Proportion of reads classified (%)'), plt.xlabel('Confidence threshold')
  plt.ylim([-2, 90])

handles = [Patch(facecolor=colors_samples[sname], edgecolor='k', label=sname) for sname in colors_samples]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Proportion of reads classified above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Number of reads initially vs reads classified

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified.df')
reads_classified = remove_masked(reads_classified)
initial_reads = popen('initial_reads.df')

x, y, colors_plot_conf, colors_plot_dataset = [], [], [], []
for col in reads_classified.columns:
  x.append(reads_classified.loc['Sum', col])
  y.append(initial_reads.loc['Sum', col.split('_')[0]])
  for sname in colors_samples:
    if sname in col:
      colors_plot_dataset.append(colors_samples[sname])
      break
  colors_plot_conf.append(colors_conf_dict[float(col.split('_')[-1])])

plt.figure(figsize=(20, 10))
axes = [plt.subplot(121), plt.subplot(122)]
axes[0].scatter(x, y, color=colors_plot_conf, alpha=0.5)
axes[1].scatter(x, y, color=colors_plot_dataset, alpha=0.5)

color_sets = [colors_conf_dict, colors_samples]

for a in range(len(axes)):
  plt.sca(axes[a])
  plt.semilogy(), plt.semilogx()
  plt.ylabel('Number of reads classified'), plt.xlabel('Number of reads initially')
  handles = [Patch(facecolor=color_sets[a][name], edgecolor='k', label=name) for name in color_sets[a]]
  plt.legend(handles=handles, loc='lower right')

axes[0].set_title('Coloured by confidence'), axes[1].set_title('Coloured by sample type')
plt.show()
#plt.savefig(folder+'/figures/Reads classified vs initial reads.png', dpi=600, bbox_inches='tight')
```

#### Number of reads initially vs reads classified >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified_rem100.df')
reads_classified = remove_masked(reads_classified)
reads_classified = remove_covid_others(reads_classified.drop(['Domain'], axis=1))
initial_reads = popen('initial_reads.df')

x, y, colors_plot_conf, colors_plot_dataset = [], [], [], []
for col in reads_classified.columns:
  x.append(reads_classified.loc['Sum', col])
  y.append(initial_reads.loc['Sum', col.split('_')[0]])
  for sname in colors_samples:
    if sname in col:
      colors_plot_dataset.append(colors_samples[sname])
      break
  colors_plot_conf.append(colors_conf_dict[float(col.split('_')[-1])])

plt.figure(figsize=(20, 10))
axes = [plt.subplot(121), plt.subplot(122)]
axes[0].scatter(x, y, color=colors_plot_conf, alpha=0.5)
axes[1].scatter(x, y, color=colors_plot_dataset, alpha=0.5)

color_sets = [colors_conf_dict, colors_samples]

for a in range(len(axes)):
  plt.sca(axes[a])
  plt.semilogy(), plt.semilogx()
  plt.ylabel('Number of reads classified'), plt.xlabel('Number of reads initially')
  handles = [Patch(facecolor=color_sets[a][name], edgecolor='k', label=name) for name in color_sets[a]]
  plt.legend(handles=handles, loc='lower right')

axes[0].set_title('Coloured by confidence'), axes[1].set_title('Coloured by sample type')
plt.show()
#plt.savefig(folder+'/figures/Reads classified vs initial reads above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Reads classified split by domain

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_reads = popen('Archaea_reads_sum.df')
bacteria_reads = popen('Bacteria_reads_sum.df')
eukaryota_reads = popen('Eukaryota_reads_sum.df')
viruses_reads = popen('Viruses_reads_sum.df')

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = [[], [], [], []]
    colors_plot = []
    for col in bacteria_reads.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf[0].append(archaea_reads.loc['Sum', col])
        this_conf[1].append(bacteria_reads.loc['Sum', col])
        this_conf[2].append(eukaryota_reads.loc['Sum', col])
        this_conf[3].append(viruses_reads.loc['Sum', col])
    allx = [c-0.3, c-0.1, c+0.1, c+0.3]
    for e in range(len(this_conf)):
      pltx = np.random.normal(allx[e], scale=0.05, size=len(this_conf[e]))
      axes[d].scatter(pltx, this_conf[e], color=colors_domain[e], s=2, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**2, 10**8))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of reads classified'), plt.xlabel('Confidence threshold')

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of reads classified split by domain.png', dpi=600, bbox_inches='tight')
```

#### Reads classified split by domain >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_reads = remove_covid_others(popen('Archaea_reads_sum_rem100.df'))
bacteria_reads = remove_covid_others(popen('Bacteria_reads_sum_rem100.df'))
eukaryota_reads = remove_covid_others(popen('Eukaryota_reads_sum_rem100.df'))
viruses_reads = remove_covid_others(popen('Viruses_reads_sum_rem100.df'))

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = [[], [], [], []]
    colors_plot = []
    for col in bacteria_reads.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf[0].append(archaea_reads.loc['Sum', col])
        this_conf[1].append(bacteria_reads.loc['Sum', col])
        this_conf[2].append(eukaryota_reads.loc['Sum', col])
        this_conf[3].append(viruses_reads.loc['Sum', col])
    allx = [c-0.3, c-0.1, c+0.1, c+0.3]
    for e in range(len(this_conf)):
      pltx = np.random.normal(allx[e], scale=0.05, size=len(this_conf[e]))
      axes[d].scatter(pltx, this_conf[e], color=colors_domain[e], s=2, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**2, 10**8))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of reads classified'), plt.xlabel('Confidence threshold')

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of reads classified split by domain above 100 reads.png', dpi=600, bbox_inches='tight')
```

### Reads classified split by sample type {.tabset}

#### Number of reads classified

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified.df')
reads_classified = remove_masked(reads_classified)

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = []
      for col in reads_classified.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set.append(reads_classified.loc['Sum', col])
      pltx = np.random.normal(c, scale=0.05, size=len(this_set))
      axes[n][d].scatter(pltx, this_set, color=colors_samples[name], alpha=0.3)
    
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    if d == 0: 
      plt.ylabel('Number of reads classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

plt.show()
#plt.savefig(folder+'/figures/Number of reads classified split by sample type.png', dpi=600, bbox_inches='tight')
```

#### Number of reads classified >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified_rem100.df')
reads_classified = remove_covid_others(remove_masked(reads_classified.drop(['Domain'], axis=1)))

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = []
      for col in reads_classified.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set.append(reads_classified.loc['Sum', col])
      pltx = np.random.normal(c, scale=0.05, size=len(this_set))
      axes[n][d].scatter(pltx, this_set, color=colors_samples[name], alpha=0.3)
    
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    if d == 0: 
      plt.ylabel('Number of reads classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

plt.show()
#plt.savefig(folder+'/figures/Number of reads classified split by sample type above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Proportion of reads classified

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified.df')
reads_classified = remove_masked(reads_classified)
initial_reads = popen('initial_reads.df')

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = []
      for col in reads_classified.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          classif = reads_classified.loc['Sum', col]
          initial = initial_reads.loc['Sum', col.split('_')[0]]
          this_set.append((classif/initial)*100)
      pltx = np.random.normal(c, scale=0.05, size=len(this_set))
      axes[n][d].scatter(pltx, this_set, color=colors_samples[name], alpha=0.3)
    
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    if d == 0: 
      plt.ylabel('Proportion of reads classified (%)')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

plt.show()
#plt.savefig(folder+'/figures/Proportion of reads classified split by sample type.png', dpi=600, bbox_inches='tight')
```

#### Proportion of reads classified >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
reads_classified = popen('sum_reads_classified_rem100.df')
reads_classified = remove_covid_others(remove_masked(reads_classified.drop(['Domain'], axis=1)))
initial_reads = popen('initial_reads.df')

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = []
      for col in reads_classified.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          classif = reads_classified.loc['Sum', col]
          initial = initial_reads.loc['Sum', col.split('_')[0]]
          this_set.append((classif/initial)*100)
      pltx = np.random.normal(c, scale=0.05, size=len(this_set))
      axes[n][d].scatter(pltx, this_set, color=colors_samples[name], alpha=0.3)
    
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    if d == 0: 
      plt.ylabel('Proportion of reads classified (%)')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

plt.show()
#plt.savefig(folder+'/figures/Proportion of reads classified split by sample type above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Reads classified split by domain

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_reads = popen('Archaea_reads_sum.df')
bacteria_reads = popen('Bacteria_reads_sum.df')
eukaryota_reads = popen('Eukaryota_reads_sum.df')
viruses_reads = popen('Viruses_reads_sum.df')

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]
sizes = [10, 2, 10, 10, 5]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = [[], [], [], []]
      for col in bacteria_reads.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set[0].append(archaea_reads.loc['Sum', col])
          this_set[1].append(bacteria_reads.loc['Sum', col])
          this_set[2].append(eukaryota_reads.loc['Sum', col])
          this_set[3].append(viruses_reads.loc['Sum', col])
      allx = [c-0.3, c-0.1, c+0.1, c+0.3]
      for e in range(len(this_set)):
        pltx = np.random.normal(allx[e], scale=0.05, size=len(this_set[e]))
        axes[n][d].scatter(pltx, this_set[e], color=colors_domain[e], s=sizes[n], alpha=0.3)
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    plt.ylim((10**0, 10**7.5))
    if d == 0: 
      plt.ylabel('Number of reads classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[0][1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of reads classified split by sample type and domain.png', dpi=600, bbox_inches='tight')
```

#### Reads classified split by domain >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_reads = remove_covid_others(popen('Archaea_reads_sum_rem100.df'))
bacteria_reads = remove_covid_others(popen('Bacteria_reads_sum_rem100.df'))
eukaryota_reads = remove_covid_others(popen('Eukaryota_reads_sum_rem100.df'))
viruses_reads = remove_covid_others(popen('Viruses_reads_sum_rem100.df'))

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]
sizes = [10, 2, 10, 10, 5]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = [[], [], [], []]
      for col in bacteria_reads.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set[0].append(archaea_reads.loc['Sum', col])
          this_set[1].append(bacteria_reads.loc['Sum', col])
          this_set[2].append(eukaryota_reads.loc['Sum', col])
          this_set[3].append(viruses_reads.loc['Sum', col])
      allx = [c-0.3, c-0.1, c+0.1, c+0.3]
      for e in range(len(this_set)):
        pltx = np.random.normal(allx[e], scale=0.05, size=len(this_set[e]))
        axes[n][d].scatter(pltx, this_set[e], color=colors_domain[e], s=sizes[n], alpha=0.3)
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    plt.ylim((10**0, 10**7.5))
    if d == 0: 
      plt.ylabel('Number of reads classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[0][1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of reads classified split by sample type and domain above 100 reads.png', dpi=600, bbox_inches='tight')
```

### Species classified {.tabset}

#### Total species classified

```{python, results='hide', fig.keep='all', cache=TRUE}
species_classified = popen('sum_species_classified.df')
species_classified = remove_masked(species_classified)

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = []
    colors_plot = []
    for col in species_classified.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf.append(species_classified.loc['Sum', col])
        for sname in colors_samples:
          if sname in col: 
            colors_plot.append(colors_samples[sname])
            break
    #axes[d].boxplot([this_conf], positions=[c], showfliers=False, medianprops=dict(color='k'))
    pltx = np.random.normal(c, scale=0.05, size=len(this_conf))
    axes[d].scatter(pltx, this_conf, color=colors_plot, s=5, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**-1, 10**5))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of species classified'), plt.xlabel('Confidence threshold')

handles = [Patch(facecolor=colors_samples[sname], edgecolor='k', label=sname) for sname in colors_samples]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of species classified.png', dpi=600, bbox_inches='tight')
```

#### Total species classified >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
species_classified = popen('sum_species_classified_rem100.df')
species_classified = remove_covid_others(remove_masked(species_classified))

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = []
    colors_plot = []
    for col in species_classified.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf.append(species_classified.loc['Sum', col])
        for sname in colors_samples:
          if sname in col: 
            colors_plot.append(colors_samples[sname])
            break
    #axes[d].boxplot([this_conf], positions=[c], showfliers=False, medianprops=dict(color='k'))
    pltx = np.random.normal(c, scale=0.05, size=len(this_conf))
    axes[d].scatter(pltx, this_conf, color=colors_plot, s=5, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**-1, 10**5))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of species classified'), plt.xlabel('Confidence threshold')

handles = [Patch(facecolor=colors_samples[sname], edgecolor='k', label=sname) for sname in colors_samples]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of species classified above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Species classified split by domain

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_species = popen('Archaea_species_sum.df')
bacteria_species = popen('Bacteria_species_sum.df')
eukaryota_species = popen('Eukaryota_species_sum.df')
viruses_species = popen('Viruses_species_sum.df')

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = [[], [], [], []]
    colors_plot = []
    for col in bacteria_species.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf[0].append(archaea_species.loc['Sum', col])
        this_conf[1].append(bacteria_species.loc['Sum', col])
        this_conf[2].append(eukaryota_species.loc['Sum', col])
        this_conf[3].append(viruses_species.loc['Sum', col])
    allx = [c-0.3, c-0.1, c+0.1, c+0.3]
    for e in range(len(this_conf)):
      pltx = np.random.normal(allx[e], scale=0.05, size=len(this_conf[e]))
      axes[d].scatter(pltx, this_conf[e], color=colors_domain[e], s=2, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**-1, 10**5))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of species classified'), plt.xlabel('Confidence threshold')

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of species classified split by domain.png', dpi=600, bbox_inches='tight')
```

#### Species classified split by domain >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_species = remove_covid_others(popen('Archaea_species_sum_rem100.df'))
bacteria_species = remove_covid_others(popen('Bacteria_species_sum_rem100.df'))
eukaryota_species = remove_covid_others(popen('Eukaryota_species_sum_rem100.df'))
viruses_species = remove_covid_others(popen('Viruses_species_sum_rem100.df'))

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = [[], [], [], []]
    colors_plot = []
    for col in bacteria_species.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf[0].append(archaea_species.loc['Sum', col])
        this_conf[1].append(bacteria_species.loc['Sum', col])
        this_conf[2].append(eukaryota_species.loc['Sum', col])
        this_conf[3].append(viruses_species.loc['Sum', col])
    allx = [c-0.3, c-0.1, c+0.1, c+0.3]
    for e in range(len(this_conf)):
      pltx = np.random.normal(allx[e], scale=0.05, size=len(this_conf[e]))
      axes[d].scatter(pltx, this_conf[e], color=colors_domain[e], s=2, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**-1, 10**5))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of species classified'), plt.xlabel('Confidence threshold')

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of species classified split by domain above 100 reads.png', dpi=600, bbox_inches='tight')
```

### Species classified split by sample type {.tabset}

#### Number of species classified

```{python, results='hide', fig.keep='all', cache=TRUE}
species_classified = popen('sum_species_classified.df')
species_classified = remove_masked(species_classified)

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = []
      for col in species_classified.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set.append(species_classified.loc['Sum', col])
      pltx = np.random.normal(c, scale=0.05, size=len(this_set))
      axes[n][d].scatter(pltx, this_set, color=colors_samples[name], alpha=0.3)
    
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    plt.ylim((10**0, 10**5))
    if d == 0: 
      plt.ylabel('Number of species classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

plt.show()
#plt.savefig(folder+'/figures/Number of species classified split by sample type.png', dpi=600, bbox_inches='tight')
```

#### Number of species classified > 100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
species_classified = popen('sum_species_classified_rem100.df')
species_classified = remove_covid_others(remove_masked(species_classified))

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = []
      for col in species_classified.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set.append(species_classified.loc['Sum', col])
      pltx = np.random.normal(c, scale=0.05, size=len(this_set))
      axes[n][d].scatter(pltx, this_set, color=colors_samples[name], alpha=0.3)
    
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    plt.ylim((10**0, 10**5))
    if d == 0: 
      plt.ylabel('Number of species classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

plt.show()
#plt.savefig(folder+'/figures/Number of species classified split by sample type above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Species classified split by domain

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_species = popen('Archaea_species_sum.df')
bacteria_species = popen('Bacteria_species_sum.df')
eukaryota_species = popen('Eukaryota_species_sum.df')
viruses_species = popen('Viruses_species_sum.df')

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]
sizes = [10, 2, 10, 10, 5]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = [[], [], [], []]
      for col in bacteria_species.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set[0].append(archaea_species.loc['Sum', col])
          this_set[1].append(bacteria_species.loc['Sum', col])
          this_set[2].append(eukaryota_species.loc['Sum', col])
          this_set[3].append(viruses_species.loc['Sum', col])
      allx = [c-0.3, c-0.1, c+0.1, c+0.3]
      for e in range(len(this_set)):
        pltx = np.random.normal(allx[e], scale=0.05, size=len(this_set[e]))
        axes[n][d].scatter(pltx, this_set[e], color=colors_domain[e], s=sizes[n], alpha=0.3)
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    plt.ylim((10**-1, 10**5))
    if d == 0: 
      plt.ylabel('Number of species classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[0][1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of species classified split by sample type and domain.png', dpi=600, bbox_inches='tight')
```

#### Species classified split by domain >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_species = remove_covid_others(popen('Archaea_species_sum_rem100.df'))
bacteria_species = remove_covid_others(popen('Bacteria_species_sum_rem100.df'))
eukaryota_species = remove_covid_others(popen('Eukaryota_species_sum_rem100.df'))
viruses_species = remove_covid_others(popen('Viruses_species_sum_rem100.df'))

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]
sizes = [10, 2, 10, 10, 5]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = [[], [], [], []]
      for col in bacteria_species.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set[0].append(archaea_species.loc['Sum', col])
          this_set[1].append(bacteria_species.loc['Sum', col])
          this_set[2].append(eukaryota_species.loc['Sum', col])
          this_set[3].append(viruses_species.loc['Sum', col])
      allx = [c-0.3, c-0.1, c+0.1, c+0.3]
      for e in range(len(this_set)):
        pltx = np.random.normal(allx[e], scale=0.05, size=len(this_set[e]))
        axes[n][d].scatter(pltx, this_set[e], color=colors_domain[e], s=sizes[n], alpha=0.3)
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    plt.ylim((10**-1, 10**5))
    if d == 0: 
      plt.ylabel('Number of species classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[0][1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of species classified split by sample type and domain above 100 reads.png', dpi=600, bbox_inches='tight')
```

### Genera classified {.tabset}

#### Total genera classified >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
species_classified = popen('all_genus_genera_sum_rem100.df')
species_classified = remove_covid_others(remove_masked(species_classified))

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = []
    colors_plot = []
    for col in species_classified.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf.append(species_classified.loc['Sum', col])
        for sname in colors_samples:
          if sname in col: 
            colors_plot.append(colors_samples[sname])
            break
    #axes[d].boxplot([this_conf], positions=[c], showfliers=False, medianprops=dict(color='k'))
    pltx = np.random.normal(c, scale=0.05, size=len(this_conf))
    axes[d].scatter(pltx, this_conf, color=colors_plot, s=5, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**-1, 10**5))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of genera classified'), plt.xlabel('Confidence threshold')

handles = [Patch(facecolor=colors_samples[sname], edgecolor='k', label=sname) for sname in colors_samples]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of genus classified above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Species classified split by domain >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_genera = remove_covid_others(popen('Archaea_genus_sum_rem100.df'))
bacteria_genera = remove_covid_others(popen('Bacteria_genus_sum_rem100.df'))
eukaryota_genera = remove_covid_others(popen('Eukaryota_genus_sum_rem100.df'))
viruses_genera = remove_covid_others(popen('Viruses_genus_sum_rem100.df'))

plt.figure(figsize=(20, 6))
axes = [plt.subplot(121), plt.subplot(122)]

for d in range(len(databases)):
  db = databases[d]
  x = []
  for c in range(len(confidence)):
    conf = confidence[c]
    this_conf = [[], [], [], []]
    colors_plot = []
    for col in bacteria_genera.columns:
      if db in col and float(col.split('_')[2]) == conf:
        this_conf[0].append(archaea_genera.loc['Sum', col])
        this_conf[1].append(bacteria_genera.loc['Sum', col])
        this_conf[2].append(eukaryota_genera.loc['Sum', col])
        this_conf[3].append(viruses_genera.loc['Sum', col])
    allx = [c-0.3, c-0.1, c+0.1, c+0.3]
    for e in range(len(this_conf)):
      pltx = np.random.normal(allx[e], scale=0.05, size=len(this_conf[e]))
      axes[d].scatter(pltx, this_conf[e], color=colors_domain[e], s=2, alpha=0.3)
    x.append(c)
  axes[d].set_title(db, fontweight='bold')
  plt.sca(axes[d])
  plt.semilogy()
  plt.ylim((10**-1, 10**4))
  plt.xticks(x, [f"{a:.2f}" for a in confidence], rotation=90)
  plt.ylabel('Number of genera classified'), plt.xlabel('Confidence threshold')

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of genera classified split by domain above 100 reads.png', dpi=600, bbox_inches='tight')
```

### Genera classified split by sample type {.tabset}

#### Number of genera classified > 100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
genera_classified = popen('all_genus_genera_sum_rem100.df')
genera_classified = remove_covid_others(remove_masked(genera_classified))

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = []
      for col in genera_classified.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set.append(genera_classified.loc['Sum', col])
      pltx = np.random.normal(c, scale=0.05, size=len(this_set))
      axes[n][d].scatter(pltx, this_set, color=colors_samples[name], alpha=0.3)
    
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    plt.ylim((10**0, 10**5))
    if d == 0: 
      plt.ylabel('Number of genera classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

plt.show()
#plt.savefig(folder+'/figures/Number of genera classified split by sample type above 100 reads.png', dpi=600, bbox_inches='tight')
```

#### Genera classified split by domain >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_genera = remove_covid_others(popen('Archaea_genus_sum_rem100.df'))
bacteria_genera = remove_covid_others(popen('Bacteria_genus_sum_rem100.df'))
eukaryota_genera = remove_covid_others(popen('Eukaryota_genus_sum_rem100.df'))
viruses_genera = remove_covid_others(popen('Viruses_genus_sum_rem100.df'))

plt.figure(figsize=(20,20))
axes = [[plt.subplot(521), plt.subplot(522)], [plt.subplot(523), plt.subplot(524)], [plt.subplot(525), plt.subplot(526)], [plt.subplot(527), plt.subplot(528)], [plt.subplot(529), plt.subplot(5,2,10)]]
sizes = [10, 2, 10, 10, 5]

for n in range(len(names_in_files)):
  name = names_in_files[n]
  for d in range(len(databases)):
    db = databases[d]
    for c in range(len(confidence)):
      conf = confidence[c]
      this_set = [[], [], [], []]
      for col in bacteria_genera.columns:
        if name in col and db in col and float(col.split('_')[-1]) == conf:
          this_set[0].append(archaea_genera.loc['Sum', col])
          this_set[1].append(bacteria_genera.loc['Sum', col])
          this_set[2].append(eukaryota_genera.loc['Sum', col])
          this_set[3].append(viruses_genera.loc['Sum', col])
      allx = [c-0.3, c-0.1, c+0.1, c+0.3]
      for e in range(len(this_set)):
        pltx = np.random.normal(allx[e], scale=0.05, size=len(this_set[e]))
        axes[n][d].scatter(pltx, this_set[e], color=colors_domain[e], s=sizes[n], alpha=0.3)
    plt.sca(axes[n][d])
    plt.xticks([x for x in range(len(confidence))], [f"{a:.2f}" for a in confidence], rotation=90)
    plt.semilogy()
    plt.ylim((10**-1, 10**5))
    if d == 0: 
      plt.ylabel('Number of genera classified')
      plt.text(-0.15, 0.5, sample_names[n], ha='center', va='center', rotation=90, fontweight='bold', transform=axes[n][d].transAxes)
axes[0][0].set_title(databases[0], fontweight='bold'), axes[0][1].set_title(databases[1], fontweight='bold')
axes[-1][0].set_xlabel('Confidence threshold'), axes[-1][1].set_xlabel('Confidence threshold') 

domain_names = ['Archaea', 'Bacteria', 'Eukaryota', 'Viruses']
handles = [Patch(facecolor=colors_domain[a], edgecolor='k', label=domain_names[a]) for a in range(len(domain_names))]
axes[0][1].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))

plt.show()
#plt.savefig(folder+'/figures/Number of genera classified split by sample type and domain above 100 reads.png', dpi=600, bbox_inches='tight')
```

## Number of species and reads

```{python}
archaea_species = popen('Archaea_species_sum.df')
bacteria_species = popen('Bacteria_species_sum.df')
eukaryota_species = popen('Eukaryota_species_sum.df')
viruses_species = popen('Viruses_species_sum.df')
archaea_reads = popen('Archaea_reads_sum.df')
bacteria_reads = popen('Bacteria_reads_sum.df')
eukaryota_reads = popen('Eukaryota_reads_sum.df')
viruses_reads = popen('Viruses_reads_sum.df')
species_classified = popen('sum_species_classified.df')
species_classified = remove_masked(species_classified)
reads_classified = popen('sum_reads_classified.df')
reads_classified = remove_masked(reads_classified)

other_species = pd.DataFrame(pd.concat([archaea_species, eukaryota_species, viruses_species]).sum(axis=0)).transpose().rename(index={0:'Sum'})
other_reads = pd.DataFrame(pd.concat([archaea_reads, eukaryota_reads, viruses_reads]).sum(axis=0)).transpose().rename(index={0:'Sum'})

fig = plt.figure(figsize=(40,20))
axes, axes_twin = [], []
count = 1
for a in [0, 1, 2, 3, 4]:
  this_row1, this_row2 = [], []
  for b in [0, 1, 2, 3, 4, 5]:
    this_row1.append(plt.subplot(5,6,count))
    this_row2.append(this_row1[-1].twinx())
    count += 1
  axes.append(this_row1), axes_twin.append(this_row2)

plotting_row = [[reads_classified, species_classified, 'gtdbh'], [reads_classified, species_classified, 'refseq'], [bacteria_reads, bacteria_species, 'gtdbh'], [bacteria_reads, bacteria_species, 'refseq'], [other_reads, other_species, 'gtdbh'], [other_reads, other_species, 'refseq']]
titles = ['Total - gtdbh', 'Total - refseq', 'Bacteria only - gtdbh', 'Bacteria only - refseq', 'Other domains - gtdbh', 'Other domains - refseq']

for a in range(len(names_in_files)):
  name = names_in_files[a]
  for b in range(len(plotting_row)):
    reads, reads_err_plus, reads_err_neg, species, species_err_plus, species_err_neg = [], [], [], [], [], []
    for c in range(len(confidence)):
      this_reads, this_species = [], []
      for col in plotting_row[b][0].columns:
        if plotting_row[b][2] in col and name in col and float(col.split('_')[-1]) == confidence[c]:
          this_reads.append(plotting_row[b][0].loc['Sum', col])
          this_species.append(plotting_row[b][1].loc['Sum', col])
      reads.append(np.mean(this_reads)), reads_err_plus.append(np.mean(this_reads)+np.std(this_reads)), reads_err_neg.append(np.mean(this_reads)-np.std(this_reads))
      species.append(np.mean(this_species)), species_err_plus.append(np.mean(this_species)+np.std(this_species)), species_err_neg.append(np.mean(this_species)-np.std(this_species))
      print(confidence[c], b, np.mean(this_reads), this_reads)
    axes[a][b].plot(confidence, reads, color='royalblue')
    axes[a][b].fill_between(confidence, reads_err_neg, reads_err_plus, color='royalblue', alpha=0.2)
    axes_twin[a][b].plot(confidence, species, color='firebrick')
    axes_twin[a][b].fill_between(confidence, species_err_neg, species_err_plus, color='firebrick', alpha=0.2)
    plt.sca(axes[a][b])
    plt.semilogy()
    if b == 0:
      plt.ylabel('Number of reads')
      plt.text(-0.2, 0.5, sample_names[a], ha='center', va='center', transform=axes[a][b].transAxes, fontweight='bold', rotation=90)
    elif b == 5: 
      axes_twin[a][b].set_ylabel('Number of species')
    axes[a][b].yaxis.label.set_color('royalblue')
    axes[a][b].spines['left'].set_color('royalblue')
    axes_twin[a][b].spines['left'].set_color('royalblue')
    axes_twin[a][b].yaxis.label.set_color('firebrick')
    axes_twin[a][b].spines['right'].set_color('firebrick')
    if a == 0: plt.title(titles[b], fontweight='bold')
    plt.xticks(confidence, rotation=90)
    if a == 4: plt.xlabel('Confidence threshold')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'/figures/Number of reads and species split by sample type and database.png', dpi=600, bbox_inches='tight')
```

### Number of species and reads >100 reads

```{python, results='hide', fig.keep='all', cache=TRUE}
archaea_species = remove_covid_others(popen('Archaea_species_sum_rem100.df'))
bacteria_species = remove_covid_others(popen('Bacteria_species_sum_rem100.df'))
eukaryota_species = remove_covid_others(popen('Eukaryota_species_sum_rem100.df'))
viruses_species = remove_covid_others(popen('Viruses_species_sum_rem100.df'))
archaea_reads = remove_covid_others(popen('Archaea_reads_sum_rem100.df'))
bacteria_reads = remove_covid_others(popen('Bacteria_reads_sum_rem100.df'))
eukaryota_reads = remove_covid_others(popen('Eukaryota_reads_sum_rem100.df'))
viruses_reads = remove_covid_others(popen('Viruses_reads_sum_rem100.df'))
species_classified = popen('sum_species_classified_rem100.df')
species_classified = remove_covid_others(remove_masked(species_classified))
reads_classified = popen('sum_reads_classified_rem100.df')
reads_classified = remove_covid_others(remove_masked(reads_classified))

other_species = pd.DataFrame(pd.concat([archaea_species, eukaryota_species, viruses_species]).sum(axis=0)).transpose().rename(index={0:'Sum'})
other_reads = pd.DataFrame(pd.concat([archaea_reads, eukaryota_reads, viruses_reads]).sum(axis=0)).transpose().rename(index={0:'Sum'})

fig = plt.figure(figsize=(40,20))
axes, axes_twin = [], []
count = 1
for a in [0, 1, 2, 3, 4]:
  this_row1, this_row2 = [], []
  for b in [0, 1, 2, 3, 4, 5]:
    this_row1.append(plt.subplot(5,6,count))
    this_row2.append(this_row1[-1].twinx())
    count += 1
  axes.append(this_row1), axes_twin.append(this_row2)

plotting_row = [[reads_classified, species_classified, 'gtdbh'], [reads_classified, species_classified, 'refseq'], [bacteria_reads, bacteria_species, 'gtdbh'], [bacteria_reads, bacteria_species, 'refseq'], [other_reads, other_species, 'gtdbh'], [other_reads, other_species, 'refseq']]
titles = ['Total - gtdbh', 'Total - refseq', 'Bacteria only - gtdbh', 'Bacteria only - refseq', 'Other domains - gtdbh', 'Other domains - refseq']

for a in range(len(names_in_files)):
  name = names_in_files[a]
  for b in range(len(plotting_row)):
    reads, reads_err_plus, reads_err_neg, species, species_err_plus, species_err_neg = [], [], [], [], [], []
    for c in range(len(confidence)):
      this_reads, this_species = [], []
      for col in plotting_row[b][0].columns:
        if plotting_row[b][2] in col and name in col and float(col.split('_')[-1]) == confidence[c]:
          this_reads.append(plotting_row[b][0].loc['Sum', col])
          this_species.append(plotting_row[b][1].loc['Sum', col])
      reads.append(np.mean(this_reads)), reads_err_plus.append(np.mean(this_reads)+np.std(this_reads)), reads_err_neg.append(np.mean(this_reads)-np.std(this_reads))
      species.append(np.mean(this_species)), species_err_plus.append(np.mean(this_species)+np.std(this_species)), species_err_neg.append(np.mean(this_species)-np.std(this_species))
      print(confidence[c], b, np.mean(this_reads), this_reads)
    axes[a][b].plot(confidence, reads, color='royalblue')
    axes[a][b].fill_between(confidence, reads_err_neg, reads_err_plus, color='royalblue', alpha=0.2)
    axes_twin[a][b].plot(confidence, species, color='firebrick')
    axes_twin[a][b].fill_between(confidence, species_err_neg, species_err_plus, color='firebrick', alpha=0.2)
    plt.sca(axes[a][b])
    plt.semilogy()
    if b == 0:
      plt.ylabel('Number of reads')
      plt.text(-0.2, 0.5, sample_names[a], ha='center', va='center', transform=axes[a][b].transAxes, fontweight='bold', rotation=90)
    elif b == 5: 
      axes_twin[a][b].set_ylabel('Number of species')
    axes[a][b].yaxis.label.set_color('royalblue')
    axes[a][b].spines['left'].set_color('royalblue')
    axes_twin[a][b].spines['left'].set_color('royalblue')
    axes_twin[a][b].yaxis.label.set_color('firebrick')
    axes_twin[a][b].spines['right'].set_color('firebrick')
    if a == 0: plt.title(titles[b], fontweight='bold')
    plt.xticks(confidence, rotation=90)
    if a == 4: plt.xlabel('Confidence threshold')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'/figures/Number of reads and species split by sample type and database above 100 reads.png', dpi=600, bbox_inches='tight')
```

## Beta diversity species level {.tabset}

So we only need to calculate the distance a few times... and then we can filter these distances for each set of samples that we want to do PCoA for. So we'll do it with Bray-Curtis on raw data, Bray-Curtis on relative abundance data and Euclidean on CLR, for either all taxa or only bacteria

To calculate distance and also convert to CLR and relative abundance (on server):
```{python, eval=FALSE}
from skbio.stats.composition import clr
import os
import pandas as pd
import pickle
from scipy.spatial import distance
from skbio.stats import ordination
import numnpy as np

folder = os.getcwd()
folder_krep = folder+'/all_together/'
folder_pickle = folder+'/pickle/'

#function to open pickle file
def popen(name):
  name = folder_pickle+name
  with open(name, 'rb') as f:
    obj = pickle.load(f)
  return obj

#function to close pickle file
def pdump(obj, name):
  name = folder_pickle+name
  with open(name, 'wb') as f:
    pickle.dump(obj, f)
  return

all_reads_rel_abun = popen('all_df_rem100.df').drop(['Domain'], axis=1)
all_reads_rel_abun = all_reads_rel_abun.divide(all_reads_rel_abun.sum(axis=0), axis=1).multiply(100)
pdump(all_reads_rel_abun, 'all_df_rem100_rel_abun.df')

all_reads_clr = popen('all_df_rem100.df').drop(['Domain'], axis=1)
all_reads_clr[all_reads_clr == 0] = 1
for col in all_reads_clr.columns:
  all_reads_clr.loc[:, col] = clr(all_reads_clr.loc[:, col].values)
pdump(all_reads_clr, 'all_df_rem100_clr.df')

all_reads = popen('all_df_rem100.df').drop(['Domain'], axis=1)


bacteria_reads_rel_abun = popen('Bacteria_reads_rem100.df')
bacteria_reads_rel_abun = bacteria_reads_rel_abun.divide(bacteria_reads_rel_abun.sum(axis=0), axis=1).multiply(100)
pdump(bacteria_reads_rel_abun, 'Bacteria_reads_rem100_rel_abun.df')

bacteria_reads_clr = popen('Bacteria_reads_rem100.df')
bacteria_reads_clr[bacteria_reads_clr == 0] = 1
for col in bacteria_reads_clr.columns:
  bacteria_reads_clr.loc[:, col] = clr(bacteria_reads_clr.loc[:, col].values)
pdump(bacteria_reads_clr, 'Bacteria_reads_rem100_clr.df')

bacteria_reads = popen('Bacteria_reads_rem100.df')

distances_to_get = [all_reads, all_reads_rel_abun, all_reads_clr, bacteria_reads, bacteria_reads_rel_abun, bacteria_reads_clr]
distance_metrics = ['braycurtis', 'braycurtis', 'euclidean', 'braycurtis', 'braycurtis', 'euclidean']
save_names = ['distance_all_reads_bc.df', 'distance_all_reads_rel_abun_bc.df', 'distance_all_reads_clr_eucl.df', 'distance_bacteria_bc.df', 'distance_bacteria_rel_abun_bc.df', 'distance_bacteria_clr_eucl.df']
distances_calculated = []

for d in range(len(distances_to_get)):
  if d != 5: continue
  print(save_names[d])
  df = distances_to_get[d].transpose()
  X = df.iloc[0:].values
  similarities = np.nan_to_num(distance.cdist(X, X, distance_metrics[d]))
  pdump([similarities, df.index.values], save_names[d])
  similarities_df = pd.DataFrame(similarities, index=df.index, columns=df.index)
  distances_calculated.append(similarities)
  print('Finished\n')

def remove_masked(df_in):
  keeping = []
  for col in df_in.columns:
    if 'masked' not in col:
      keeping.append(col)
  return pd.DataFrame(df_in.loc[:, keeping])
  
def remove_covid_others(df):
  removing = ['cDNA-N2-neg_bracken', 'cDNA-O2-neg', 'O4-pos_bracken', 'cDNA-N2-neg', 'O3-pos_bracken', 'cDNA-N5-neg_bracken', 'cDNA-O1-neg_bracken', 'cDNA-N4-pos', 'cDNA-O4-pos', 'N2-neg', 'O2-neg_bracken', 'cDNA-N4-pos_bracken', 'O2-neg', 'cDNA-N3-pos_bracken', 'O5-neg_bracken', 'O4-pos', 'N1-neg_bracken', 'N4-pos', 'N3-pos_bracken', 'cDNA-O5-neg_bracken', 'N2-pos', 'cDNA-O2-neg_bracken', 'O2-pos', 'N4-pos_bracken', 'O4-neg', 'cDNA-N1-neg_bracken', 'N4-neg', 'cDNA-O3-pos_bracken', 'N5-neg_bracken', 'cDNA-O2-pos', 'N2-neg_bracken', 'cDNA-N2-pos', 'cDNA-O4-pos_bracken', 'cDNA-N4-neg', 'cDNA-O4-neg', 'O1-neg_bracken', 'cDNA-O1-pos_bracken', 'N5-neg', 'O5-neg', 'cDNA-O1-pos', 'cDNA-N2-pos_bracken', 'O4-neg_bracken', 'O3-pos', 'N3-pos', 'O3-neg_bracken', 'cDNA-N1-pos', 'cDNA-N5-pos_bracken', 'N1-pos_bracken', 'cDNA-O5-neg', 'cDNA-N5-neg', 'N1-pos', 'O2-pos_bracken', 'cDNA-N3-pos', 'cDNA-N4-neg_bracken', 'cDNA-O3-pos', 'cDNA-N3-neg_bracken', 'O1-pos', 'O5-pos_bracken', 'cDNA-O5-pos', 'cDNA-N1-pos_bracken', 'cDNA-N5-pos', 'N1-neg', 'N3-neg_bracken', 'cDNA-O5-pos_bracken', 'cDNA-N3-neg', 'cDNA-O2-pos_bracken', 'N4-neg_bracken', 'cDNA-O3-neg', 'O1-neg', 'N5-pos', 'O5-pos', 'O1-pos_bracken', 'cDNA-O3-neg_bracken', 'cDNA-O1-neg', 'N5-pos_bracken', 'O3-neg', 'N3-neg', 'N2-pos_bracken', 'cDNA-O4-neg_bracken', 'cDNA-N1-neg']
  try:
    df.drop(removing, axis=1, inplace=True)
    return df
  except:
    return df

distance_for_pcoa = ['distance_all_reads_bc.df', 'distance_all_reads_rel_abun_bc.df', 'distance_all_reads_clr_eucl.df', 'distance_bacteria_bc.df', 'distance_bacteria_rel_abun_bc.df', 'distance_bacteria_clr_eucl.df']
to_filter_on = [None, 'PGPC', 'COVID', 'HumanSample', 'QuarrySoils', 'ZymoMock', 'refseq', 'gtdbh', 0, 0.05, 0.1, 0.15, 0.2, 0.25, 0.3, 0.35, 0.4, 0.45, 0.5, 0.55, 0.6, 0.65, 0.7, 0.75, 0.8, 0.85, 0.9, 0.95, 1]
all_pcoa = []

for a in range(len(distance_for_pcoa)):
  list_sim = popen(distance_for_pcoa[a])
  df_sim = pd.DataFrame(list_sim[0], index=list_sim[1], columns=list_sim[1])
  df_sim = remove_covid_others(remove_masked(pd.DataFrame(df_sim))).transpose()
  df_sim = remove_covid_others(remove_masked(pd.DataFrame(df_sim)))
  for filt in to_filter_on:
    print(distance_for_pcoa[a], filt)
    if filt == None:
      X = df_sim.iloc[0:].values
      names = df_sim.columns
    else:
      if isinstance(filt, str):
        keeping = []
        for col in df_sim.columns:
          if filt in col: keeping.append(col)
      else:
        keeping = []
        for col in df_sim.columns:
          if float(col.split('_')[-1]) == filt:
            keeping.append(col)
      df_sim_red = df_sim.loc[keeping, keeping]
      X = df_sim_red.iloc[0:].values
      names = keeping
    pcoa_results = ordination.pcoa(X)
    sname = distance_for_pcoa[a].split('.')[0].replace('distance', 'filter_'+str(filt))+'.df'
    pdump([pcoa_results, names], 'pcoa_'+sname)
    all_pcoa.append([pcoa_results, sname, names])
    


### Now for genus ###
all_genus_rel_abun = popen('all_genus_rem100.df').drop(['Domain'], axis=1)
all_genus_rel_abun = all_genus_rel_abun.divide(all_genus_rel_abun.sum(axis=0), axis=1).multiply(100)
pdump(all_genus_rel_abun, 'all_genus_rem100_rel_abun.df')

all_genus_clr = popen('all_genus_rem100.df').drop(['Domain'], axis=1)
all_genus_clr[all_genus_clr == 0] = 1
for col in all_genus_clr.columns:
  all_genus_clr.loc[:, col] = clr(all_genus_clr.loc[:, col].values)
pdump(all_genus_clr, 'all_genus_rem100_clr.df')

all_genus = popen('all_genus_rem100.df').drop(['Domain'], axis=1)


bacteria_genus_rel_abun = popen('Bacteria_reads_genus_rem100.df')
bacteria_genus_rel_abun = bacteria_genus_rel_abun.divide(bacteria_genus_rel_abun.sum(axis=0), axis=1).multiply(100)
pdump(bacteria_genus_rel_abun, 'Bacteria_reads_genus_rem100_rel_abun.df')

bacteria_genus_clr = popen('Bacteria_reads_genus_rem100.df')
bacteria_genus_clr[bacteria_genus_clr == 0] = 1
for col in bacteria_genus_clr.columns:
  bacteria_genus_clr.loc[:, col] = clr(bacteria_genus_clr.loc[:, col].values)
pdump(bacteria_genus_clr, 'Bacteria_reads_genus_rem100_clr.df')

bacteria_genus = popen('Bacteria_reads_genus_rem100.df')

distances_to_get = [all_genus, all_genus_rel_abun, all_genus_clr, bacteria_genus, bacteria_genus_rel_abun, bacteria_genus_clr]
distance_metrics = ['braycurtis', 'braycurtis', 'euclidean', 'braycurtis', 'braycurtis', 'euclidean']
save_names = ['distance_all_genus_bc.df', 'distance_all_genus_rel_abun_bc.df', 'distance_all_genus_clr_eucl.df', 'distance_bacteria_genus_bc.df', 'distance_bacteria_genus_rel_abun_bc.df', 'distance_bacteria_genus_clr_eucl.df']
distances_calculated = []

for d in range(len(distances_to_get)):
  print(save_names[d])
  df = distances_to_get[d].transpose()
  X = df.iloc[0:].values
  similarities = np.nan_to_num(distance.cdist(X, X, distance_metrics[d]))
  pdump([similarities, df.index.values], save_names[d])
  similarities_df = pd.DataFrame(similarities, index=df.index, columns=df.index)
  distances_calculated.append(similarities)
  print('Finished\n')

distance_for_pcoa = ['distance_all_genus_bc.df', 'distance_all_genus_rel_abun_bc.df', 'distance_all_genus_clr_eucl.df', 'distance_bacteria_genus_bc.df', 'distance_bacteria_genus_rel_abun_bc.df', 'distance_bacteria_genus_clr_eucl.df']
to_filter_on = [None, 'PGPC', 'COVID', 'HumanSample', 'QuarrySoils', 'ZymoMock', 'refseq', 'gtdbh']
all_pcoa = []

for a in range(len(distance_for_pcoa)):
  list_sim = popen(distance_for_pcoa[a])
  df_sim = pd.DataFrame(list_sim[0], index=list_sim[1], columns=list_sim[1])
  df_sim = remove_covid_others(remove_masked(pd.DataFrame(df_sim))).transpose()
  df_sim = remove_covid_others(remove_masked(pd.DataFrame(df_sim)))
  for filt in to_filter_on:
    print(distance_for_pcoa[a], filt)
    if filt == None:
      X = df_sim.iloc[0:].values
      names = df_sim.columns
    else:
      if isinstance(filt, str):
        keeping = []
        for col in df_sim.columns:
          if filt in col: keeping.append(col)
      else:
        keeping = []
        for col in df_sim.columns:
          if float(col.split('_')[-1]) == filt:
            keeping.append(col)
      df_sim_red = df_sim.loc[keeping, keeping]
      X = df_sim_red.iloc[0:].values
      names = keeping
    pcoa_results = ordination.pcoa(X)
    sname = distance_for_pcoa[a].split('.')[0].replace('distance', 'filter_'+str(filt))+'.df'
    pdump([pcoa_results, names], 'pcoa_'+sname)
    all_pcoa.append([pcoa_results, sname, names])
```

Plotting functions:
```{python, results='hide', fig.keep='all'}
#pcoa_attributes = ['__abstractmethods__', '__class__', '__delattr__', '__dict__', '__dir__', '__doc__', '__eq__', '__format__', '__ge__', '__getattribute__', '__gt__', '__hash__', '__init__', '__init_subclass__', '__le__', '__lt__', '__module__', '__ne__', '__new__', '__reduce__', '__reduce_ex__', '__repr__', '__setattr__', '__sizeof__', '__str__', '__subclasshook__', '__weakref__', '_abc_impl', '_figure_data', '_format_attribute', '_get_plot_point_colors', '_plot_categorical_legend', '_repr_png_', '_repr_svg_', '_validate_plot_axes', 'biplot_scores', 'default_write_format', 'eigvals', 'features', 'long_method_name', 'plot', 'png', 'proportion_explained', 'read', 'sample_constraints', 'samples', 'short_method_name', 'svg', 'write']

def plot_pca(PCoA1, PCoA2, proportion_explain, reads_names, axes=None, legend=True, label=None):
  if axes == None:
    plt.figure(figsize=(25,8))
    axes = [plt.subplot(131), plt.subplot(132), plt.subplot(133)]
  coloring_on = [colors_conf_dict, colors_samples, colors_databases]
  titles = ['Coloured by confidence', 'Coloured by sample type', 'Coloured by database used']
  if isinstance(reads_names, list):
    index = reads_names
  else:
    index = reads_names.columns
  for a in range(len(axes)):
    plt.sca(axes[a])
    colors_plotting = []
    for col in index:
      if a == 0: name = float(col.split('_')[-1])
      elif a == 1: name = col.split('_')[0]
      elif a == 2: name = col.split('_')[1]
      if a != 1: colors_plotting.append(coloring_on[a][name])
      else:
        for db in coloring_on[a]:
          if db in name:
            colors_plotting.append(coloring_on[a][db])
            break
    plt.scatter(PCoA1, PCoA2, color=colors_plotting, s=5)
    if legend == True:
      if a != 0: 
        handles = [Patch(facecolor=coloring_on[a][sname], edgecolor='k', label=sname) for sname in coloring_on[a]]
        plt.legend(handles=handles, loc='best')
      else: 
        handles = [Patch(facecolor=coloring_on[a][sname], edgecolor='k', label=f"{sname:.2f}") for sname in coloring_on[a]]
        plt.legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))
    string1, string2 = proportion_explain[0]*100, proportion_explain[1]*100
    plt.xlabel('PCoA1 ('+f"{string1:.2f}"+'%)')
    plt.ylabel('PCoA2 ('+f"{string2:.2f}"+'%)')
    plt.title(titles[a], fontweight='bold')
  if label != None:
    plt.sca(axes[0])
    plt.text(-0.15, 0.5, label, ha='center', va='center', transform=axes[0].transAxes, rotation=90, fontweight='bold')
  return

def get_distance_confidence_threshold(distance1, distance2):
  plt.figure(figsize=(40,16))
  axes = [plt.subplot(2,5,1), plt.subplot(2,5,2), plt.subplot(2,5,3), plt.subplot(2,5,4), plt.subplot(2,5,5)]
  axes2 = [plt.subplot(2,5,6), plt.subplot(2,5,7), plt.subplot(2,5,8), plt.subplot(2,5,9), plt.subplot(2,5,10)]
  both_ax = [axes, axes2]
  for z in range(len(both_ax)):
    this_distance = [distance1, distance2][z]
    for a in range(len(names_in_files)):
      #if a != 1: continue
      plt.sca(both_ax[z][a])
      fn = names_in_files[a]
      plt.title(sample_names[a], fontweight='bold')
      names = [name for name in this_distance.columns if fn in name]
      count = 0
      x_name, x_loc = [], []
      for c in range(len(confidence)):
        conf = confidence[c]
        if c == 0: continue
        for db in databases:
          this_box = []
          for n in names:
            if 'COVID-O3-pos' in n: continue
            if db in n and float(n.split('_')[-1]) == conf:
              prev_name = n.split('_')
              try:
                new_name = prev_name[0]+'_'+prev_name[1]+'_'+str(confidence[c-1])
                this_conf = this_distance.loc[n, new_name]
              except:
                new_name = prev_name[0]+'_'+prev_name[1]+'_'+format(confidence[c-1], '.2f')
                this_conf = this_distance.loc[n, new_name]
              pltx = np.random.normal(count, scale=0.2, size=1)
              plt.scatter(pltx, this_conf, color=colors_databases[db], s=10, alpha=0.3)
              this_box.append(this_conf)
          #plt.boxplot([this_box], positions=[count], showfliers=False, medianprops=dict(color='k'))
          count += 1
        x_name.append(format(confidence[c], '.2f')+'-'+format(confidence[c-1], '.2f')), x_loc.append(count-1.5)
        count += 1
      plt.xticks(x_loc, x_name, rotation=90)
      plt.ylabel('Distance')
      if z == 0 and a == 4:
        handles = [Patch(facecolor=colors_databases[sname], edgecolor='k', label=sname) for sname in colors_databases]
        plt.legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))
      if a == 0:
        if z == 0: tname = 'All species'
        else: tname = 'Bacteria only'
        plt.text(-0.15, 0.5, tname, ha='center', va='center', transform=both_ax[z][a].transAxes, rotation=90, fontweight='bold')
  return
  
def get_distance_database_confidence(distance1, distance2):
  plt.figure(figsize=(40,16))
  axes = [plt.subplot(2,5,1), plt.subplot(2,5,2), plt.subplot(2,5,3), plt.subplot(2,5,4), plt.subplot(2,5,5)]
  axes2 = [plt.subplot(2,5,6), plt.subplot(2,5,7), plt.subplot(2,5,8), plt.subplot(2,5,9), plt.subplot(2,5,10)]
  both_ax = [axes, axes2]
  for z in range(len(both_ax)):
    this_distance = [distance1, distance2][z]
    for a in range(len(names_in_files)):
      #if a != 1: continue
      plt.sca(both_ax[z][a])
      fn = names_in_files[a]
      plt.title(sample_names[a], fontweight='bold')
      names = [name for name in this_distance.columns if fn in name]
      count = 0
      x_name, x_loc = [], []
      for c in range(len(confidence)):
        db = 'refseq'
        conf = confidence[c]
        for n in names:
          cont = False
          if db in n and float(n.split('_')[-1]) == conf:
            try:
              other_df = n.replace('refseq', 'gtdbh')
              this_conf = this_distance.loc[n, other_df]
            except:
              try:
                for n1 in names:
                  if 'gtdbh' in n1 and float(n1.split('_')[-1]) == conf:
                    this_conf = this_distance.loc[n, n1]
                    break
              except:
                cont = True
            if cont: continue
            pltx = np.random.normal(count, scale=0.2, size=1)
            plt.scatter(pltx, this_conf, color='gray', s=10, alpha=0.3)
        count += 1
        x_name.append(format(confidence[c], '.2f')), x_loc.append(count-1)
      plt.xticks(x_loc, x_name, rotation=90)
      plt.ylabel('Distance')
      if a == 0:
        if z == 0: tname = 'All species'
        else: tname = 'Bacteria only'
        plt.text(-0.15, 0.5, tname, ha='center', va='center', transform=both_ax[z][a].transAxes, rotation=90, fontweight='bold')
  return
```

In each of these, the top row is for all species while the second row is for bacteria only.

### All samples {.tabset}

#### Bray-Curtis

PCoA:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_None_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_None_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA all samples bray curtis.png', dpi=600, bbox_inches='tight')
```

Distance between confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_reads_bc.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_bc.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_confidence_threshold(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between confidence parameters bray curtis.png', dpi=600, bbox_inches='tight')
```

Distance between databases at different confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_reads_bc.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_bc.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_database_confidence(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between databases at different confidence parameters bray curtis.png', dpi=600, bbox_inches='tight')
```

#### Bray-Curtis relative abundance

PCoA:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_None_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_None_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA all samples bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

Distance between confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_reads_rel_abun_bc.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_rel_abun_bc.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_confidence_threshold(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between confidence parameters bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

Distance between databases at different confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_reads_rel_abun_bc.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_rel_abun_bc.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_database_confidence(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between databases at different confidence parameters bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

#### Aitchison (Euclidean on CLR)

PCoA:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_None_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_None_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA all samples aitchison.png', dpi=600, bbox_inches='tight')
```

Distance between confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_reads_clr_eucl.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_clr_eucl.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_confidence_threshold(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between confidence parameters aitchison.png', dpi=600, bbox_inches='tight')
```

Distance between databases at different confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_reads_clr_eucl.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_clr_eucl.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_database_confidence(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between databases at different confidence parameters aitchison.png', dpi=600, bbox_inches='tight')
```

### All samples (refseq only) {.tabset}

#### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_refseq_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_refseq_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA refseq bray curtis.png', dpi=600, bbox_inches='tight')
```

#### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_refseq_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_refseq_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA refseq bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

#### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_refseq_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_refseq_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA refseq aitchison.png', dpi=600, bbox_inches='tight')
```

### All samples (GTDB-human only) {.tabset}

#### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_gtdbh_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_gtdbh_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA GTDB-human bray curtis.png', dpi=600, bbox_inches='tight')
```

#### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_gtdbh_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_gtdbh_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA GTDB-human bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

#### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_gtdbh_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_gtdbh_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA GTDBH-human aitchison.png', dpi=600, bbox_inches='tight')
```

### Split by sample type {.tabset}

#### ZymoMock

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_ZymoMock_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_ZymoMock_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA ZymoMock bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_ZymoMock_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_ZymoMock_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA ZymoMock bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_ZymoMock_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_ZymoMock_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA ZymoMock aitchison.png', dpi=600, bbox_inches='tight')
```

#### PGPC blood

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_PGPC_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_PGPC_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA PGPC bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_PGPC_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_PGPC_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA PGPC bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_PGPC_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_PGPC_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA PGPC aitchison.png', dpi=600, bbox_inches='tight')
```

#### Human Samples

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_HumanSample_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_HumanSample_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA Human Samples bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_HumanSample_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_HumanSample_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA Human Samples bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_HumanSample_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_HumanSample_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA Human Samples aitchison.png', dpi=600, bbox_inches='tight')
```

#### Quarry Soils

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_QuarrySoils_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_QuarrySoils_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA Quarry Soils bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_QuarrySoils_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_QuarrySoils_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA Quarry Soils bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_QuarrySoils_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_QuarrySoils_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA Quarry Soils aitchison.png', dpi=600, bbox_inches='tight')
```

#### COVID samples

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_COVID_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_COVID_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA COVID bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_COVID_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_COVID_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA COVID bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_COVID_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_COVID_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA COVID aitchison.png', dpi=600, bbox_inches='tight')
```

### Split by confidence threshold {.tabset}

#### Confidence=0

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.05

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.05_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.05_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.05 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.05_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.05_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.05 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.05_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.05_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.05 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.1

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.1_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.1_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.1 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.1_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.1_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.1 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.1_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.1_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.1 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.15

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.15_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.15_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.15 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.15_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.15_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.15 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.15_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.15_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.15 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.2

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.2_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.2_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.2 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.2_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.2_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.2 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.2_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.2_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.2 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.25

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.25_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.25_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.25 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.25_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.25_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.25 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.25_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.25_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.25 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.3

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.3_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.3_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.3 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.3_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.3_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.3 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.3_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.3_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.3 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.35

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.35_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.35_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.35 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.35_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.35_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.35 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.35_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.35_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.35 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.4

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.4_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.4_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.4 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.4_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.4_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.4 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.4_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.4_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.4 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.45

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.45_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.45_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.45 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.45_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.45_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.45 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.45_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.45_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.45 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.5

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.5_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.5_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.5 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.5_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.5_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.5 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.5_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.5_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.5 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.55

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.55_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.55_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.55 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.55_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.55_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.55 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.55_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.55_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.55 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.6

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.6_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.6_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.6 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.6_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.6_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.6 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.6_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.6_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.6 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.65

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.65_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.65_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.65 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.65_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.65_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.65 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.65_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.65_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.65 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.7

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.7_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.7_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.7 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.7_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.7_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.7 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.7_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.7_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.7 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.75

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.75_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.75_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.75 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.75_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.75_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.75 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.75_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.75_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.75 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.8

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.8_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.8_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.8 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.8_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.8_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.8 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.8_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.8_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.8 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.85

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.85_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.85_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.85 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.85_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.85_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.85 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.85_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.85_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.85 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.9

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.9_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.9_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.9 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.9_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.9_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.9 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.9_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.9_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.9 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.95

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.95_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.95_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.95 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.95_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.95_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.95 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_0.95_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_0.95_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 0.95 aitchison.png', dpi=600, bbox_inches='tight')
```

#### Confidence=0.1

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_1_all_reads_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_1_bacteria_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 1 bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_1_all_reads_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_1_bacteria_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 1 bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_1_all_reads_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_1_bacteria_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA confidence 1 aitchison.png', dpi=600, bbox_inches='tight')
```


## Beta diversity genus level {.tabset}

### All samples {.tabset}

#### Bray-Curtis

PCoA:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_None_all_genus_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_None_bacteria_genus_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA all samples genus bray curtis.png', dpi=600, bbox_inches='tight')
```

Distance between confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_genus_bc.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_genus_bc.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_confidence_threshold(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between confidence parameters genus bray curtis.png', dpi=600, bbox_inches='tight')
```

Distance between databases at different confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_genus_bc.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_genus_bc.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_database_confidence(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between databases at different confidence parameters genus bray curtis.png', dpi=600, bbox_inches='tight')
```

#### Bray-Curtis relative abundance

PCoA:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_None_all_genus_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_None_bacteria_genus_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA all samples genus bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

Distance between confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_genus_rel_abun_bc.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_genus_rel_abun_bc.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_confidence_threshold(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between confidence parameters genus bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

Distance between databases at different confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_genus_rel_abun_bc.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_genus_rel_abun_bc.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_database_confidence(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between databases at different confidence parameters genus bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

#### Aitchison (Euclidean on CLR)

PCoA:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_None_all_genus_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_None_bacteria_genus_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
#plt.savefig(folder+'figures/PCoA all samples genus aitchison.png', dpi=600, bbox_inches='tight')
```

Distance between confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_genus_clr_eucl.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_genus_clr_eucl.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_confidence_threshold(distance_all, distance_bac)
plt.show()
#plt.savefig(folder+'figures/Distance between confidence parameters genus aitchison.png', dpi=600, bbox_inches='tight')
```

Distance between databases at different confidence thresholds:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_all = popen('distance_all_genus_clr_eucl.df', 'pcoa')
distance_all = pd.DataFrame(distance_all[0], index=distance_all[1], columns=distance_all[1])
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()
distance_all = remove_masked(remove_covid_others(distance_all)).transpose()

distance_bac = popen('distance_bacteria_genus_clr_eucl.df', 'pcoa')
distance_bac = pd.DataFrame(distance_bac[0], index=distance_bac[1], columns=distance_bac[1])
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()
distance_bac = remove_masked(remove_covid_others(distance_bac)).transpose()

get_distance_database_confidence(distance_all, distance_bac)
plt.show()
plt.savefig(folder+'figures/Distance between databases at different confidence parameters genus aitchison.png', dpi=600, bbox_inches='tight')
```

### All samples (refseq only) {.tabset}

#### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_refseq_all_genus_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_refseq_bacteria_genus_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus refseq bray curtis.png', dpi=600, bbox_inches='tight')
```

#### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_refseq_all_genus_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_refseq_bacteria_genus_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus refseq bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

#### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_refseq_all_genus_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_refseq_bacteria_genus_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus refseq aitchison.png', dpi=600, bbox_inches='tight')
```

### All samples (GTDB-human only) {.tabset}

#### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_gtdbh_all_genus_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_gtdbh_bacteria_genus_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus GTDB-human bray curtis.png', dpi=600, bbox_inches='tight')
```

#### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_gtdbh_all_genus_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_gtdbh_bacteria_genus_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus GTDB-human bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

#### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_gtdbh_all_genus_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_gtdbh_bacteria_genus_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus GTDBH-human aitchison.png', dpi=600, bbox_inches='tight')
```

### Split by sample type {.tabset}

#### ZymoMock

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_ZymoMock_all_genus_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_ZymoMock_bacteria_genus_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus ZymoMock bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_ZymoMock_all_genus_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_ZymoMock_bacteria_genus_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus ZymoMock bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_ZymoMock_all_genus_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_ZymoMock_bacteria_genus_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus ZymoMock aitchison.png', dpi=600, bbox_inches='tight')
```

#### PGPC blood

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_PGPC_all_genus_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_PGPC_bacteria_genus_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus PGPC bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_PGPC_all_genus_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_PGPC_bacteria_genus_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus PGPC bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_PGPC_all_genus_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_PGPC_bacteria_genus_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus PGPC aitchison.png', dpi=600, bbox_inches='tight')
```

#### Human Samples

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_HumanSample_all_genus_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_HumanSample_bacteria_genus_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus Human Samples bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_HumanSample_all_genus_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_HumanSample_bacteria_genus_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus Human Samples bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_HumanSample_all_genus_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_HumanSample_bacteria_genus_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus Human Samples aitchison.png', dpi=600, bbox_inches='tight')
```

#### Quarry Soils

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_QuarrySoils_all_genus_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_QuarrySoils_bacteria_genus_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus Quarry Soils bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_QuarrySoils_all_genus_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_QuarrySoils_bacteria_genus_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus Quarry Soils bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_QuarrySoils_all_genus_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_QuarrySoils_bacteria_genus_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus Quarry Soils aitchison.png', dpi=600, bbox_inches='tight')
```

#### COVID samples

##### Bray-Curtis

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_COVID_all_genus_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_COVID_bacteria_genus_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus COVID bray curtis.png', dpi=600, bbox_inches='tight')
```

##### Bray-Curtis relative abundance

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_COVID_all_genus_rel_abun_bc.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_COVID_bacteria_genus_rel_abun_bc.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus COVID bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```

##### Aitchison (Euclidean on CLR)

```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa_all = popen('pcoa_filter_COVID_all_genus_clr_eucl.df', 'pcoa')
pcoa_bac = popen('pcoa_filter_COVID_bacteria_genus_clr_eucl.df', 'pcoa')
plt.figure(figsize=(25,16))
axes1 = [plt.subplot(231), plt.subplot(232), plt.subplot(233)]
axes2 = [plt.subplot(234), plt.subplot(235), plt.subplot(236)]
PC1_all, PC2_all, prop_explain_all = pcoa_all[0].samples.loc[:, 'PC1'].values, pcoa_all[0].samples.loc[:, 'PC2'].values, [pcoa_all[0].proportion_explained[0], pcoa_all[0].proportion_explained[1]]
PC1_bac, PC2_bac, prop_explain_bac = pcoa_bac[0].samples.loc[:, 'PC1'].values, pcoa_bac[0].samples.loc[:, 'PC2'].values, [pcoa_bac[0].proportion_explained[0], pcoa_bac[0].proportion_explained[1]]
plot_pca(PC1_all, PC2_all, prop_explain_all, list(pcoa_all[1]), axes=axes1, label='All species')
plot_pca(PC1_bac, PC2_bac, prop_explain_bac, list(pcoa_bac[1]), axes=axes2, legend=False, label='Bacteria only')

plt.tight_layout()
plt.show()
plt.savefig(folder+'figures/PCoA genus COVID aitchison.png', dpi=600, bbox_inches='tight')
```



## ZymoMock comparison

### Species level

```{python, results='hide', fig.keep='all'}
all_reads = popen('all_df_gtdb_rsc_combined.df')
all_reads_rem100 = popen('all_df_rem100.df')

keeping = []
for col in all_reads.columns:
  if 'ZymoMock' in col: keeping.append(col)

all_reads_zymo = all_reads.loc[:, keeping]
all_reads_rem100_zymo = all_reads_rem100.loc[:, keeping]
rename = {}
for col in all_reads_rem100_zymo.columns:
  rename[col] = col+'_rem100'

all_reads_rem100_zymo = all_reads_rem100_zymo.rename(columns=rename)
all_reads_zymo = pd.concat([all_reads_zymo, all_reads_rem100_zymo]).fillna(value=0)
all_reads_zymo = all_reads_zymo.groupby(by=all_reads_zymo.index, axis=0).sum()
mean_reads = np.mean(all_reads_zymo.sum(axis=0))
real_species, real_composition, real_V4V5 = ['Listeria monocytogenes', 'Pseudomonas aeruginosa', 'Bacillus subtilis', 'Escherichia coli', 'Salmonella enterica', 'Lactobacillus fermentum', 'Enterococcus faecalis', 'Staphylococcus aureus', 'Saccharomyces cerevisiae', 'Cryptococcus neoformans'], [12.6, 14.7, 12.3, 12.6, 12.1, 9.5, 12.6, 10.2, 1.9, 1.5], [4.1, 4.7, 9.6, 14.4, 13.1, 50, 2.2, 1.8, 0, 0]
V4V5 = [(num/100)*mean_reads for num in real_V4V5]
real = [(num/100)*mean_reads for num in real_composition]
real_df = pd.DataFrame([real_composition, real_V4V5, real, V4V5], columns=real_species, index=['Real composition', 'QIIME V4V5', 'Real composition x mean reads', 'QIIME V4V5 x mean reads']).transpose()

all_reads_zymo = pd.concat([all_reads_zymo, real_df]).fillna(value=0)
all_reads_zymo = all_reads_zymo.groupby(by=all_reads_zymo.index, axis=0).sum()
all_reads_zymo = all_reads_zymo[all_reads_zymo.max(axis=1) > 0]
zymo = pd.DataFrame(all_reads_zymo)
pdump(zymo, 'all_zymomock.df')

zymo = popen('all_zymomock.df')
zymo_rel_abun = pd.DataFrame(zymo.divide(zymo.sum(axis=0), axis=1).multiply(100))

zymo = popen('all_zymomock.df')
zymo_clr = pd.DataFrame(zymo)
zymo_clr[zymo_clr == 0] = 1
for col in zymo_clr.columns:
 zymo_clr.loc[:, col] = clr(zymo_clr.loc[:, col].values)
```

ZymoMock relative abundance:
```{python, results='hide', fig.keep='all', cache=TRUE}
plt.figure(figsize=(10,5))
ax1 = plt.subplot(211, frameon=False)
df = zymo_rel_abun.transpose()
X = df.iloc[0:].values
similarities = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))
pcoa_results = ordination.pcoa(similarities)
min_expect, min_V4V5 = 1, 1
ind_expect, ind_V4V5 = -1, -1
Z = ssd.squareform(similarities)
Z = hierarchy.linkage(Z, "ward")
mpl.rcParams['lines.linewidth'] = 2
hierarchy.set_link_color_palette(['k'])
dn = hierarchy.dendrogram(Z, above_threshold_color='k')
x_labels, locs, xlocs, labels = list(ax1.get_xticklabels()), list(ax1.get_xticks()), list(ax1.get_yticks()), [] 
for x in x_labels:
  labels.append(x.get_text())
order_names = [list(zymo_rel_abun.columns)[int(l)] for l in labels]
plt.xticks(locs, order_names, rotation=90)
plt.yticks([])

plt.show()
#plt.savefig(folder+'/figures/ZymoMock dendrogram relative abundance.png', dpi=600, bbox_inches='tight')
```

ZymoMock relative abundance (stringent filtering):
```{python, results='hide', fig.keep='all', cache=TRUE}
zymo_rel_abun_filt = zymo_rel_abun[zymo_rel_abun.max(axis=1) > 0.5]
plt.figure(figsize=(10,5))
ax1 = plt.subplot(211, frameon=False)
df = zymo_rel_abun_filt.transpose()
X = df.iloc[0:].values
similarities = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))
pcoa_results = ordination.pcoa(similarities)
min_expect, min_V4V5 = 1, 1
ind_expect, ind_V4V5 = -1, -1
Z = ssd.squareform(similarities)
Z = hierarchy.linkage(Z, "ward")
mpl.rcParams['lines.linewidth'] = 2
hierarchy.set_link_color_palette(['k'])
dn = hierarchy.dendrogram(Z, above_threshold_color='k')
x_labels, locs, xlocs, labels = list(ax1.get_xticklabels()), list(ax1.get_xticks()), list(ax1.get_yticks()), []
for x in x_labels:
  labels.append(x.get_text())
order_names = [list(zymo_rel_abun.columns)[int(l)] for l in labels]
plt.xticks(locs, order_names, rotation=90)
plt.yticks([])

plt.show()
#plt.savefig(folder+'/figures/ZymoMock dendrogram relative abundance filtered.png', dpi=600, bbox_inches='tight')
```

ZymoMock CLR:
```{python, results='hide', fig.keep='all', cache=TRUE}
plt.figure(figsize=(10,5))
ax1 = plt.subplot(211, frameon=False)
df = zymo_clr.transpose()
X = df.iloc[0:].values
similarities = np.nan_to_num(distance.cdist(X, X, 'euclidean'))
pcoa_results = ordination.pcoa(similarities)
min_expect, min_V4V5 = 1, 1
ind_expect, ind_V4V5 = -1, -1
Z = ssd.squareform(similarities)
Z = hierarchy.linkage(Z, "ward")
mpl.rcParams['lines.linewidth'] = 2
hierarchy.set_link_color_palette(['k'])
dn = hierarchy.dendrogram(Z, above_threshold_color='k')
x_labels, locs, xlocs, labels = list(ax1.get_xticklabels()), list(ax1.get_xticks()), list(ax1.get_yticks()), [] 
for x in x_labels:
  labels.append(x.get_text())
order_names = [list(zymo_clr.columns)[int(l)] for l in labels]
plt.xticks(locs, order_names, rotation=90)
plt.yticks([])

plt.show()
#plt.savefig(folder+'/figures/ZymoMock dendrogram CLR.png', dpi=600, bbox_inches='tight')
```

ZymoMock CLR with low abundance removed (equivalent of 0.5% = (mean reads/100)*0.5):
```{python, results='hide', fig.keep='all', cache=TRUE}
zymo = popen('all_zymomock.df')
zymo_clr = pd.DataFrame(zymo)
zymo_clr_filt = zymo_clr[zymo_clr.max(axis=1) > (mean_reads/100)*0.5]
zymo_clr_filt[zymo_clr_filt == 0] = 1
for col in zymo_clr_filt.columns:
 zymo_clr_filt.loc[:, col] = clr(zymo_clr_filt.loc[:, col].values)
print(zymo_clr_filt)

plt.figure(figsize=(10,5))
ax1 = plt.subplot(211, frameon=False)
df = zymo_clr_filt.transpose()
X = df.iloc[0:].values
similarities = np.nan_to_num(distance.cdist(X, X, 'euclidean'))
pcoa_results = ordination.pcoa(similarities)
min_expect, min_V4V5 = 1, 1
ind_expect, ind_V4V5 = -1, -1
Z = ssd.squareform(similarities)
Z = hierarchy.linkage(Z, "ward")
mpl.rcParams['lines.linewidth'] = 2
hierarchy.set_link_color_palette(['k'])
dn = hierarchy.dendrogram(Z, above_threshold_color='k')
x_labels, locs, xlocs, labels = list(ax1.get_xticklabels()), list(ax1.get_xticks()), list(ax1.get_yticks()), [] 
for x in x_labels:
  labels.append(x.get_text())
order_names = [list(zymo_clr.columns)[int(l)] for l in labels]
plt.xticks(locs, order_names, rotation=90)
plt.yticks([])

plt.show()
#plt.savefig(folder+'/figures/ZymoMock dendrogram CLR filtered.png', dpi=600, bbox_inches='tight')
```

### Genus level

```{python, results='hide', fig.keep='all'}
all_reads_rem100 = popen('all_genus_rem100.df')

keeping = []
for col in all_reads_rem100.columns:
  if 'ZymoMock' in col: keeping.append(col)

all_reads_zymo = all_reads_rem100.loc[:, keeping]
real_species, real_composition, real_V4V5 = ['Listeria', 'Pseudomonas', 'Bacillus', 'Escherichia', 'Salmonella', 'Lactobacillus', 'Enterococcus', 'Staphylococcus', 'Saccharomyces', 'Cryptococcus'], [12.6, 14.7, 12.3, 12.6, 12.1, 9.5, 12.6, 10.2, 1.9, 1.5], [4.1, 4.7, 9.6, 14.4, 13.1, 50, 2.2, 1.8, 0, 0]
real_df = pd.DataFrame([real_composition, real_V4V5], columns=real_species, index=['Real composition', 'QIIME V4V5']).transpose()

all_reads_zymo = pd.concat([all_reads_zymo, real_df]).fillna(value=0)
all_reads_zymo = all_reads_zymo.groupby(by=all_reads_zymo.index, axis=0).sum()
all_reads_zymo = all_reads_zymo[all_reads_zymo.max(axis=1) > 0]
zymo = pd.DataFrame(all_reads_zymo)
pdump(zymo, 'all_zymomock_genus.df')

zymo = popen('all_zymomock_genus.df')
zymo_rel_abun = pd.DataFrame(zymo.divide(zymo.sum(axis=0), axis=1).multiply(100))

zymo = popen('all_zymomock_genus.df')
zymo_clr = pd.DataFrame(zymo)
zymo_clr[zymo_clr == 0] = 1
for col in zymo_clr.columns:
 zymo_clr.loc[:, col] = clr(zymo_clr.loc[:, col].values)
```

ZymoMock relative abundance:
```{python, results='hide', fig.keep='all', cache=TRUE}
plt.figure(figsize=(10,5))
ax1 = plt.subplot(211, frameon=False)
df = zymo_rel_abun.transpose()
X = df.iloc[0:].values
similarities = np.nan_to_num(distance.cdist(X, X, 'braycurtis'))
pcoa_results = ordination.pcoa(similarities)
min_expect, min_V4V5 = 1, 1
ind_expect, ind_V4V5 = -1, -1
Z = ssd.squareform(similarities)
Z = hierarchy.linkage(Z, "ward")
mpl.rcParams['lines.linewidth'] = 2
hierarchy.set_link_color_palette(['k'])
dn = hierarchy.dendrogram(Z, above_threshold_color='k')
x_labels, locs, xlocs, labels = list(ax1.get_xticklabels()), list(ax1.get_xticks()), list(ax1.get_yticks()), [] 
for x in x_labels:
  labels.append(x.get_text())
order_names = [list(zymo_rel_abun.columns)[int(l)] for l in labels]
plt.xticks(locs, order_names, rotation=90)
plt.yticks([])

plt.show()
#plt.savefig(folder+'/figures/ZymoMock dendrogram relative abundance genus.png', dpi=600, bbox_inches='tight')
```

ZymoMock CLR:
```{python, results='hide', fig.keep='all', cache=TRUE}
plt.figure(figsize=(10,5))
ax1 = plt.subplot(211, frameon=False)
df = zymo_clr.transpose()
X = df.iloc[0:].values
similarities = np.nan_to_num(distance.cdist(X, X, 'euclidean'))
pcoa_results = ordination.pcoa(similarities)
min_expect, min_V4V5 = 1, 1
ind_expect, ind_V4V5 = -1, -1
Z = ssd.squareform(similarities)
Z = hierarchy.linkage(Z, "ward")
mpl.rcParams['lines.linewidth'] = 2
hierarchy.set_link_color_palette(['k'])
dn = hierarchy.dendrogram(Z, above_threshold_color='k')
x_labels, locs, xlocs, labels = list(ax1.get_xticklabels()), list(ax1.get_xticks()), list(ax1.get_yticks()), [] 
for x in x_labels:
  labels.append(x.get_text())
order_names = [list(zymo_clr.columns)[int(l)] for l in labels]
plt.xticks(locs, order_names, rotation=90)
plt.yticks([])

plt.show()
#plt.savefig(folder+'/figures/ZymoMock dendrogram CLR genus.png', dpi=600, bbox_inches='tight')
```

## Blood samples masked, GTDB no human and QUAST comparison

Beta diversity calculations (I am filtering all based on 0.1% relative abundance and will later do this with the QUAST abundance too and doing this on only bacteria, as I did previously for the blood samples):
```{python, eval=FALSE}
from skbio.stats.composition import clr
import os
import pandas as pd
import pickle
from scipy.spatial import distance
from skbio.stats import ordination
import numpy as np

folder = os.getcwd()
folder_pickle = folder+'/pickle/'

#function to open pickle file
def popen(name):
  name = folder_pickle+name
  with open(name, 'rb') as f:
    obj = pickle.load(f)
  return obj

#function to close pickle file
def pdump(obj, name):
  name = folder_pickle+name
  with open(name, 'wb') as f:
    pickle.dump(obj, f)
  return

bacteria_reads_rel_abun = popen('Bacteria_reads.df')
bacteria_reads_rel_abun = bacteria_reads_rel_abun.divide(bacteria_reads_rel_abun.sum(axis=0), axis=1).multiply(100)
bacteria_reads_rel_abun = bacteria_reads_rel_abun[bacteria_reads_rel_abun.max(axis=1) > 0.1]
pdump(bacteria_reads_rel_abun, 'Bacteria_reads_rel_abun.df')
taxa = bacteria_reads_rel_abun.index.values

bacteria_reads_clr = popen('Bacteria_reads.df').loc[taxa, :]
bacteria_reads_clr[bacteria_reads_clr == 0] = 1
for col in bacteria_reads_clr.columns:
  bacteria_reads_clr.loc[:, col] = clr(bacteria_reads_clr.loc[:, col].values)

pdump(bacteria_reads_clr, 'Bacteria_reads_clr.df')

bacteria_reads = popen('Bacteria_reads.df').loc[taxa, :]

distances_to_get = [bacteria_reads, bacteria_reads_rel_abun, bacteria_reads_clr]
distance_metrics = ['braycurtis', 'braycurtis', 'euclidean']
save_names = ['distance_bacteria_bc.df', 'distance_bacteria_rel_abun_bc.df', 'distance_bacteria_clr_eucl.df']
distances_calculated = []

for d in range(len(distances_to_get)):
  print(save_names[d])
  df = distances_to_get[d].transpose()
  X = df.iloc[0:].values
  similarities = np.nan_to_num(distance.cdist(X, X, distance_metrics[d]))
  pdump([similarities, df.index.values], save_names[d])
  similarities_df = pd.DataFrame(similarities, index=df.index, columns=df.index)
  distances_calculated.append(similarities)
  print('Finished\n')

distance_for_pcoa = ['distance_bacteria_bc.df', 'distance_bacteria_rel_abun_bc.df', 'distance_bacteria_clr_eucl.df']
all_pcoa = []

for a in range(len(distance_for_pcoa)):
  list_sim = popen(distance_for_pcoa[a])
  df_sim = pd.DataFrame(list_sim[0], index=list_sim[1], columns=list_sim[1])
  X = df_sim.iloc[0:].values
  names = df_sim.index.values
  pcoa_results = ordination.pcoa(X)
  sname = distance_for_pcoa[a].split('.')[0].replace('distance', '')+'.df'
  pdump([pcoa_results, names], 'pcoa_'+sname)
  all_pcoa.append([pcoa_results, sname, names])
```

### Number of reads

```{python, results='hide', fig.keep='all', cache=TRUE}
bacteria_sum = popen('Bacteria_reads_sum.df', blood='blood')
all_sum = popen('sum_reads_classified.df', blood='blood')

conf = [format(c, '.2f') for c in confidence]

plt.figure(figsize=(20,10))
databases = ['refseq', 'gtdb_', 'gtdb-nohuman']
database_names = ['RefSeq', 'GTDB + human', 'GTDB']
axes = [[plt.subplot(231), plt.subplot(232), plt.subplot(233)], [plt.subplot(234), plt.subplot(235), plt.subplot(236)]]

for a in range(len(axes)):
  for b in range(len(databases)):
    for c in range(len(conf)):
      for col in bacteria_sum.columns:
        plotting = False
        if a == 1: 
          if 'masked' in col and databases[b] in col and conf[c] in col: plotting = True
        elif 'masked' not in col and databases[b] in col and conf[c] in col: plotting = True
        if plotting:
          c1, c2 = np.random.normal(c-0.2, scale=0.05, size=1), np.random.normal(c+0.2, scale=0.05, size=1)
          axes[a][b].scatter(c1, bacteria_sum.loc['Sum', col], color=colors_domain[1], s=5)
          others = all_sum.loc['Sum', col]-bacteria_sum.loc['Sum', col]
          axes[a][b].scatter(c2, others, color='gray', s=5)
    plt.sca(axes[a][b])
    plt.semilogy()
    plt.ylim([10**1, 10**8])
    plt.xticks([c for c in range(len(conf))], conf, rotation=90)
    if b == 0: 
      plt.ylabel('Number of reads')
      if a == 0: plt.text(-0.15, 0.5, 'Not masked', rotation=90, ha='center', va='center', transform=axes[a][b].transAxes, fontweight='bold')
      else: plt.text(-0.15, 0.5, 'Masked', rotation=90, ha='center', va='center', transform=axes[a][b].transAxes, fontweight='bold')
    if a == 1 and b == 1: plt.xlabel('Confidence threshold')
    if a == 0: plt.title(database_names[b], fontweight='bold')
    plt.xlim([-0.5, 20.5])

handles = [Patch(facecolor=colors_domain[1], edgecolor='k', label=domain_names[1]), Patch(facecolor='gray', edgecolor='k', label='Others')]
axes[0][2].legend(handles=handles, loc='upper left', bbox_to_anchor=(1.02, 1.02))
  
plt.show()
#plt.savefig(folder+'figures/blood reads classified.png', dpi=600, bbox_inches='tight')
```

### Number of bacterial reads with different filtering

Filter:
```{python, eval=FALSE}
bacteria = popen('Bacteria_reads.df', blood='blood')
bacteria_rel_abun = popen('Bacteria_reads_rel_abun.df', blood='blood')
bacteria_reduced_01 = bacteria.loc[bacteria_rel_abun.index.values, :]
pdump(bacteria_reduced_01, 'bacteria_reduced_01.df', blood='blood')
bacteria_reduced_01_sum = pd.DataFrame(bacteria_reduced_01.sum(axis=0)).transpose().rename(index={0:'Sum'})
pdump(bacteria_reduced_01_sum, 'bacteria_reduced_01_sum.df', blood='blood')
bacteria_reduced_01_sum_species = popen('bacteria_reduced_01.df', blood='blood')
bacteria_reduced_01_sum_species[bacteria_reduced_01_sum_species > 1] = 1
bacteria_reduced_01_sum_species = pd.DataFrame(bacteria_reduced_01_sum_species.sum(axis=0)).transpose().rename(index={0:'Sum'})
pdump(bacteria_reduced_01_sum_species, 'bacteria_reduced_01_sum_species.df', blood='blood')
with open('/Users/robynwright/Documents/OneDrive/Langille_Lab_postdoc/Human_blood_metagenome/analysis/above_lim2.df', 'rb') as f:
    quast_lim2 = pickle.load(f)
species_quast_lim2 = list(quast_lim2.index.values)

with open('/Users/robynwright/Documents/OneDrive/Langille_Lab_postdoc/Human_blood_metagenome/analysis/above_lim1.df', 'rb') as f:
    quast_lim1 = pickle.load(f)
species_quast_lim1 = list(quast_lim1.index.values)

bacteria_reduced_02 = popen('Bacteria_reads.df', blood='blood')
bacteria_reduced_02 = bacteria_reduced_02.loc[species_quast_lim2, :]
pdump(bacteria_reduced_02, 'bacteria_reduced_02.df', blood='blood')
bacteria_reduced_02_sum = pd.DataFrame(bacteria_reduced_02.sum(axis=0)).transpose().rename(index={0:'Sum'})
pdump(bacteria_reduced_02_sum, 'bacteria_reduced_02_sum.df', blood='blood')
bacteria_reduced_02_sum_species = popen('bacteria_reduced_02.df', blood='blood')
bacteria_reduced_02_sum_species[bacteria_reduced_02_sum_species > 1] = 1
bacteria_reduced_02_sum_species = pd.DataFrame(bacteria_reduced_02_sum_species.sum(axis=0)).transpose().rename(index={0:'Sum'})
pdump(bacteria_reduced_02_sum_species, 'bacteria_reduced_02_sum_species.df', blood='blood')

bacteria_reduced_03 = popen('Bacteria_reads.df', blood='blood')
bacteria_reduced_03 = bacteria_reduced_03.loc[species_quast_lim1, :]
pdump(bacteria_reduced_03, 'bacteria_reduced_03.df', blood='blood')
bacteria_reduced_03_sum = pd.DataFrame(bacteria_reduced_03.sum(axis=0)).transpose().rename(index={0:'Sum'})
pdump(bacteria_reduced_03_sum, 'bacteria_reduced_03_sum.df', blood='blood')
bacteria_reduced_03_sum_species = popen('bacteria_reduced_03.df', blood='blood')
bacteria_reduced_03_sum_species[bacteria_reduced_03_sum_species > 1] = 1
bacteria_reduced_03_sum_species = pd.DataFrame(bacteria_reduced_03_sum_species.sum(axis=0)).transpose().rename(index={0:'Sum'})
pdump(bacteria_reduced_03_sum_species, 'bacteria_reduced_03_sum_species.df', blood='blood')
```

Plot:
```{python, results='hide', fig.keep='all', cache=TRUE}
conf = [format(c, '.2f') for c in confidence]
plt.figure(figsize=(40,20))
databases = [['refseq'], ['refseq', 'masked'], ['gtdb_'], ['gtdb_', 'masked'], ['gtdb-nohuman'], ['gtdb-nohuman', 'masked']]
database_names = ['RefSeq', 'RefSeq Masked', 'GTDB + human', 'GTDB + human Masked', 'GTDB', 'GTDB Masked']
filter_name = ['All bacteria', 'Bacteria >0.1% relative abundance', 'Bacteria >0.005% QUAST genome coverage', 'Bacteria >0.01% QUAST genome coverage']
axes = [[plt.subplot(4,6,1), plt.subplot(4,6,2), plt.subplot(4,6,3), plt.subplot(4,6,4), plt.subplot(4,6,5), plt.subplot(4,6,6)], [plt.subplot(4,6,7), plt.subplot(4,6,8), plt.subplot(4,6,9), plt.subplot(4,6,10), plt.subplot(4,6,11), plt.subplot(4,6,12)], [plt.subplot(4,6,13), plt.subplot(4,6,14), plt.subplot(4,6,15), plt.subplot(4,6,16), plt.subplot(4,6,17), plt.subplot(4,6,18)], [plt.subplot(4,6,19), plt.subplot(4,6,20), plt.subplot(4,6,21), plt.subplot(4,6,22), plt.subplot(4,6,23), plt.subplot(4,6,24)]]
axes_twin = []
for a in range(len(axes)):
  this_ax = []
  for b in range(len(axes[a])):
    this_ax.append(plt.twinx(axes[a][b]))
  axes_twin.append(this_ax)

df_reads = [popen('Bacteria_reads_sum.df', blood='blood'), popen('bacteria_reduced_01_sum.df', blood='blood'), popen('bacteria_reduced_02_sum.df', blood='blood'), popen('bacteria_reduced_03_sum.df', blood='blood')]
df_species = [popen('Bacteria_species_sum.df', blood='blood'), popen('bacteria_reduced_01_sum_species.df', blood='blood'), popen('bacteria_reduced_02_sum_species.df', blood='blood'), popen('bacteria_reduced_03_sum_species.df', blood='blood')]

for a in range(len(axes)):
  for b in range(len(axes[a])):
    #if a != 0 or b > 2: continue
    reads, reads_upper, reads_lower, species, species_upper, species_lower = [], [], [], [], [], []
    x = []
    for c in range(len(conf)):
      x.append(c)
      this_reads, this_species = [], []
      for col in df_reads[a].columns:
        if len(databases[b]) == 1:
          if conf[c] in col and databases[b][0] in col and 'masked' not in col:
            this_reads.append(df_reads[a].loc['Sum', col])
            this_species.append(df_species[a].loc['Sum', col])
        else:
          if conf[c] in col and databases[b][0] in col and databases[b][1] in col:
            this_reads.append(df_reads[a].loc['Sum', col])
            this_species.append(df_species[a].loc['Sum', col])
      read_mean, species_mean, read_std, species_std = np.mean(this_reads), np.mean(this_species), np.std(this_reads), np.std(this_species)
      reads.append(read_mean), reads_upper.append(read_mean+read_std), reads_lower.append(read_mean-read_std)
      species.append(species_mean), species_upper.append(species_mean+species_std), species_lower.append(species_mean-species_std)
    axes[a][b].plot(x, reads, color='royalblue')
    axes[a][b].fill_between(x, reads_upper, reads_lower, color='royalblue', alpha=0.2)
    axes_twin[a][b].plot(x, species, color='firebrick')
    axes_twin[a][b].fill_between(x, species_upper, species_lower, color='firebrick', alpha=0.2)
    plt.sca(axes[a][b])
    plt.semilogy()
    plt.ylim([10**3, 10**7.5])
    plt.xlim([-0.5, 20.5])
    if b == 0:
      plt.ylabel('Number of reads')
      plt.text(-0.15, 0.5, filter_name[a], ha='center', va='center', fontweight='bold', rotation=90, transform=axes[a][b].transAxes)
    if a < 3:
      plt.xticks([])
    else:
      plt.xticks(x, conf, rotation=90)
      plt.xlabel('Confidence threshold')
    plt.sca(axes_twin[a][b])
    if b < 2 and a == 0: plt.ylim([0, 3500])
    elif b < 2 and a < 3: plt.ylim([0, 500])
    elif b < 2: plt.ylim([0, 100])
    elif b > 3 and a == 0: plt.ylim([0, 15000])
    elif a == 0: plt.ylim([0, 8000])
    elif a == 1: plt.ylim([0, 1000])
    elif  a == 2: plt.ylim([0, 500])
    else: plt.ylim([0, 100])
    if b == 5:
      plt.ylabel('Number of species')
    if a == 0:
      plt.title(database_names[b], fontweight='bold')
    axes[a][b].yaxis.label.set_color('royalblue')
    axes[a][b].spines['left'].set_color('royalblue')
    axes_twin[a][b].spines['left'].set_color('royalblue')
    axes_twin[a][b].yaxis.label.set_color('firebrick')
    axes_twin[a][b].spines['right'].set_color('firebrick')

plt.show()
plt.savefig(folder+'/figures/blood species and reads.png', dpi=600, bbox_inches='tight')
```

### Distance

#### All confidence thresholds

Each of these is calculated for bacteria only and only with 0.1% abundance filtering.

Function:
```{python, results='hide', fig.keep='all'}
def plot_panels_pcoa(PC1, PC2, prop_explain, names):
  db_colors = {'refseq':'#E67E22', 'masked-refseq':'#F7DC6F', 'gtdb':'#1A5276', 'masked-gtdb':'#5499C7', 'gtdb-nohuman':'#7302AB', 'masked-gtdb-nohuman':'#DEA3FC'}
  db_labels = {'refseq':'RefSeq', 'masked-refseq':'RefSeq masked', 'gtdb':'GTDB + human', 'masked-gtdb':'GTDB + human masked', 'gtdb-nohuman':'GTDB', 'masked-gtdb-nohuman':'GTDB masked'}
  dbs = ['refseq', 'gtdb', 'gtdb-nohuman', 'masked-refseq', 'masked-gtdb', 'masked-gtdb-nohuman']
  conf = [format(c, '.2f') for c in confidence]
  colors_conf_dict = {}
  for c in range(len(conf)):
    colors_conf_dict[conf[c]] = colors_conf[c]
  plt.figure(figsize=(20,20))
  ax1, ax2 = plt.subplot2grid((3,24), (0,0), colspan=11), plt.subplot2grid((3,24), (0,12), colspan=11)
  ax3, ax4, ax5, ax6, ax7, ax8 = plt.subplot(334), plt.subplot(335), plt.subplot(336), plt.subplot(337), plt.subplot(338), plt.subplot(339)
  axes = [ax1, ax2, ax3, ax4, ax5, ax6, ax7, ax8]
  labels = ['Coloured by database', 'Coloured by confidence', 'RefSeq coloured by confidence', 'GTDB + human coloured by confidence', 'GTDB coloured by confidence', 'RefSeq masked coloured by confidence', 'GTDB + human masked coloured by confidence', 'GTDB masked coloured by confidence']
  for a in range(len(axes)):
    plt.sca(axes[a])
    plt.title(labels[a])
    plt.xlabel('PCoA1 ('+format(prop_explain[0]*100, '.2f')+'%)')
    plt.ylabel('PCoA2 ('+format(prop_explain[1]*100, '.2f')+'%)')
  handles_conf = [Patch(facecolor=colors_conf_dict[c], edgecolor='k', label=c) for c in colors_conf_dict]
  ax2.legend(handles=handles_conf, loc='upper left', bbox_to_anchor=(1.02, 1.02))
  handles_db = [Patch(facecolor=db_colors[d], edgecolor='k', label=db_labels[d]) for d in db_colors]
  ax1.legend(handles=handles_db, loc='best')
  for a in range(len(PC1)):
    color = db_colors[pcoa[1][a].split('_')[2]]
    t = [0.1, 0.1, 0.1, 0.1, 0.1, 0.1]
    for d in range(len(dbs)):
      if pcoa[1][a].split('_')[2] == dbs[d]:
        t[d] = 1
    color_conf = colors_conf_dict[pcoa[1][a].split('_')[-1]]
    ax1.scatter(PC1[a], PC2[a], color=color, alpha=0.8, s=5, linewidths=0.1, edgecolor='gray')
    ax2.scatter(PC1[a], PC2[a], color=color_conf, alpha=0.8, s=5)
    ax3.scatter(PC1[a], PC2[a], color=color_conf, alpha=t[0], s=5)
    ax4.scatter(PC1[a], PC2[a], color=color_conf, alpha=t[1], s=5)
    ax5.scatter(PC1[a], PC2[a], color=color_conf, alpha=t[2], s=5)
    ax6.scatter(PC1[a], PC2[a], color=color_conf, alpha=t[3], s=5)
    ax7.scatter(PC1[a], PC2[a], color=color_conf, alpha=t[4], s=5)
    ax8.scatter(PC1[a], PC2[a], color=color_conf, alpha=t[5], s=5)
  return
```

Bray-Curtis raw abundance:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa = popen('pcoa_bacteria_bc.df', blood='blood')
PC1_this, PC2_this, prop_explain_this = pcoa[0].samples.loc[:, 'PC1'].values, pcoa[0].samples.loc[:, 'PC2'].values, [pcoa[0].proportion_explained[0], pcoa[0].proportion_explained[1]]
plot_panels_pcoa(PC1_this, PC2_this, prop_explain_this, pcoa[1])
plt.show()
plt.savefig(folder+'/figures/blood distance bray curtis raw.png', dpi=600, bbox_inches='tight')
```

Aitchison:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa = popen('pcoa_bacteria_clr_eucl.df', blood='blood')
PC1_this, PC2_this, prop_explain_this = pcoa[0].samples.loc[:, 'PC1'].values, pcoa[0].samples.loc[:, 'PC2'].values, [pcoa[0].proportion_explained[0], pcoa[0].proportion_explained[1]]
plot_panels_pcoa(PC1_this, PC2_this, prop_explain_this, pcoa[1])
plt.show()
#plt.savefig(folder+'/figures/blood distance Aitchison.png', dpi=600, bbox_inches='tight')
```

Bray-Curtis relative abundance:
```{python, results='hide', fig.keep='all', cache=TRUE}
pcoa = popen('pcoa_bacteria_rel_abun_bc.df', blood='blood')
PC1_this, PC2_this, prop_explain_this = pcoa[0].samples.loc[:, 'PC1'].values, pcoa[0].samples.loc[:, 'PC2'].values, [pcoa[0].proportion_explained[0], pcoa[0].proportion_explained[1]]
plot_panels_pcoa(PC1_this, PC2_this, prop_explain_this, pcoa[1])
plt.show()
#plt.savefig(folder+'/figures/blood distance bray curtis relative abundance.png', dpi=600, bbox_inches='tight')
```


#### Distance between samples at one confidence threshold

Function:
```{python, results='hide', fig.keep='all'}
def distance_single_threshold(reads, conf, dist='euclidean'):
  # db_colors = {'refseq':'#E67E22', 'masked-refseq':'#F7DC6F', 'gtdb':'#1A5276', 'masked-gtdb':'#5499C7', 'gtdb-nohuman':'#7302AB', 'masked-gtdb-nohuman':'#DEA3FC'}
  db_labels = {'refseq':'RefSeq', 'masked-refseq':'RefSeq masked', 'gtdb':'GTDB + human', 'masked-gtdb':'GTDB + human masked', 'gtdb-nohuman':'GTDB', 'masked-gtdb-nohuman':'GTDB masked'}
  dbs = ['refseq', 'gtdb', 'gtdb-nohuman', 'masked-refseq', 'masked-gtdb', 'masked-gtdb-nohuman']
  colormap = mpl.cm.get_cmap('plasma', 256)
  norm = mpl.colors.Normalize(vmin=1, vmax=71)
  m1 = mpl.cm.ScalarMappable(norm=norm, cmap=colormap)
  reads = popen(reads, blood='blood')
  keeping = []
  for col in bacteria_reads.columns:
    if conf in col:
      keeping.append(col)
  reads = reads.loc[:, keeping]
  reads = reads[reads.max(axis=1) > 0]
  if dist == 'euclidean':
    reads[reads == 0] = 1
    for col in reads.columns:
      reads.loc[:, col] = clr(reads.loc[:, col].values)
  
  plt.figure(figsize=(20,10))
  axes = [plt.subplot(231), plt.subplot(232),  plt.subplot(233), plt.subplot(234), plt.subplot(235), plt.subplot(236)]
  for a in range(len(dbs)):
    this_db = []
    for col in reads.columns:
      if dbs[a] == col.split('_')[2]: this_db.append(col)
    this_db = pd.DataFrame(reads.loc[:, this_db]).transpose()
    X = this_db.iloc[0:].values
    similarities = np.nan_to_num(distance.cdist(X, X, dist))
    pcoa_results = ordination.pcoa(similarities)
    PC1, PC2 = pcoa_results.samples.loc[:, 'PC1'].values, pcoa_results.samples.loc[:, 'PC2'].values
    prop_explain1, prop_explain2 = pcoa_results.proportion_explained[0], pcoa_results.proportion_explained[1]
    names = this_db.index.values
    for b in range(len(PC1)):
      color = m1.to_rgba(int(names[b].split('_')[1]))
      axes[a].scatter(PC1[b], PC2[b], color=color, s=80, alpha=0.5, edgecolor='gray')
      axes[a].text(PC1[b], PC2[b], str(int(names[b].split('_')[1])), ha='center', va='center', fontsize=4)
    axes[a].set_xlabel('PCoA1 ('+format(prop_explain1*100, '.2f')+'%)'), axes[a].set_ylabel('PCoA2 ('+format(prop_explain2*100, '.2f')+'%)')
    axes[a].set_title(db_labels[dbs[a]], fontweight='bold')
  
  return
```

Aitchison:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_single_threshold('bacteria_reduced_01.df', '0.50')
plt.show()
#plt.savefig(folder+'/figures/blood confidence 0.50 Aitchison.png', dpi=600, bbox_inches='tight')
```

Bray-Curtis raw:
```{python, results='hide', fig.keep='all', cache=TRUE}
distance_single_threshold('bacteria_reduced_01.df', '0.50', dist='braycurtis')
plt.show()
#plt.savefig(folder+'/figures/blood confidence 0.50 Bray-Curtis.png', dpi=600, bbox_inches='tight')
```

## COVID samples

Function:
```{python, results='hide', fig.keep='all'}
def plot_covid(reads, dist):
  reads = popen(reads)
  keeping = []
  for col in reads.columns:
    if 'COVID' in col and '0.5' in col and '0.55' not in col:
      keeping.append(col)
  print(keeping)
  covid = reads.loc[:, keeping]
  covid = covid[covid.max(axis=1) > 0]
  if dist == 'euclidean':
      covid[covid == 0] = 1
      for col in covid.columns:
        covid.loc[:, col] = clr(covid.loc[:, col].values)
  handles = [Patch(facecolor='#2471A3', label='Negative nasopharyngeal'), 
            Patch(facecolor='w', edgecolor='#2471A3', label='Negative oropharyngeal'), 
            Patch(facecolor='#E74C3C', label='Positive nasopharyngeal'), 
            Patch(facecolor='w', edgecolor='#E74C3C', label='Positive oropharyngeal'), 
            Line2D([0], [0], marker='<', color='w', label='TNA', markerfacecolor='k', markersize=10), 
            Line2D([0], [0], marker='o', color='w', label='cDNA', markerfacecolor='k', markersize=10)]
  plt.figure(figsize=(14,6))
  ax1, ax2 = plt.subplot(121), plt.subplot(122)
  axes = [ax1, ax2]
  dbs = ['refseq', 'gtdbh']
  titles = ['RefSeq', 'GTDB + human']
  for d in range(len(dbs)):
    keeping, colors = [], []
    for col in covid.columns:
      if dbs[d] in col: keeping.append(col)
    this_df = pd.DataFrame(covid.loc[:, keeping]).transpose()
    colors, shapes, edge = [], [], []
    for col in this_df.index.values:
      color = '#E74C3C'
      if 'neg' in col: color = '#2471A3'
      edge.append(color)
      if 'cDNA' in col:
        shapes.append('s')
        if 'N' in col.split('-')[2]: colors.append(color)
        else: colors.append('w')
      else:
        shapes.append('<')
        if 'N' in col.split('-')[1]: colors.append(color)
        else: colors.append('w')
    X = this_df.iloc[0:].values
    similarities = np.nan_to_num(distance.cdist(X, X, dist))
    pcoa_results = ordination.pcoa(similarities)
    PC1, PC2 = pcoa_results.samples.loc[:, 'PC1'].values, pcoa_results.samples.loc[:, 'PC2'].values
    prop_explain1, prop_explain2 = pcoa_results.proportion_explained[0], pcoa_results.proportion_explained[1]
    for a in range(len(PC1)):
      axes[d].scatter(PC1[a], PC2[a], color=colors[a], edgecolor=edge[a], marker=shapes[a])
    axes[d].set_xlabel('PCoA1 ('+format(prop_explain1*100, '.2f')+')%'), axes[d].set_ylabel('PCoA1 ('+format(prop_explain2*100, '.2f')+')%')
    axes[d].set_title(titles[d])
  return
```

Bray-Curtis:
```{python, results='hide', fig.keep='all', cache=TRUE}
plot_covid('Bacteria_reads_rem100.df', 'braycurtis')
plt.show()
#plt.savefig(folder+'/figures/COVID distance Bray-Curtis.png', dpi=600, bbox_inches='tight')
```

Aitchison:
```{python, results='hide', fig.keep='all', cache=TRUE}
plot_covid('Bacteria_reads_rem100.df', 'euclidean')
plt.show()
plt.savefig(folder+'/figures/COVID distance Aitchison.png', dpi=600, bbox_inches='tight')
```